// @backstage/plugin-catalog-react/dist/components/EntityDataTable/presets.esm.js
import { columnFactories } from './columnFactories.esm.js';

const entityColumnPresets = {
  component: {
    columns: [
      columnFactories.createEntityRefColumn({ defaultKind: "component" }),
      columnFactories.createOwnerColumn(),
      columnFactories.createSpecTypeColumn(),
      columnFactories.createSpecLifecycleColumn(),
      columnFactories.createMetadataDescriptionColumn()
    ],
    helpLink: "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-component"
  },
  resource: {
    columns: [
      columnFactories.createEntityRefColumn({ defaultKind: "resource" }),
      columnFactories.createOwnerColumn(),
      columnFactories.createSpecTypeColumn(),
      columnFactories.createSpecLifecycleColumn(),
      columnFactories.createMetadataDescriptionColumn()
    ],
    helpLink: "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-resource"
  },
  system: {
    columns: [
      columnFactories.createEntityRefColumn({ defaultKind: "system" }),
      columnFactories.createOwnerColumn(),
      columnFactories.createMetadataDescriptionColumn()
    ],
    helpLink: "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-system"
  },
  domain: {
    columns: [
      columnFactories.createEntityRefColumn({ defaultKind: "domain" }),
      columnFactories.createOwnerColumn(),
      columnFactories.createMetadataDescriptionColumn()
    ],
    helpLink: "https://backstage.io/docs/features/software-catalog/descriptor-format#kind-domain"
  }
};

export { entityColumnPresets };
