// @backstage/plugin-catalog-react/dist/testUtils/catalogApiMock.esm.js
import { createApiFactory } from '@backstage/frontend-plugin-api';
import { InMemoryCatalogClient } from '@backstage/catalog-client/testUtils';
import { catalogApiRef } from '@backstage/plugin-catalog-react';
import { createApiMock, attachMockApiFactory } from '@backstage/frontend-test-utils';

function catalogApiMock(options) {
  const instance = new InMemoryCatalogClient(options);
  return attachMockApiFactory(catalogApiRef, instance);
}
((catalogApiMock2) => {
  catalogApiMock2.factory = (options) => createApiFactory({
    api: catalogApiRef,
    deps: {},
    factory: () => new InMemoryCatalogClient(options)
  });
  catalogApiMock2.mock = createApiMock(catalogApiRef, () => ({
    getEntities: jest.fn(),
    getEntitiesByRefs: jest.fn(),
    queryEntities: jest.fn(),
    getEntityAncestors: jest.fn(),
    getEntityByRef: jest.fn(),
    removeEntityByUid: jest.fn(),
    refreshEntity: jest.fn(),
    getEntityFacets: jest.fn(),
    getLocations: jest.fn(),
    queryLocations: jest.fn(),
    streamLocations: jest.fn(),
    getLocationById: jest.fn(),
    getLocationByRef: jest.fn(),
    addLocation: jest.fn(),
    removeLocationById: jest.fn(),
    getLocationByEntity: jest.fn(),
    validateEntity: jest.fn(),
    analyzeLocation: jest.fn(),
    streamEntities: jest.fn(),
    streamEntityPages: jest.fn()
  }));
})(catalogApiMock || (catalogApiMock = {}));

export { catalogApiMock };
