// @backstage/backend-dev-utils/dist/ipcClient.cjs.js
'use strict';

const requestType = "@backstage/cli/channel/request";
const responseType = "@backstage/cli/channel/response";
const IPC_TIMEOUT_MS = 5e3;
class BackstageIpcClient {
  #messageId = 0;
  #sendMessage;
  #handlers = /* @__PURE__ */ new Map();
  /**
   * Creates a new client if we're in a child process with IPC and BACKSTAGE_CLI_CHANNEL is set.
   */
  static create() {
    const sendMessage = process.send?.bind(process);
    const client = sendMessage && process.env.BACKSTAGE_CLI_CHANNEL ? new BackstageIpcClient(sendMessage) : void 0;
    if (client) {
      process.on(
        "message",
        (message, _sendHandle) => client.handleMessage(message, _sendHandle)
      );
    }
    return client;
  }
  constructor(sendMessage) {
    this.#sendMessage = sendMessage;
  }
  handleMessage(message, _sendHandle) {
    const isResponse = (msg) => msg?.type === responseType;
    if (isResponse(message)) {
      this.#handlers.get(message.id)?.(message);
    }
  }
  /**
   * Send a request to the parent process and wait for a response.
   */
  async request(method, body) {
    return new Promise((resolve, reject) => {
      const id = this.#messageId++;
      const request = {
        type: requestType,
        id,
        method,
        body
      };
      let timeout = void 0;
      const responseHandler = (response) => {
        clearTimeout(timeout);
        timeout = void 0;
        this.#handlers.delete(id);
        if ("error" in response) {
          const error = new Error(response.error.message);
          if (response.error.name) {
            error.name = response.error.name;
          }
          reject(error);
        } else {
          resolve(response.body);
        }
      };
      timeout = setTimeout(() => {
        reject(new Error(`IPC request '${method}' with ID ${id} timed out`));
        this.#handlers.delete(id);
      }, IPC_TIMEOUT_MS);
      timeout.unref();
      this.#handlers.set(id, responseHandler);
      this.#sendMessage(request, void 0, void 0, (e) => {
        if (e) {
          reject(e);
        }
      });
    });
  }
}
const ipcClient = BackstageIpcClient.create();

exports.BackstageIpcClient = BackstageIpcClient;
exports.ipcClient = ipcClient;
