// @backstage/ui/dist/components/PasswordField/PasswordField.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useEffect, useState } from 'react';
import { TextField, Input, Button } from 'react-aria-components';
import clsx from 'clsx';
import { FieldLabel } from '../FieldLabel/FieldLabel.esm.js';
import { FieldError } from '../FieldError/FieldError.esm.js';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { PasswordFieldDefinition } from './definition.esm.js';
import { RiEyeLine, RiEyeOffLine } from '@remixicon/react';
import stylesPasswordField from './PasswordField.module.css.esm.js';

const PasswordField = forwardRef(
  (props, ref) => {
    const {
      label,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledBy
    } = props;
    useEffect(() => {
      if (!label && !ariaLabel && !ariaLabelledBy) {
        console.warn(
          "PasswordField requires either a visible label, aria-label, or aria-labelledby for accessibility"
        );
      }
    }, [label, ariaLabel, ariaLabelledBy]);
    const {
      classNames: classNamesPasswordField,
      dataAttributes,
      cleanedProps
    } = useStyles(PasswordFieldDefinition, {
      size: "small",
      ...props
    });
    const {
      className,
      description,
      icon,
      isRequired,
      secondaryLabel,
      placeholder,
      ...rest
    } = cleanedProps;
    const secondaryLabelText = secondaryLabel || (isRequired ? "Required" : null);
    const [isVisible, setIsVisible] = useState(false);
    return /* @__PURE__ */ jsxs(
      TextField,
      {
        className: clsx(
          classNamesPasswordField.root,
          stylesPasswordField[classNamesPasswordField.root],
          className
        ),
        ...dataAttributes,
        "aria-label": ariaLabel,
        "aria-labelledby": ariaLabelledBy,
        type: "password",
        ...rest,
        ref,
        children: [
          /* @__PURE__ */ jsx(
            FieldLabel,
            {
              label,
              secondaryLabel: secondaryLabelText,
              description
            }
          ),
          /* @__PURE__ */ jsxs(
            "div",
            {
              className: clsx(
                classNamesPasswordField.inputWrapper,
                stylesPasswordField[classNamesPasswordField.inputWrapper]
              ),
              "data-size": dataAttributes["data-size"],
              children: [
                icon && /* @__PURE__ */ jsx(
                  "div",
                  {
                    className: clsx(
                      classNamesPasswordField.inputIcon,
                      stylesPasswordField[classNamesPasswordField.inputIcon]
                    ),
                    "data-size": dataAttributes["data-size"],
                    "aria-hidden": "true",
                    children: icon
                  }
                ),
                /* @__PURE__ */ jsx(
                  Input,
                  {
                    className: clsx(
                      classNamesPasswordField.input,
                      stylesPasswordField[classNamesPasswordField.input]
                    ),
                    ...icon && { "data-icon": true },
                    placeholder,
                    type: isVisible ? "text" : "password"
                  }
                ),
                /* @__PURE__ */ jsx(
                  Button,
                  {
                    "data-size": dataAttributes["data-size"],
                    "data-variant": "tertiary",
                    "aria-label": isVisible ? "Hide value" : "Show value",
                    "aria-controls": isVisible ? "text" : "password",
                    "aria-expanded": isVisible,
                    onPress: () => setIsVisible((v) => !v),
                    className: clsx(
                      classNamesPasswordField.inputVisibility,
                      stylesPasswordField[classNamesPasswordField.inputVisibility]
                    ),
                    children: isVisible ? /* @__PURE__ */ jsx(RiEyeLine, {}) : /* @__PURE__ */ jsx(RiEyeOffLine, {})
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ jsx(FieldError, {})
        ]
      }
    );
  }
);
PasswordField.displayName = "PasswordField";

export { PasswordField };
