// @backstage/ui/dist/components/Dialog/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Dialog.module.css.esm.js';

const DialogDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DialogOverlay",
    dialog: "bui-Dialog",
    content: "bui-DialogContent"
  },
  propDefs: {
    children: {},
    className: {},
    width: {},
    height: {},
    style: {}
  }
});
const DialogHeaderDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DialogHeader",
    title: "bui-DialogHeaderTitle"
  },
  propDefs: {
    children: {},
    className: {}
  }
});
const DialogBodyDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DialogBody"
  },
  propDefs: {
    children: {},
    className: {}
  }
});
const DialogFooterDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-DialogFooter"
  },
  propDefs: {
    children: {},
    className: {}
  }
});

export { DialogBodyDefinition, DialogDefinition, DialogFooterDefinition, DialogHeaderDefinition };
