// @backstage/ui/dist/components/Grid/Grid.esm.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { GridDefinition, GridItemDefinition } from './definition.esm.js';

const GridRoot = forwardRef((props, ref) => {
  const { ownProps, dataAttributes, utilityStyle, restProps } = useDefinition(
    GridDefinition,
    { columns: "auto", gap: "4", ...props }
  );
  const { classes, childrenWithBgProvider } = ownProps;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: classes.root,
      style: { ...utilityStyle, ...ownProps.style },
      ...dataAttributes,
      ...restProps,
      children: childrenWithBgProvider
    }
  );
});
const GridItem = forwardRef((props, ref) => {
  const { ownProps, dataAttributes, utilityStyle, restProps } = useDefinition(
    GridItemDefinition,
    props
  );
  const { classes, childrenWithBgProvider } = ownProps;
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: classes.root,
      style: { ...utilityStyle, ...ownProps.style },
      ...dataAttributes,
      ...restProps,
      children: childrenWithBgProvider
    }
  );
});
const Grid = {
  Root: GridRoot,
  Item: GridItem
};

export { Grid };
