// @backstage/plugin-catalog-react/dist/components/EntityRelationCard/EntityRelationCard.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { Alert, Text, Link } from '@backstage/ui';
import { useTranslationRef } from '@backstage/frontend-plugin-api';
import { EntityInfoCard } from '../EntityInfoCard/EntityInfoCard.esm.js';
import { EntityDataTable } from '../EntityDataTable/EntityDataTable.esm.js';
import '@backstage/catalog-model';
import '../EntityRefLink/EntityRefLink.esm.js';
import 'react';
import 'lodash/get';
import { catalogReactTranslationRef } from '../../translation.esm.js';
import '../EntityDataTable/presets.esm.js';
import { useEntity } from '../../hooks/useEntity.esm.js';
import '../../hooks/useEntityListProvider.esm.js';
import 'react-use/esm/useAsync';
import 'lodash/isEqual';
import 'lodash/sortBy';
import '@backstage/core-plugin-api';
import '../../api.esm.js';
import { useRelatedEntities } from '../../hooks/useRelatedEntities.esm.js';
import 'react-use/esm/useObservable';
import '../../apis/EntityPresentationApi/EntityPresentationApi.esm.js';
import '../../apis/StarredEntitiesApi/StarredEntitiesApi.esm.js';
import 'zen-observable';

function EntityRelationCard(props) {
  const {
    title,
    relationType,
    entityKind,
    columnConfig,
    emptyState,
    className
  } = props;
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const { entity } = useEntity();
  const { entities, loading, error } = useRelatedEntities(entity, {
    type: relationType,
    kind: entityKind
  });
  return /* @__PURE__ */ jsx(EntityInfoCard, { title, className, children: error ? /* @__PURE__ */ jsx(Alert, { status: "warning", icon: true, title: error.message, role: "status" }) : /* @__PURE__ */ jsx(
    EntityDataTable,
    {
      columnConfig,
      data: entities ?? [],
      loading,
      emptyState: emptyState && /* @__PURE__ */ jsxs(Text, { as: "p", children: [
        emptyState.message,
        " ",
        emptyState.helpLink && /* @__PURE__ */ jsx(
          Link,
          {
            href: emptyState.helpLink,
            target: "_blank",
            rel: "noopener noreferrer",
            variant: "body-medium",
            children: t("entityRelationCard.emptyHelpLinkTitle")
          }
        )
      ] })
    }
  ) });
}

export { EntityRelationCard };
