// @backstage/plugin-auth-node/dist/oauth/OAuthCookieManager.cjs.js
'use strict';

var types = require('@backstage/types');

const THOUSAND_DAYS_MS = 1e3 * 24 * 60 * 60 * 1e3;
const TEN_MINUTES_MS = 600 * 1e3;
const MAX_COOKIE_SIZE_CHARACTERS = 4e3;
const defaultCookieConfigurer = ({
  callbackUrl,
  providerId,
  appOrigin
}) => {
  const { hostname: domain, pathname, protocol } = new URL(callbackUrl);
  const secure = protocol === "https:";
  let sameSite = "lax";
  if (new URL(appOrigin).hostname !== domain && secure) {
    sameSite = "none";
  }
  const path = pathname.endsWith(`${providerId}/handler/frame`) ? pathname.slice(0, -"/handler/frame".length) : `${pathname}/${providerId}`;
  return { path, secure, sameSite };
};
class OAuthCookieManager {
  constructor(options) {
    this.options = options;
    this.cookieConfigurer = options.cookieConfigurer ?? defaultCookieConfigurer;
    this.nonceCookie = `${options.providerId}-nonce`;
    this.refreshTokenCookie = `${options.providerId}-refresh-token`;
    this.grantedScopeCookie = `${options.providerId}-granted-scope`;
    this.maxAge = options.sessionDuration ? types.durationToMilliseconds(options.sessionDuration) : THOUSAND_DAYS_MS;
  }
  cookieConfigurer;
  nonceCookie;
  refreshTokenCookie;
  grantedScopeCookie;
  maxAge;
  getConfig(origin, pathSuffix = "") {
    const cookieConfig = this.cookieConfigurer({
      providerId: this.options.providerId,
      baseUrl: this.options.baseUrl,
      callbackUrl: this.options.callbackUrl,
      appOrigin: origin ?? this.options.defaultAppOrigin
    });
    return {
      httpOnly: true,
      sameSite: "lax",
      ...cookieConfig,
      path: cookieConfig.path + pathSuffix
    };
  }
  setNonce(res, nonce, origin) {
    this.setCookie(
      res,
      this.nonceCookie,
      nonce,
      TEN_MINUTES_MS,
      origin,
      "/handler"
    );
  }
  setRefreshToken(res, refreshToken, origin) {
    this.setCookie(
      res,
      this.refreshTokenCookie,
      refreshToken,
      this.maxAge,
      origin
    );
  }
  removeRefreshToken(res, origin) {
    this.removeCookie(res, this.refreshTokenCookie, origin);
  }
  removeGrantedScopes(res, origin) {
    this.removeCookie(res, this.grantedScopeCookie, origin);
  }
  setGrantedScopes(res, scope, origin) {
    this.setCookie(res, this.grantedScopeCookie, scope, this.maxAge, origin);
  }
  getNonce(req) {
    return this.getCookie(req, this.nonceCookie);
  }
  getRefreshToken(req) {
    return this.getCookie(req, this.refreshTokenCookie);
  }
  getGrantedScopes(req) {
    return this.getCookie(req, this.grantedScopeCookie);
  }
  setCookie(res, name, val, maxAge, origin, pathSuffix = "") {
    const options = {
      maxAge,
      ...this.getConfig(origin, pathSuffix)
    };
    const req = res.req;
    const newCookieShouldBeChunked = val.length > MAX_COOKIE_SIZE_CHARACTERS;
    const existingChunkCount = OAuthCookieManager.countExistingCookieChunks(
      req,
      name
    );
    const chunkedFormatExists = existingChunkCount > 0;
    if (this.cookieConfigurer === defaultCookieConfigurer) {
      this.removeLegacyCookieWithDomain(res, name, existingChunkCount);
    }
    if (chunkedFormatExists) {
      this.removeChunkedCookie(res, name, existingChunkCount);
    }
    if (newCookieShouldBeChunked) {
      this.setChunkedCookie(req, res, name, val, options);
    } else {
      res.cookie(name, val, options);
    }
  }
  removeLegacyCookieWithDomain(res, name, chunkCount) {
    const { hostname: domain } = new URL(this.options.callbackUrl);
    res.cookie(name, "", {
      ...this.getRemoveCookieOptions(),
      domain
    });
    this.removeChunkedCookie(res, name, chunkCount, { domain });
  }
  setChunkedCookie(req, res, name, val, options) {
    const nonChunkedFormatExists = !!req.cookies[name];
    if (nonChunkedFormatExists) {
      res.cookie(name, "", this.getRemoveCookieOptions());
    }
    const chunkedCookieArray = this.splitCookieToChunks(
      val,
      MAX_COOKIE_SIZE_CHARACTERS
    );
    chunkedCookieArray.forEach((chunkValue, chunkNumber) => {
      res.cookie(
        OAuthCookieManager.getCookieChunkName(name, chunkNumber),
        chunkValue,
        options
      );
    });
  }
  getCookie(req, name) {
    const existingChunkCount = OAuthCookieManager.countExistingCookieChunks(
      req,
      name
    );
    const isChunked = existingChunkCount > 0;
    if (isChunked) {
      return this.getChunkedCookie(req, name, existingChunkCount);
    }
    return req.cookies[name];
  }
  getChunkedCookie(req, name, chunkCount) {
    const chunkedCookieArray = [];
    for (let chunkNumber = 0; chunkNumber < chunkCount; chunkNumber++) {
      const chunk = req.cookies[OAuthCookieManager.getCookieChunkName(name, chunkNumber)];
      chunkedCookieArray.push(chunk);
    }
    return chunkedCookieArray.join("");
  }
  removeCookie(res, name, origin) {
    const req = res.req;
    const existingChunkCount = OAuthCookieManager.countExistingCookieChunks(
      req,
      name
    );
    const chunkedFormatExists = existingChunkCount > 0;
    const nonChunkedFormatExists = !!req.cookies[name];
    if (nonChunkedFormatExists) {
      res.cookie(name, "", this.getRemoveCookieOptions(origin));
    }
    if (chunkedFormatExists) {
      this.removeChunkedCookie(res, name, existingChunkCount, {
        origin
      });
    }
  }
  removeChunkedCookie(res, name, chunkCount, { domain, origin } = {}) {
    for (let chunkNumber = 0; chunkNumber < chunkCount; chunkNumber++) {
      const key = OAuthCookieManager.getCookieChunkName(name, chunkNumber);
      const baseOptions = this.getRemoveCookieOptions(origin);
      const options = domain ? { ...baseOptions, domain } : baseOptions;
      res.cookie(key, "", options);
    }
  }
  splitCookieToChunks(val, chunkSize) {
    const numChunks = Math.ceil(val.length / chunkSize);
    const chunkedCookieArray = Array(numChunks);
    let offset = 0;
    for (let i = 0; i < numChunks; i++) {
      chunkedCookieArray[i] = val.substring(offset, offset + chunkSize);
      offset += chunkSize;
    }
    return chunkedCookieArray;
  }
  static countExistingCookieChunks(req, name) {
    for (let chunkNumber = 0; ; chunkNumber++) {
      const key = OAuthCookieManager.getCookieChunkName(name, chunkNumber);
      const exists = !!req.cookies[key];
      if (!exists) {
        return chunkNumber;
      }
    }
  }
  static getCookieChunkName(name, chunkIndex) {
    return `${name}-${chunkIndex}`;
  }
  getRemoveCookieOptions(origin) {
    return {
      maxAge: 0,
      ...this.getConfig(origin)
    };
  }
}

exports.OAuthCookieManager = OAuthCookieManager;
