// @backstage/frontend-plugin-api/dist/components/ExtensionBoundary.esm.js
import { jsx, Fragment } from 'react/jsx-runtime';
import { lazy, useMemo, Suspense, useEffect } from 'react';
import { AnalyticsContext } from '../analytics/AnalyticsContext.esm.js';
import { useAnalytics } from '../analytics/useAnalytics.esm.js';
import { ErrorDisplayBoundary } from './ErrorDisplayBoundary.esm.js';
import { ErrorApiBoundary } from './ErrorApiBoundary.esm.js';
import { routableExtensionRenderedEvent } from '../core-plugin-api/src/analytics/Tracker.esm.js';
import '../apis/definitions/AppTreeApi.esm.js';
import '../apis/definitions/auth.esm.js';
import '../apis/definitions/AlertApi.esm.js';
import '../apis/definitions/AppLanguageApi.esm.js';
import '../apis/definitions/AppThemeApi.esm.js';
import '../apis/definitions/SwappableComponentsApi.esm.js';
import '../apis/definitions/ConfigApi.esm.js';
import '../apis/definitions/DiscoveryApi.esm.js';
import { errorApiRef } from '../apis/definitions/ErrorApi.esm.js';
import '../apis/definitions/FeatureFlagsApi.esm.js';
import '../apis/definitions/FetchApi.esm.js';
import '../apis/definitions/IconsApi.esm.js';
import '../apis/definitions/IdentityApi.esm.js';
import '../apis/definitions/DialogApi.esm.js';
import '../apis/definitions/OAuthRequestApi.esm.js';
import '../apis/definitions/RouteResolutionApi.esm.js';
import '../apis/definitions/StorageApi.esm.js';
import '../apis/definitions/AnalyticsApi.esm.js';
import '../apis/definitions/TranslationApi.esm.js';
import { useApi } from '../apis/system/useApi.esm.js';
import { pluginWrapperApiRef } from '../apis/definitions/PluginWrapperApi.esm.js';
import { coreExtensionData } from '../wiring/coreExtensionData.esm.js';
import 'zod';
import 'zod-to-json-schema';
import '../wiring/createExtensionBlueprint.esm.js';
import { AppNodeProvider } from './AppNodeProvider.esm.js';
import { Progress } from './DefaultSwappableComponents.esm.js';

function useOptionalErrorApi() {
  try {
    return useApi(errorApiRef);
  } catch {
    return void 0;
  }
}
function useOptionalPluginWrapperApi() {
  try {
    return useApi(pluginWrapperApiRef);
  } catch {
    return void 0;
  }
}
const RouteTracker = (props) => {
  const { enabled, children } = props;
  const analytics = useAnalytics();
  useEffect(() => {
    if (enabled) {
      analytics.captureEvent(routableExtensionRenderedEvent, "");
    }
  }, [analytics, enabled]);
  return /* @__PURE__ */ jsx(Fragment, { children });
};
function ExtensionBoundary(props) {
  const { node, children } = props;
  const errorApi = useOptionalErrorApi();
  const hasRoutePathOutput = Boolean(
    node.instance?.getData(coreExtensionData.routePath)
  );
  const plugin = node.spec.plugin;
  const pluginId = plugin.id ?? "app";
  const pluginWrapperApi = useOptionalPluginWrapperApi();
  const PluginWrapper = useMemo(() => {
    return pluginWrapperApi?.getPluginWrapper(pluginId);
  }, [pluginWrapperApi, pluginId]);
  const attributes = {
    extensionId: node.spec.id,
    pluginId
  };
  let content = /* @__PURE__ */ jsx(AnalyticsContext, { attributes, children: /* @__PURE__ */ jsx(RouteTracker, { enabled: hasRoutePathOutput, children }) });
  if (PluginWrapper) {
    content = /* @__PURE__ */ jsx(PluginWrapper, { children: content });
  }
  if (props.errorPresentation === "error-api") {
    content = /* @__PURE__ */ jsx(ErrorApiBoundary, { node, errorApi, children: content });
  } else {
    content = /* @__PURE__ */ jsx(ErrorDisplayBoundary, { plugin, children: content });
  }
  return /* @__PURE__ */ jsx(AppNodeProvider, { node, children: /* @__PURE__ */ jsx(Suspense, { fallback: /* @__PURE__ */ jsx(Progress, {}), children: content }) });
}
((ExtensionBoundary2) => {
  function lazy$1(appNode, loader) {
    const ExtensionComponent = lazy(
      () => loader().then((element) => ({ default: () => element }))
    );
    return /* @__PURE__ */ jsx(ExtensionBoundary2, { node: appNode, children: /* @__PURE__ */ jsx(ExtensionComponent, {}) });
  }
  ExtensionBoundary2.lazy = lazy$1;
  function lazyComponent(appNode, loader) {
    const ExtensionComponent = lazy(
      () => loader().then((Component) => ({ default: Component }))
    );
    return (props) => /* @__PURE__ */ jsx(ExtensionBoundary2, { node: appNode, children: /* @__PURE__ */ jsx(ExtensionComponent, { ...props }) });
  }
  ExtensionBoundary2.lazyComponent = lazyComponent;
})(ExtensionBoundary || (ExtensionBoundary = {}));

export { ExtensionBoundary };
