// @backstage/ui/dist/components/InternalLinkProvider/InternalLinkProvider.esm.js
import { jsx, Fragment } from 'react/jsx-runtime';
import { createContext, useContext, useEffect, useState, useCallback } from 'react';
import { RouterProvider } from 'react-aria-components';
import { useNavigate, useHref } from 'react-router-dom';
import { isExternalLink } from '../../utils/isExternalLink.esm.js';

function isInternalLink(href) {
  return !!href && !isExternalLink(href);
}
function RoutedContainer({ children }) {
  const navigate = useNavigate();
  return /* @__PURE__ */ jsx(RouterProvider, { navigate, useHref, children });
}
function useRoutingRegistration() {
  const [count, setCount] = useState(0);
  const register = useCallback(() => {
    setCount((c) => c + 1);
    return () => setCount((c) => c - 1);
  }, []);
  return { hasRoutedChildren: count > 0, contextValue: { register } };
}
function createRoutingRegistration() {
  const RoutingContext = createContext(null);
  function RoutingProvider({ children }) {
    const { hasRoutedChildren, contextValue } = useRoutingRegistration();
    const content = /* @__PURE__ */ jsx(RoutingContext.Provider, { value: contextValue, children });
    if (hasRoutedChildren) {
      return /* @__PURE__ */ jsx(RoutedContainer, { children: content });
    }
    return content;
  }
  function useRoutingRegistrationEffect(href) {
    const routingCtx = useContext(RoutingContext);
    const hasInternalHref = isInternalLink(href);
    useEffect(() => {
      if (hasInternalHref && routingCtx) {
        return routingCtx.register();
      }
      return void 0;
    }, [hasInternalHref, routingCtx]);
  }
  return { RoutingContext, RoutingProvider, useRoutingRegistrationEffect };
}
function InternalLinkProvider({
  href,
  children
}) {
  if (!isInternalLink(href)) {
    return /* @__PURE__ */ jsx(Fragment, { children });
  }
  return /* @__PURE__ */ jsx(RoutedContainer, { children });
}

export { InternalLinkProvider, RoutedContainer, createRoutingRegistration, isInternalLink, useRoutingRegistration };
