// @backstage/ui/dist/components/PasswordField/PasswordField.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useEffect, useState } from 'react';
import { TextField, Input, Button } from 'react-aria-components';
import { FieldLabel } from '../FieldLabel/FieldLabel.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../FieldLabel/FieldLabel.module.css.esm.js';
import { FieldError } from '../FieldError/FieldError.esm.js';
import '../FieldError/FieldError.module.css.esm.js';
import { PasswordFieldDefinition } from './definition.esm.js';
import { RiEyeLine, RiEyeOffLine } from '@remixicon/react';

const PasswordField = forwardRef(
  (props, ref) => {
    const { ownProps, restProps, dataAttributes } = useDefinition(
      PasswordFieldDefinition,
      props
    );
    const { classes, label, icon, secondaryLabel, placeholder, description } = ownProps;
    useEffect(() => {
      if (!label && !restProps["aria-label"] && !restProps["aria-labelledby"]) {
        console.warn(
          "PasswordField requires either a visible label, aria-label, or aria-labelledby for accessibility"
        );
      }
    }, [label, restProps["aria-label"], restProps["aria-labelledby"]]);
    const secondaryLabelText = secondaryLabel || (restProps.isRequired ? "Required" : null);
    const [isVisible, setIsVisible] = useState(false);
    return /* @__PURE__ */ jsxs(
      TextField,
      {
        className: classes.root,
        ...dataAttributes,
        type: "password",
        ...restProps,
        ref,
        children: [
          /* @__PURE__ */ jsx(
            FieldLabel,
            {
              label,
              secondaryLabel: secondaryLabelText,
              description
            }
          ),
          /* @__PURE__ */ jsxs(
            "div",
            {
              className: classes.inputWrapper,
              "data-size": dataAttributes["data-size"],
              children: [
                icon && /* @__PURE__ */ jsx(
                  "div",
                  {
                    className: classes.inputIcon,
                    "data-size": dataAttributes["data-size"],
                    "aria-hidden": "true",
                    children: icon
                  }
                ),
                /* @__PURE__ */ jsx(
                  Input,
                  {
                    className: classes.input,
                    ...icon && { "data-icon": true },
                    placeholder,
                    type: isVisible ? "text" : "password"
                  }
                ),
                /* @__PURE__ */ jsx(
                  Button,
                  {
                    "data-size": dataAttributes["data-size"],
                    "data-variant": "tertiary",
                    "aria-label": isVisible ? "Hide value" : "Show value",
                    "aria-controls": isVisible ? "text" : "password",
                    "aria-expanded": isVisible,
                    onPress: () => setIsVisible((v) => !v),
                    className: classes.inputVisibility,
                    children: isVisible ? /* @__PURE__ */ jsx(RiEyeLine, {}) : /* @__PURE__ */ jsx(RiEyeOffLine, {})
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ jsx(FieldError, {})
        ]
      }
    );
  }
);
PasswordField.displayName = "PasswordField";

export { PasswordField };
