// @backstage/catalog-client/dist/testUtils/InMemoryCatalogClient.esm.js
import { CATALOG_FILTER_EXISTS } from '@backstage/catalog-client';
import { stringifyEntityRef, parseEntityRef, DEFAULT_NAMESPACE } from '@backstage/catalog-model';
import { InputError, NotFoundError, NotImplementedError } from '@backstage/errors';
import { filterPredicateToFilterFunction } from '@backstage/filter-predicates';
import lodash from 'lodash';
import { traverse } from '../plugins/catalog-backend/src/database/operations/stitcher/buildEntitySearch.esm.js';
import { DEFAULT_STREAM_ENTITIES_LIMIT } from '../constants.esm.js';

function makeCursor(data) {
  return btoa(JSON.stringify(data));
}
function parseCursor(cursor) {
  return JSON.parse(atob(cursor));
}
const FILTER_EXISTS_SENTINEL = "\0CATALOG_FILTER_EXISTS";
function serializeFilterValue(v) {
  if (v === CATALOG_FILTER_EXISTS) return FILTER_EXISTS_SENTINEL;
  if (Array.isArray(v)) {
    return v.map(
      (x) => x === CATALOG_FILTER_EXISTS ? FILTER_EXISTS_SENTINEL : x
    );
  }
  return v;
}
function deserializeFilterValue(v) {
  if (v === FILTER_EXISTS_SENTINEL) return CATALOG_FILTER_EXISTS;
  if (Array.isArray(v)) {
    return v.map(
      (x) => x === FILTER_EXISTS_SENTINEL ? CATALOG_FILTER_EXISTS : x
    );
  }
  return v;
}
function serializeFilter(filter) {
  if (!filter) return void 0;
  return [filter].flat().map(
    (f) => Object.fromEntries(
      Object.entries(f).map(([k, v]) => [k, serializeFilterValue(v)])
    )
  );
}
function deserializeFilter(filter) {
  if (!filter) return void 0;
  return filter.map(
    (f) => Object.fromEntries(
      Object.entries(f).map(([k, v]) => [
        k,
        deserializeFilterValue(v)
      ])
    )
  );
}
function buildEntitySearch(entity) {
  const rows = traverse(entity);
  if (entity.metadata?.name) {
    rows.push({
      key: "metadata.name",
      value: entity.metadata.name.toLocaleLowerCase("en-US")
    });
  }
  if (entity.metadata?.namespace) {
    rows.push({
      key: "metadata.namespace",
      value: entity.metadata.namespace.toLocaleLowerCase("en-US")
    });
  }
  if (entity.metadata?.uid) {
    rows.push({
      key: "metadata.uid",
      value: entity.metadata.uid.toLocaleLowerCase("en-US")
    });
  }
  if (!entity.metadata.namespace) {
    rows.push({ key: "metadata.namespace", value: DEFAULT_NAMESPACE });
  }
  for (const relation of entity.relations ?? []) {
    rows.push({
      key: `relations.${relation.type.toLocaleLowerCase("en-US")}`,
      value: relation.targetRef.toLocaleLowerCase("en-US")
    });
  }
  return rows;
}
function createFilter(filterOrFilters) {
  if (!filterOrFilters) {
    return () => true;
  }
  const filters = [filterOrFilters].flat();
  return (entity) => {
    const rows = buildEntitySearch(entity);
    return filters.some((filter) => {
      for (const [key, expectedValue] of Object.entries(filter)) {
        const searchValues = rows.filter((row) => row.key === key.toLocaleLowerCase("en-US")).map((row) => row.value?.toString().toLocaleLowerCase("en-US"));
        if (searchValues.length === 0) {
          return false;
        }
        if (expectedValue === CATALOG_FILTER_EXISTS) {
          continue;
        }
        if (Array.isArray(expectedValue)) {
          return expectedValue.some(
            (value) => searchValues?.includes(String(value).toLocaleLowerCase("en-US"))
          );
        }
        if (!searchValues?.includes(
          String(expectedValue).toLocaleLowerCase("en-US")
        )) {
          return false;
        }
      }
      return true;
    });
  };
}
function getPathArrayAndValue(input, field) {
  return field.split(".").reduce(
    ([pathArray, inputSubset], pathPart, index, fieldParts) => {
      if (lodash.hasIn(inputSubset, pathPart)) {
        return [pathArray.concat(pathPart), inputSubset[pathPart]];
      } else if (fieldParts[index + 1] !== void 0) {
        fieldParts[index + 1] = `${pathPart}.${fieldParts[index + 1]}`;
        return [pathArray, inputSubset];
      }
      return [pathArray, void 0];
    },
    [[], input]
  );
}
function applyFieldsFilter(entity, fields) {
  if (!fields?.length) {
    return entity;
  }
  const output = {};
  for (const field of fields) {
    const [pathArray, value] = getPathArrayAndValue(entity, field);
    if (value !== void 0) {
      lodash.set(output, pathArray, value);
    }
  }
  return output;
}
function applyOrdering(entities, order) {
  if (!order) {
    return entities;
  }
  const orders = [order].flat();
  if (orders.length === 0) {
    return entities;
  }
  const searchMap = /* @__PURE__ */ new Map();
  for (const entity of entities) {
    searchMap.set(entity, buildEntitySearch(entity));
  }
  return [...entities].sort((a, b) => {
    const aRows = searchMap.get(a);
    const bRows = searchMap.get(b);
    for (const { field, order: dir } of orders) {
      const key = field.toLocaleLowerCase("en-US");
      const aRow = aRows.find((r) => r.key.toLocaleLowerCase("en-US") === key);
      const bRow = bRows.find((r) => r.key.toLocaleLowerCase("en-US") === key);
      const aValue = aRow?.value !== null && aRow?.value !== void 0 ? String(aRow.value).toLocaleLowerCase("en-US") : null;
      const bValue = bRow?.value !== null && bRow?.value !== void 0 ? String(bRow.value).toLocaleLowerCase("en-US") : null;
      if (aValue === null && bValue === null) continue;
      if (aValue === null) return 1;
      if (bValue === null) return -1;
      if (aValue < bValue) return dir === "asc" ? -1 : 1;
      if (aValue > bValue) return dir === "asc" ? 1 : -1;
    }
    const aRef = stringifyEntityRef(a);
    const bRef = stringifyEntityRef(b);
    if (aRef < bRef) return -1;
    if (aRef > bRef) return 1;
    return 0;
  });
}
function applyFullTextFilter(entities, fullTextFilter) {
  if (!fullTextFilter?.term?.trim()) {
    return entities;
  }
  const term = fullTextFilter.term.trim().toLocaleLowerCase("en-US");
  const fields = fullTextFilter.fields?.map((f) => f.toLocaleLowerCase("en-US"));
  return entities.filter((entity) => {
    const rows = buildEntitySearch(entity);
    return rows.some((row) => {
      if (fields?.length && !fields.includes(row.key.toLocaleLowerCase("en-US"))) {
        return false;
      }
      if (row.value === null || row.value === void 0) return false;
      const value = String(row.value).toLocaleLowerCase("en-US");
      return value.includes(term);
    });
  });
}
class InMemoryCatalogClient {
  #entities;
  constructor(options) {
    this.#entities = options?.entities?.slice() ?? [];
  }
  async getEntities(request) {
    const filter = createFilter(request?.filter);
    let items = this.#entities.filter(filter);
    items = applyOrdering(items, request?.order);
    let offset = request?.offset ?? 0;
    let limit = request?.limit;
    if (request?.after) {
      try {
        const cursor = parseCursor(request.after);
        if (cursor.offset !== void 0) offset = cursor.offset;
        if (cursor.limit !== void 0) limit = cursor.limit;
      } catch {
        throw new InputError("Invalid cursor");
      }
    }
    if (offset > 0) {
      items = items.slice(offset);
    }
    if (limit !== void 0) {
      items = items.slice(0, limit);
    }
    return {
      items: request?.fields ? items.map((e) => applyFieldsFilter(e, request.fields)) : items
    };
  }
  async getEntitiesByRefs(request) {
    const filter = createFilter(request.filter);
    const queryFilter = request.query ? filterPredicateToFilterFunction(request.query) : void 0;
    const refMap = this.#createEntityRefMap();
    const items = request.entityRefs.map((ref) => refMap.get(ref)).map(
      (e) => e && filter(e) && (!queryFilter || queryFilter(e)) ? e : void 0
    );
    return {
      items: request.fields ? items.map((e) => e ? applyFieldsFilter(e, request.fields) : void 0) : items
    };
  }
  async queryEntities(request) {
    let filter;
    let query;
    let orderFields;
    let fullTextFilter;
    let offset;
    let limit;
    if (request && "cursor" in request) {
      let c;
      try {
        c = parseCursor(request.cursor);
      } catch {
        throw new InputError("Invalid cursor");
      }
      filter = deserializeFilter(c.filter);
      query = c.query;
      orderFields = c.orderFields;
      fullTextFilter = c.fullTextFilter;
      offset = c.offset;
      limit = request.limit;
    } else {
      filter = request?.filter;
      query = request?.query;
      orderFields = request?.orderFields;
      fullTextFilter = request?.fullTextFilter;
      offset = request?.offset ?? 0;
      limit = request?.limit;
    }
    let items = this.#entities.filter(createFilter(filter));
    if (query) {
      items = items.filter(filterPredicateToFilterFunction(query));
    }
    if (fullTextFilter) {
      const orderFieldsList = orderFields ? [orderFields].flat() : [];
      items = applyFullTextFilter(items, {
        ...fullTextFilter,
        fields: fullTextFilter.fields ?? [
          orderFieldsList[0]?.field ?? "metadata.uid"
        ]
      });
    }
    items = applyOrdering(items, orderFields);
    const totalItems = items.length;
    if (limit === void 0 && offset === 0) {
      return {
        items: request?.fields ? items.map((e) => applyFieldsFilter(e, request?.fields)) : items,
        pageInfo: {},
        totalItems
      };
    }
    const effectiveLimit = limit ?? totalItems;
    const pageItems = items.slice(offset, offset + effectiveLimit);
    const cursorBase = {
      filter: serializeFilter(filter),
      query,
      orderFields,
      fullTextFilter,
      totalItems
    };
    const pageInfo = {};
    if (offset + effectiveLimit < totalItems) {
      pageInfo.nextCursor = makeCursor({
        ...cursorBase,
        offset: offset + effectiveLimit
      });
    }
    if (offset > 0) {
      pageInfo.prevCursor = makeCursor({
        ...cursorBase,
        offset: Math.max(0, offset - effectiveLimit)
      });
    }
    return {
      items: request?.fields ? pageItems.map((e) => applyFieldsFilter(e, request?.fields)) : pageItems,
      totalItems,
      pageInfo
    };
  }
  async getEntityAncestors(request) {
    const entity = this.#createEntityRefMap().get(request.entityRef);
    if (!entity) {
      throw new NotFoundError(`Entity with ref ${request.entityRef} not found`);
    }
    return {
      items: [{ entity, parentEntityRefs: [] }],
      rootEntityRef: request.entityRef
    };
  }
  async getEntityByRef(entityRef) {
    return this.#createEntityRefMap().get(
      stringifyEntityRef(parseEntityRef(entityRef))
    );
  }
  async removeEntityByUid(uid) {
    const index = this.#entities.findIndex((e) => e.metadata.uid === uid);
    if (index !== -1) {
      this.#entities.splice(index, 1);
    }
  }
  async refreshEntity(_entityRef) {
  }
  async getEntityFacets(request) {
    const filter = createFilter(request.filter);
    let filteredEntities = this.#entities.filter(filter);
    if (request.query) {
      filteredEntities = filteredEntities.filter(
        filterPredicateToFilterFunction(request.query)
      );
    }
    const facets = Object.fromEntries(
      request.facets.map((facet) => {
        const facetValues = /* @__PURE__ */ new Map();
        for (const entity of filteredEntities) {
          const rows = buildEntitySearch(entity);
          const uniqueValues = new Set(
            rows.filter(
              (row) => row.key.toLocaleLowerCase("en-US") === facet.toLocaleLowerCase("en-US")
            ).map((row) => row.value).filter((v) => v !== null && v !== void 0).map((v) => String(v))
          );
          for (const value of uniqueValues) {
            facetValues.set(value, (facetValues.get(value) ?? 0) + 1);
          }
        }
        const counts = Array.from(facetValues.entries()).map(
          ([value, count]) => ({ value, count })
        );
        return [facet, counts];
      })
    );
    return {
      facets
    };
  }
  async getLocations(_request) {
    throw new NotImplementedError("Method not implemented.");
  }
  async queryLocations(_request) {
    throw new NotImplementedError("Method not implemented.");
  }
  async *streamLocations(_request) {
    throw new NotImplementedError("Method not implemented.");
  }
  async getLocationById(_id) {
    throw new NotImplementedError("Method not implemented.");
  }
  async getLocationByRef(_locationRef) {
    throw new NotImplementedError("Method not implemented.");
  }
  async addLocation(_location) {
    throw new NotImplementedError("Method not implemented.");
  }
  async removeLocationById(_id) {
    throw new NotImplementedError("Method not implemented.");
  }
  async getLocationByEntity(_entityRef) {
    throw new NotImplementedError("Method not implemented.");
  }
  async validateEntity(_entity, _locationRef) {
    throw new NotImplementedError("Method not implemented.");
  }
  async analyzeLocation(_location) {
    throw new NotImplementedError("Method not implemented.");
  }
  async *streamEntities(request) {
    let cursor = void 0;
    const limit = request?.pageSize ?? DEFAULT_STREAM_ENTITIES_LIMIT;
    do {
      const res = await this.queryEntities(
        cursor ? { ...request, limit, cursor } : { ...request, limit }
      );
      yield res.items;
      cursor = res.pageInfo.nextCursor;
    } while (cursor);
  }
  #createEntityRefMap() {
    return new Map(this.#entities.map((e) => [stringifyEntityRef(e), e]));
  }
}

export { InMemoryCatalogClient };
