// @backstage/plugin-catalog-backend-module-github/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha = require('@backstage/plugin-catalog-node/alpha');
var pluginEventsNode = require('@backstage/plugin-events-node');
var GithubEntityProvider = require('./cjs/GithubEntityProvider-CJpx_erw.cjs.js');
require('@backstage/catalog-client');
require('@backstage/integration');
require('@octokit/rest');
require('lodash');
require('git-url-parse');
require('@backstage/backend-common');
require('path');
require('@backstage/plugin-catalog-node');
require('@octokit/graphql');
require('uuid');
require('@backstage/backend-tasks');
require('@backstage/catalog-model');
require('minimatch');

const githubCatalogModule = backendPluginApi.createBackendModule({
  pluginId: "catalog",
  moduleId: "github",
  register(env) {
    env.registerInit({
      deps: {
        analyzers: alpha.catalogAnalysisExtensionPoint,
        auth: backendPluginApi.coreServices.auth,
        catalog: alpha.catalogProcessingExtensionPoint,
        config: backendPluginApi.coreServices.rootConfig,
        discovery: backendPluginApi.coreServices.discovery,
        events: pluginEventsNode.eventsServiceRef,
        logger: backendPluginApi.coreServices.logger,
        scheduler: backendPluginApi.coreServices.scheduler
      },
      async init({
        catalog,
        config,
        events,
        logger,
        scheduler,
        analyzers,
        discovery,
        auth
      }) {
        analyzers.addScmLocationAnalyzer(
          new GithubEntityProvider.GithubLocationAnalyzer({
            discovery,
            config,
            auth
          })
        );
        catalog.addEntityProvider(
          GithubEntityProvider.GithubEntityProvider.fromConfig(config, {
            events,
            logger,
            scheduler
          })
        );
      }
    });
  }
});

exports.default = githubCatalogModule;
