// @backstage/plugin-catalog-node/dist/scmEvents/DefaultCatalogScmEventsService.cjs.js
'use strict';

class DefaultCatalogScmEventsService {
  #subscribers;
  #metrics;
  constructor(metrics) {
    this.#subscribers = /* @__PURE__ */ new Set();
    const meter = metrics.getMeter("default");
    this.#metrics = {
      actions: meter.createCounter("catalog.events.scm.actions", {
        description: "Number of actions taken as a result of SCM event messages",
        unit: "short"
      })
    };
  }
  subscribe(subscriber) {
    this.#subscribers.add(subscriber);
    return {
      unsubscribe: () => {
        this.#subscribers.delete(subscriber);
      }
    };
  }
  async publish(events) {
    await Promise.all(
      Array.from(this.#subscribers).map(async (subscriber) => {
        try {
          await subscriber.onEvents(events);
        } catch (error) {
        }
      })
    );
  }
  markEventActionTaken(options) {
    this.#metrics.actions.add(options.count ?? 1, { action: options.action });
  }
}

exports.DefaultCatalogScmEventsService = DefaultCatalogScmEventsService;
