// @backstage/plugin-app/dist/extensions/AppRoot.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { useState } from 'react';
import { createExtension, coreExtensionData, createExtensionInput, discoveryApiRef, errorApiRef, fetchApiRef, AppRootWrapperBlueprint, RouterBlueprint, SignInPageBlueprint, routeResolutionApiRef } from '@backstage/frontend-plugin-api';
import { identityApiRef, useApi, configApiRef } from '@backstage/core-plugin-api';
import { isProtectedApp } from '../packages/core-app-api/src/app/isProtectedApp.esm.js';
import { BrowserRouter } from 'react-router-dom';
import { RouteTracker } from '../packages/frontend-app-api/src/routing/RouteTracker.esm.js';
import { getBasePath } from '../packages/frontend-app-api/src/routing/getBasePath.esm.js';

const AppRoot = createExtension({
  name: "root",
  attachTo: { id: "app", input: "root" },
  inputs: {
    router: createExtensionInput([RouterBlueprint.dataRefs.component], {
      singleton: true,
      optional: true
    }),
    signInPage: createExtensionInput([SignInPageBlueprint.dataRefs.component], {
      singleton: true,
      optional: true
    }),
    children: createExtensionInput([coreExtensionData.reactElement], {
      singleton: true
    }),
    elements: createExtensionInput([coreExtensionData.reactElement]),
    wrappers: createExtensionInput([
      AppRootWrapperBlueprint.dataRefs.component
    ])
  },
  output: [coreExtensionData.reactElement],
  factory({ inputs, apis }) {
    if (inputs.router && inputs.router.node.spec.plugin?.id !== "app") {
      console.warn(
        `DEPRECATION WARNING: Router should only be installed as an extension in the app plugin. You can either use appPlugin.override(), or a module for the app plugin. The following extension will be ignored in the future: ${inputs.router.node.spec.id}`
      );
    }
    if (inputs.signInPage && inputs.signInPage.node.spec.plugin?.id !== "app") {
      console.warn(
        `DEPRECATION WARNING: SignInPage should only be installed as an extension in the app plugin. You can either use appPlugin.override(), or a module for the app plugin. The following extension will be ignored in the future: ${inputs.signInPage.node.spec.id}`
      );
    }
    if (isProtectedApp()) {
      const identityApi = apis.get(identityApiRef);
      if (!identityApi) {
        throw new Error("App requires an Identity API implementation");
      }
      const appIdentityProxy = toAppIdentityProxy(identityApi);
      const discoveryApi = apis.get(discoveryApiRef);
      const errorApi = apis.get(errorApiRef);
      const fetchApi = apis.get(fetchApiRef);
      if (!discoveryApi || !errorApi || !fetchApi) {
        throw new Error(
          "App is running in protected mode but missing required APIs"
        );
      }
      appIdentityProxy.enableCookieAuth({
        discoveryApi,
        errorApi,
        fetchApi
      });
    }
    let content = inputs.children.get(
      coreExtensionData.reactElement
    );
    for (const wrapper of inputs.wrappers) {
      const Component = wrapper.get(AppRootWrapperBlueprint.dataRefs.component);
      const pluginId = wrapper.node.spec.plugin.id;
      if (Component) {
        content = /* @__PURE__ */ jsx(Component, { children: content });
        if (pluginId !== "app") {
          console.warn(
            `DEPRECATION WARNING: AppRootWrappers should only be installed as an extension in the app plugin. You can either use appPlugin.override(), or a module for the app plugin. The following extension will be ignored in the future: ${wrapper.node.spec.id}`
          );
        }
      }
    }
    return [
      coreExtensionData.reactElement(
        /* @__PURE__ */ jsx(
          AppRouter,
          {
            SignInPageComponent: inputs.signInPage?.get(
              SignInPageBlueprint.dataRefs.component
            ),
            RouterComponent: inputs.router?.get(
              RouterBlueprint.dataRefs.component
            ),
            extraElements: inputs.elements?.map(
              (el) => el.get(coreExtensionData.reactElement)
            ),
            children: content
          }
        )
      )
    ];
  }
});
function SignInPageWrapper({
  component: Component,
  appIdentityProxy,
  children
}) {
  const [identityApi, setIdentityApi] = useState();
  const configApi = useApi(configApiRef);
  const basePath = getBasePath(configApi);
  if (!identityApi) {
    return /* @__PURE__ */ jsx(Component, { onSignInSuccess: setIdentityApi });
  }
  appIdentityProxy.setTarget(identityApi, {
    signOutTargetUrl: basePath || "/"
  });
  return /* @__PURE__ */ jsx(Fragment, { children });
}
function toAppIdentityProxy(identityApi) {
  if (!("enableCookieAuth" in identityApi)) {
    throw new Error("Unexpected Identity API implementation");
  }
  return identityApi;
}
function DefaultRouter(props) {
  const configApi = useApi(configApiRef);
  const basePath = getBasePath(configApi);
  return /* @__PURE__ */ jsx(BrowserRouter, { basename: basePath, children: props.children });
}
function AppRouter(props) {
  const {
    children,
    SignInPageComponent,
    RouterComponent = DefaultRouter,
    extraElements = []
  } = props;
  const configApi = useApi(configApiRef);
  const appIdentityProxy = toAppIdentityProxy(useApi(identityApiRef));
  const routeResolutionsApi = useApi(routeResolutionApiRef);
  const basePath = getBasePath(configApi);
  if (!("getRouteObjects" in routeResolutionsApi)) {
    throw new Error("Unexpected route resolution API implementation");
  }
  const routeObjects = routeResolutionsApi.getRouteObjects();
  if (!SignInPageComponent) {
    appIdentityProxy.setTarget(
      {
        getUserId: () => "guest",
        getIdToken: async () => void 0,
        getProfile: () => ({
          email: "guest@example.com",
          displayName: "Guest"
        }),
        getProfileInfo: async () => ({
          email: "guest@example.com",
          displayName: "Guest"
        }),
        getBackstageIdentity: async () => ({
          type: "user",
          userEntityRef: "user:default/guest",
          ownershipEntityRefs: ["user:default/guest"]
        }),
        getCredentials: async () => ({}),
        signOut: async () => {
        }
      },
      { signOutTargetUrl: basePath || "/" }
    );
    return /* @__PURE__ */ jsxs(RouterComponent, { children: [
      ...extraElements,
      /* @__PURE__ */ jsx(RouteTracker, { routeObjects }),
      children
    ] });
  }
  return /* @__PURE__ */ jsxs(RouterComponent, { children: [
    ...extraElements,
    /* @__PURE__ */ jsx(RouteTracker, { routeObjects }),
    /* @__PURE__ */ jsx(
      SignInPageWrapper,
      {
        component: SignInPageComponent,
        appIdentityProxy,
        children
      }
    )
  ] });
}

export { AppRoot, AppRouter };
