// @backstage/plugin-app/dist/extensions/AppThemeApi.esm.js
import { jsx } from 'react/jsx-runtime';
import { UnifiedThemeProvider, themes } from '@backstage/theme';
import DarkIcon from '@material-ui/icons/Brightness2';
import LightIcon from '@material-ui/icons/WbSunny';
import { ApiBlueprint, createExtensionInput, appThemeApiRef, ThemeBlueprint } from '@backstage/frontend-plugin-api';
import 'zen-observable';
import '@backstage/core-plugin-api';
import '../packages/core-app-api/src/apis/implementations/auth/saml/types.esm.js';
import { AppThemeSelector } from '../packages/core-app-api/src/apis/implementations/AppThemeApi/AppThemeSelector.esm.js';
import '@backstage/config';

const AppThemeApi = ApiBlueprint.makeWithOverrides({
  name: "app-theme",
  inputs: {
    themes: createExtensionInput([ThemeBlueprint.dataRefs.theme], {
      replaces: [{ id: "app", input: "themes" }]
    })
  },
  factory: (originalFactory, { inputs }) => {
    return originalFactory(
      (defineParams) => defineParams({
        api: appThemeApiRef,
        deps: {},
        factory: () => {
          const nonAppExtensions = inputs.themes.filter(
            (i) => i.node.spec.plugin?.id !== "app"
          );
          if (nonAppExtensions.length > 0) {
            const list = nonAppExtensions.map((i) => i.node.spec.id).join(", ");
            console.warn(
              `DEPRECATION WARNING: Theme should only be installed as an extension in the app plugin. You can either use appPlugin.override(), or a module for the app plugin. The following extension will be ignored in the future: ${list}`
            );
          }
          return AppThemeSelector.createWithStorage(
            inputs.themes.map((i) => i.get(ThemeBlueprint.dataRefs.theme))
          );
        }
      })
    );
  }
});
const LightTheme = ThemeBlueprint.make({
  name: "light",
  params: {
    theme: {
      id: "light",
      title: "Light Theme",
      variant: "light",
      icon: /* @__PURE__ */ jsx(LightIcon, {}),
      Provider: ({ children }) => /* @__PURE__ */ jsx(UnifiedThemeProvider, { theme: themes.light, children })
    }
  }
});
const DarkTheme = ThemeBlueprint.make({
  name: "dark",
  params: {
    theme: {
      id: "dark",
      title: "Dark Theme",
      variant: "dark",
      icon: /* @__PURE__ */ jsx(DarkIcon, {}),
      Provider: ({ children }) => /* @__PURE__ */ jsx(UnifiedThemeProvider, { theme: themes.dark, children })
    }
  }
});

export { AppThemeApi, DarkTheme, LightTheme };
