// @backstage/plugin-app/dist/extensions/PluginWrapperApi.esm.js
import { pluginWrapperApiRef, PluginWrapperBlueprint } from '@backstage/frontend-plugin-api/alpha';
import { ApiBlueprint, createExtensionInput } from '@backstage/frontend-plugin-api';
import { DefaultPluginWrapperApi } from '../apis/PluginWrapperApi/DefaultPluginWrapperApi.esm.js';

const PluginWrapperApi = ApiBlueprint.makeWithOverrides({
  name: "plugin-wrapper",
  inputs: {
    wrappers: createExtensionInput([PluginWrapperBlueprint.dataRefs.wrapper])
  },
  factory: (originalFactory, { inputs }) => {
    return originalFactory(
      (defineParams) => defineParams({
        api: pluginWrapperApiRef,
        deps: {},
        factory: () => {
          return DefaultPluginWrapperApi.fromWrappers(
            inputs.wrappers.map((wrapperInput) => ({
              loader: wrapperInput.get(PluginWrapperBlueprint.dataRefs.wrapper),
              pluginId: wrapperInput.node.spec.plugin.id ?? "app"
            }))
          );
        }
      })
    );
  }
});

export { PluginWrapperApi };
