// @backstage/plugin-catalog-node/dist/alpha.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var catalogClient = require('@backstage/catalog-client');
var alpha = require('@backstage/plugin-catalog-common/alpha');
var pluginPermissionNode = require('@backstage/plugin-permission-node');
var pluginCatalogNode = require('@backstage/plugin-catalog-node');
var extensions = require('./extensions.cjs.js');
var catalogScmEventsServiceRef = require('./scmEvents/catalogScmEventsServiceRef.cjs.js');

const catalogEntityPermissionResourceRef = pluginPermissionNode.createPermissionResourceRef().with({
  pluginId: "catalog",
  resourceType: alpha.RESOURCE_TYPE_CATALOG_ENTITY
});
const catalogServiceRef = backendPluginApi.createServiceRef({
  id: "catalog-client-legacy",
  defaultFactory: async (service) => backendPluginApi.createServiceFactory({
    service,
    deps: {
      discoveryApi: backendPluginApi.coreServices.discovery
    },
    async factory({ discoveryApi }) {
      return new catalogClient.CatalogClient({ discoveryApi });
    }
  })
});
const catalogLocationsExtensionPoint = pluginCatalogNode.catalogLocationsExtensionPoint;
const catalogProcessingExtensionPoint = pluginCatalogNode.catalogProcessingExtensionPoint;
const catalogAnalysisExtensionPoint = pluginCatalogNode.catalogAnalysisExtensionPoint;

exports.catalogModelExtensionPoint = extensions.catalogModelExtensionPoint;
exports.catalogPermissionExtensionPoint = extensions.catalogPermissionExtensionPoint;
exports.catalogScmEventsServiceRef = catalogScmEventsServiceRef.catalogScmEventsServiceRef;
exports.catalogAnalysisExtensionPoint = catalogAnalysisExtensionPoint;
exports.catalogEntityPermissionResourceRef = catalogEntityPermissionResourceRef;
exports.catalogLocationsExtensionPoint = catalogLocationsExtensionPoint;
exports.catalogProcessingExtensionPoint = catalogProcessingExtensionPoint;
exports.catalogServiceRef = catalogServiceRef;
