// @backstage/plugin-scaffolder-node/dist/scaffolderService.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var errors = require('@backstage/errors');
var integration = require('@backstage/integration');
var pluginScaffolderCommon = require('@backstage/plugin-scaffolder-common');

class DefaultScaffolderService {
  #auth;
  #client;
  #discovery;
  constructor(options) {
    this.#auth = options.auth;
    this.#client = options.client;
    this.#discovery = options.discovery;
  }
  async getTemplateParameterSchema(request, options) {
    return this.#client.getTemplateParameterSchema(
      request.templateRef,
      await this.#getOptions(options)
    );
  }
  async scaffold(request, options) {
    return this.#client.scaffold(request, await this.#getOptions(options));
  }
  async getTask(request, options) {
    return this.#client.getTask(
      request.taskId,
      await this.#getOptions(options)
    );
  }
  async cancelTask(request, options) {
    return this.#client.cancelTask(
      request.taskId,
      await this.#getOptions(options)
    );
  }
  async retry(request, options) {
    return this.#client.retry(request.taskId, await this.#getOptions(options));
  }
  async listTasks(request, options) {
    const { token } = await this.#getOptions(options);
    const baseUrl = await this.#discovery.getBaseUrl("scaffolder");
    const params = new URLSearchParams();
    if (request.createdBy) {
      params.set("createdBy", request.createdBy);
    }
    if (request.limit !== void 0) {
      params.set("limit", String(request.limit));
    }
    if (request.offset !== void 0) {
      params.set("offset", String(request.offset));
    }
    const query = params.toString();
    const url = `${baseUrl}/v2/tasks${query ? `?${query}` : ""}`;
    const response = await fetch(url, {
      headers: {
        "Content-Type": "application/json",
        ...token && { Authorization: `Bearer ${token}` }
      }
    });
    if (!response.ok) {
      throw await errors.ResponseError.fromResponse(response);
    }
    const body = await response.json();
    return {
      items: body.tasks,
      totalItems: body.totalTasks ?? 0
    };
  }
  async listActions(_request, options) {
    return this.#client.listActions(
      options ? await this.#getOptions(options) : {}
    );
  }
  async listTemplatingExtensions(_request, options) {
    return this.#client.listTemplatingExtensions(
      options ? await this.#getOptions(options) : {}
    );
  }
  async getLogs(request, options) {
    const { token } = await this.#getOptions(options);
    const baseUrl = await this.#discovery.getBaseUrl("scaffolder");
    const params = new URLSearchParams();
    if (request.after !== void 0) {
      params.set("after", String(request.after));
    }
    const query = params.toString();
    const taskId = encodeURIComponent(request.taskId);
    const url = `${baseUrl}/v2/tasks/${taskId}/events${query ? `?${query}` : ""}`;
    const response = await fetch(url, {
      headers: {
        "Content-Type": "application/json",
        ...token && { Authorization: `Bearer ${token}` }
      }
    });
    if (!response.ok) {
      throw await errors.ResponseError.fromResponse(response);
    }
    return response.json();
  }
  async dryRun(request, options) {
    return this.#client.dryRun(request, await this.#getOptions(options));
  }
  async autocomplete(request, options) {
    return this.#client.autocomplete(request, await this.#getOptions(options));
  }
  async #getOptions(options) {
    return this.#auth.getPluginRequestToken({
      onBehalfOf: options.credentials,
      targetPluginId: "scaffolder"
    });
  }
}
const scaffolderServiceRef = backendPluginApi.createServiceRef({
  id: "scaffolder-client",
  defaultFactory: async (service) => backendPluginApi.createServiceFactory({
    service,
    deps: {
      auth: backendPluginApi.coreServices.auth,
      discovery: backendPluginApi.coreServices.discovery,
      config: backendPluginApi.coreServices.rootConfig
    },
    async factory({ auth, discovery, config }) {
      const integrations = integration.ScmIntegrations.fromConfig(config);
      const client = new pluginScaffolderCommon.ScaffolderClient({
        discoveryApi: discovery,
        fetchApi: { fetch },
        scmIntegrationsApi: integrations
      });
      return new DefaultScaffolderService({
        auth,
        client,
        discovery
      });
    }
  })
});

exports.scaffolderServiceRef = scaffolderServiceRef;
