// @backstage/plugin-catalog-node/dist/testUtils/catalogServiceMock.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var testUtils = require('@backstage/catalog-client/testUtils');
var pluginCatalogNode = require('@backstage/plugin-catalog-node');
var backendTestUtils = require('@backstage/backend-test-utils');

function catalogServiceMock(options) {
  return new testUtils.InMemoryCatalogClient(options);
}
((catalogServiceMock2) => {
  catalogServiceMock2.factory = (options) => backendPluginApi.createServiceFactory({
    service: pluginCatalogNode.catalogServiceRef,
    deps: {},
    factory: () => new testUtils.InMemoryCatalogClient(options)
  });
  catalogServiceMock2.mock = backendTestUtils.createServiceMock(
    pluginCatalogNode.catalogServiceRef,
    () => ({
      getEntities: jest.fn(),
      getEntitiesByRefs: jest.fn(),
      queryEntities: jest.fn(),
      getEntityAncestors: jest.fn(),
      getEntityByRef: jest.fn(),
      removeEntityByUid: jest.fn(),
      refreshEntity: jest.fn(),
      getEntityFacets: jest.fn(),
      getLocations: jest.fn(),
      queryLocations: jest.fn(),
      streamLocations: jest.fn(),
      getLocationById: jest.fn(),
      getLocationByRef: jest.fn(),
      addLocation: jest.fn(),
      removeLocationById: jest.fn(),
      getLocationByEntity: jest.fn(),
      validateEntity: jest.fn(),
      analyzeLocation: jest.fn(),
      streamEntities: jest.fn()
    })
  );
})(catalogServiceMock || (catalogServiceMock = {}));

exports.catalogServiceMock = catalogServiceMock;
