// @backstage/ui/dist/hooks/useBreakpoint.esm.js
import { useMemo } from 'react';
import { useSyncExternalStore } from 'use-sync-external-store/shim';

const breakpoints = [
  { name: "Initial", id: "initial", value: 0 },
  { name: "Extra Small", id: "xs", value: 640 },
  { name: "Small", id: "sm", value: 768 },
  { name: "Medium", id: "md", value: 1024 },
  { name: "Large", id: "lg", value: 1280 },
  { name: "Extra Large", id: "xl", value: 1536 }
];
const breakpointIndex = new Map(
  breakpoints.map((bp, i) => [bp.id, i])
);
function bpIndex(key) {
  return breakpointIndex.get(key) ?? 0;
}
const IS_SERVER = typeof window === "undefined" || typeof window.matchMedia === "undefined";
function computeBreakpoint() {
  if (IS_SERVER) {
    return "initial";
  }
  for (let i = breakpoints.length - 1; i >= 0; i--) {
    if (window.matchMedia(`(min-width: ${breakpoints[i].value}px)`).matches) {
      return breakpoints[i].id;
    }
  }
  return "initial";
}
let current;
const listeners = /* @__PURE__ */ new Set();
let initialized = false;
function ensureInitialized() {
  if (initialized || IS_SERVER) {
    return;
  }
  initialized = true;
  current = computeBreakpoint();
  for (const bp of breakpoints) {
    const mql = window.matchMedia(`(min-width: ${bp.value}px)`);
    mql.addEventListener("change", () => {
      const next = computeBreakpoint();
      if (next !== current) {
        current = next;
        for (const cb of listeners) {
          cb();
        }
      }
    });
  }
}
function subscribe(callback) {
  ensureInitialized();
  listeners.add(callback);
  return () => {
    listeners.delete(callback);
  };
}
function getSnapshot() {
  ensureInitialized();
  return current ?? "initial";
}
function getServerSnapshot() {
  return "initial";
}
const useBreakpoint = () => {
  const breakpoint = useSyncExternalStore(
    subscribe,
    getSnapshot,
    getServerSnapshot
  );
  return useMemo(
    () => ({
      breakpoint,
      up: (key) => bpIndex(breakpoint) >= bpIndex(key),
      down: (key) => bpIndex(breakpoint) < bpIndex(key)
    }),
    [breakpoint]
  );
};

export { breakpoints, useBreakpoint };
