// @backstage/plugin-app-react/dist/blueprints/index.esm.js
import { AppRootWrapperBlueprint as AppRootWrapperBlueprint$1, ThemeBlueprint as ThemeBlueprint$1, SwappableComponentBlueprint as SwappableComponentBlueprint$1, SignInPageBlueprint as SignInPageBlueprint$1, RouterBlueprint as RouterBlueprint$1, NavContentBlueprint as NavContentBlueprint$1, IconBundleBlueprint as IconBundleBlueprint$1, TranslationBlueprint as TranslationBlueprint$1 } from '@backstage/frontend-plugin-api';

const AppRootWrapperBlueprint = AppRootWrapperBlueprint$1;
const ThemeBlueprint = ThemeBlueprint$1;
const SwappableComponentBlueprint = SwappableComponentBlueprint$1;
const SignInPageBlueprint = SignInPageBlueprint$1;
const RouterBlueprint = RouterBlueprint$1;
const NavContentBlueprint = NavContentBlueprint$1;
const IconBundleBlueprint = IconBundleBlueprint$1;
const TranslationBlueprint = TranslationBlueprint$1;

export { AppRootWrapperBlueprint, IconBundleBlueprint, NavContentBlueprint, RouterBlueprint, SignInPageBlueprint, SwappableComponentBlueprint, ThemeBlueprint, TranslationBlueprint };
