// @backstage/integration/dist/gitlab/GitLabIntegration.esm.js
import { basicIntegrations, defaultScmResolveUrl } from '../helpers.esm.js';
import { readGitLabIntegrationConfigs } from './config.esm.js';
import pThrottle from 'p-throttle';

class GitLabIntegration {
  constructor(integrationConfig) {
    this.integrationConfig = integrationConfig;
    this.fetchImpl = this.createFetchStrategy();
  }
  static factory = ({ config }) => {
    const configs = readGitLabIntegrationConfigs(
      config.getOptionalConfigArray("integrations.gitlab") ?? []
    );
    return basicIntegrations(
      configs.map((c) => new GitLabIntegration(c)),
      (i) => i.config.host
    );
  };
  fetchImpl;
  get type() {
    return "gitlab";
  }
  get title() {
    return this.integrationConfig.host;
  }
  get config() {
    return this.integrationConfig;
  }
  resolveUrl(options) {
    return defaultScmResolveUrl(options);
  }
  resolveEditUrl(url) {
    return replaceGitLabUrlType(url, "edit");
  }
  fetch(input, init) {
    return this.fetchImpl(input, init);
  }
  createFetchStrategy() {
    let fetchFn = async (url, options) => {
      return fetch(url, { ...options, mode: "same-origin" });
    };
    const retryConfig = this.integrationConfig.retry;
    if (retryConfig) {
      fetchFn = this.withRetry(fetchFn, retryConfig);
      if (retryConfig.maxApiRequestsPerMinute && retryConfig.maxApiRequestsPerMinute > 0) {
        fetchFn = pThrottle({
          limit: retryConfig.maxApiRequestsPerMinute,
          interval: 6e4
        })(fetchFn);
      }
    }
    return fetchFn;
  }
  withRetry(fetchFn, retryConfig) {
    const maxRetries = retryConfig?.maxRetries ?? 0;
    const retryStatusCodes = retryConfig?.retryStatusCodes ?? [];
    if (maxRetries <= 0 || retryStatusCodes.length === 0) {
      return fetchFn;
    }
    return async (url, options) => {
      const abortSignal = options?.signal;
      let response;
      let attempt = 0;
      for (; ; ) {
        response = await fetchFn(url, options);
        if (!retryStatusCodes.includes(response.status)) {
          break;
        }
        if (attempt++ >= maxRetries) {
          break;
        }
        const retryAfter = response.headers.get("Retry-After");
        const delay = retryAfter ? parseInt(retryAfter, 10) * 1e3 : Math.min(100 * Math.pow(2, attempt - 1), 1e4);
        await sleep(delay, abortSignal);
      }
      return response;
    };
  }
}
async function sleep(durationMs, abortSignal) {
  if (abortSignal?.aborted) {
    return;
  }
  await new Promise((resolve) => {
    let timeoutHandle = void 0;
    const done = () => {
      if (timeoutHandle) {
        clearTimeout(timeoutHandle);
      }
      abortSignal?.removeEventListener("abort", done);
      resolve();
    };
    timeoutHandle = setTimeout(done, durationMs);
    abortSignal?.addEventListener("abort", done);
  });
}
function replaceGitLabUrlType(url, type) {
  return url.replace(/\/\-\/(blob|tree|edit)\//, `/-/${type}/`);
}

export { GitLabIntegration, replaceGitLabUrlType, sleep };
