// @backstage/core-app-api/dist/apis/implementations/AppThemeApi/AppThemeSelector.esm.js
import { BehaviorSubject } from '../../../lib/subjects.esm.js';

const STORAGE_KEY = "theme";
class AppThemeSelector {
  constructor(themes) {
    this.themes = themes;
  }
  static createWithStorage(themes) {
    const selector = new AppThemeSelector(themes);
    if (!window.localStorage) {
      return selector;
    }
    const initialThemeId = window.localStorage.getItem(STORAGE_KEY) ?? void 0;
    selector.setActiveThemeId(initialThemeId);
    const subscription = selector.activeThemeId$().subscribe((themeId) => {
      if (themeId) {
        window.localStorage.setItem(STORAGE_KEY, themeId);
      } else {
        window.localStorage.removeItem(STORAGE_KEY);
      }
    });
    const storageListener = (event) => {
      if (event.key === STORAGE_KEY) {
        const themeId = localStorage.getItem(STORAGE_KEY) ?? void 0;
        selector.setActiveThemeId(themeId);
      }
    };
    window.addEventListener("storage", storageListener);
    selector.#storageSubscription = subscription;
    selector.#storageListener = storageListener;
    return selector;
  }
  activeThemeId;
  subject = new BehaviorSubject(void 0);
  // References for cleanup when using createWithStorage
  #storageSubscription;
  #storageListener;
  getInstalledThemes() {
    return this.themes.slice();
  }
  activeThemeId$() {
    return this.subject;
  }
  getActiveThemeId() {
    return this.activeThemeId;
  }
  setActiveThemeId(themeId) {
    this.activeThemeId = themeId;
    this.subject.next(themeId);
  }
  /**
   * Cleans up resources created by createWithStorage().
   * Call this method when the selector is no longer needed to prevent memory leaks.
   * This is particularly useful in testing scenarios or when the app is unmounted.
   */
  dispose() {
    if (this.#storageSubscription) {
      this.#storageSubscription.unsubscribe();
      this.#storageSubscription = void 0;
    }
    if (this.#storageListener) {
      window.removeEventListener("storage", this.#storageListener);
      this.#storageListener = void 0;
    }
  }
}

export { AppThemeSelector };
