// @backstage/ui/dist/components/Select/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Select.module.css.esm.js';

const SelectDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Select",
    popover: "bui-SelectPopover"
  },
  propDefs: {
    icon: {},
    size: { dataAttribute: true, default: "small" },
    options: {},
    searchable: {},
    searchPlaceholder: {},
    label: {},
    secondaryLabel: {},
    description: {},
    isRequired: {},
    className: {}
  }
});
const SelectTriggerDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-SelectTrigger",
      chevron: "bui-SelectTriggerChevron",
      value: "bui-SelectValue"
    },
    bg: "consumer",
    propDefs: {
      icon: {}
    }
  }
);
const SelectContentDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-SelectSearchWrapper",
      search: "bui-SelectSearch",
      searchClear: "bui-SelectSearchClear"
    },
    propDefs: {
      searchable: {},
      searchPlaceholder: { default: "Search..." },
      options: {}
    }
  }
);
const SelectListBoxDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-SelectList",
      item: "bui-SelectItem",
      itemIndicator: "bui-SelectItemIndicator",
      itemLabel: "bui-SelectItemLabel",
      noResults: "bui-SelectNoResults"
    },
    propDefs: {
      options: {}
    }
  }
);

export { SelectContentDefinition, SelectDefinition, SelectListBoxDefinition, SelectTriggerDefinition };
