// @backstage/ui/dist/components/TablePagination/TablePagination.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import clsx from 'clsx';
import { useId } from 'react-aria';
import { Text } from '../Text/Text.esm.js';
import { ButtonIcon } from '../ButtonIcon/ButtonIcon.esm.js';
import { useMemo } from 'react';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../ButtonIcon/ButtonIcon.module.css.esm.js';
import { Select } from '../Select/Select.esm.js';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { TablePaginationDefinition } from './definition.esm.js';
import styles from './TablePagination.module.css.esm.js';
import { RiArrowLeftSLine, RiArrowRightSLine } from '@remixicon/react';

const DEFAULT_PAGE_SIZE_OPTIONS = [
  { label: "Show 5 results", value: 5 },
  { label: "Show 10 results", value: 10 },
  { label: "Show 20 results", value: 20 },
  { label: "Show 30 results", value: 30 },
  { label: "Show 40 results", value: 40 },
  { label: "Show 50 results", value: 50 }
];
function getOptionValue(option) {
  return typeof option === "number" ? option : option.value;
}
function isNumberArray(options) {
  return options.length > 0 && typeof options[0] === "number";
}
function normalizePageSizeOptions(options) {
  if (isNumberArray(options)) {
    return options.map((value) => ({
      label: `Show ${value} results`,
      value
    }));
  }
  return options;
}
function TablePagination({
  pageSize,
  pageSizeOptions = DEFAULT_PAGE_SIZE_OPTIONS,
  offset,
  totalCount,
  hasNextPage,
  hasPreviousPage,
  onNextPage,
  onPreviousPage,
  onPageSizeChange,
  showPageSizeOptions = true,
  getLabel
}) {
  const { classNames } = useStyles(TablePaginationDefinition, {});
  const labelId = useId();
  const normalizedOptions = useMemo(
    () => normalizePageSizeOptions(pageSizeOptions),
    [pageSizeOptions]
  );
  const effectivePageSize = useMemo(() => {
    const isValid = pageSizeOptions.some(
      (opt) => getOptionValue(opt) === pageSize
    );
    if (isValid) {
      return pageSize;
    }
    const firstValue = getOptionValue(pageSizeOptions[0]);
    console.warn(
      `TablePagination: pageSize ${pageSize} is not in pageSizeOptions, using ${firstValue} instead`
    );
    return firstValue;
  }, [pageSize, pageSizeOptions]);
  const hasItems = totalCount !== void 0 && totalCount !== 0;
  let label = `${totalCount} items`;
  if (getLabel) {
    label = getLabel({ pageSize: effectivePageSize, offset, totalCount });
  } else if (offset !== void 0) {
    const fromCount = offset + 1;
    const toCount = Math.min(offset + effectivePageSize, totalCount ?? 0);
    label = `${fromCount} - ${toCount} of ${totalCount}`;
  }
  return /* @__PURE__ */ jsxs("div", { className: clsx(classNames.root, styles[classNames.root]), children: [
    /* @__PURE__ */ jsx("div", { className: clsx(classNames.left, styles[classNames.left]), children: showPageSizeOptions && /* @__PURE__ */ jsx(
      Select,
      {
        name: "pageSize",
        size: "small",
        "aria-label": "Select table page size",
        options: normalizedOptions.map((opt) => ({
          label: opt.label,
          value: String(opt.value)
        })),
        defaultValue: effectivePageSize.toString(),
        onChange: (value) => {
          const newPageSize = Number(value);
          onPageSizeChange?.(newPageSize);
        },
        className: clsx(classNames.select, styles[classNames.select])
      }
    ) }),
    /* @__PURE__ */ jsxs("div", { className: clsx(classNames.right, styles[classNames.right]), children: [
      hasItems && /* @__PURE__ */ jsx(Text, { as: "p", variant: "body-medium", id: labelId, children: label }),
      /* @__PURE__ */ jsx(
        ButtonIcon,
        {
          variant: "secondary",
          size: "small",
          onClick: onPreviousPage,
          isDisabled: !hasPreviousPage,
          icon: /* @__PURE__ */ jsx(RiArrowLeftSLine, {}),
          "aria-label": "Previous table page",
          "aria-describedby": hasItems ? labelId : void 0
        }
      ),
      /* @__PURE__ */ jsx(
        ButtonIcon,
        {
          variant: "secondary",
          size: "small",
          onClick: onNextPage,
          isDisabled: !hasNextPage,
          icon: /* @__PURE__ */ jsx(RiArrowRightSLine, {}),
          "aria-label": "Next table page",
          "aria-describedby": hasItems ? labelId : void 0
        }
      )
    ] })
  ] });
}

export { TablePagination };
