// @backstage/ui/dist/components/SearchAutocomplete/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './SearchAutocomplete.module.css.esm.js';

const SearchAutocompleteDefinition = defineComponent()({
  styles,
  bg: "consumer",
  classNames: {
    root: "bui-SearchAutocomplete",
    searchField: "bui-SearchAutocompleteSearchField",
    searchFieldInput: "bui-SearchAutocompleteInput",
    searchFieldClear: "bui-SearchAutocompleteClear",
    popover: "bui-SearchAutocompletePopover",
    inner: "bui-SearchAutocompleteInner",
    listBox: "bui-SearchAutocompleteListBox",
    loadingState: "bui-SearchAutocompleteLoadingState",
    emptyState: "bui-SearchAutocompleteEmptyState"
  },
  propDefs: {
    "aria-label": {},
    "aria-labelledby": {},
    size: { dataAttribute: true, default: "small" },
    placeholder: { default: "Search" },
    inputValue: {},
    onInputChange: {},
    popoverWidth: {},
    popoverPlacement: {},
    children: {},
    isLoading: {},
    defaultOpen: {},
    className: {},
    style: {}
  }
});
const SearchAutocompleteItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-SearchAutocompleteItem",
    itemContent: "bui-SearchAutocompleteItemContent"
  },
  propDefs: {
    children: {},
    className: {}
  }
});

export { SearchAutocompleteDefinition, SearchAutocompleteItemDefinition };
