// @backstage/ui/dist/components/SearchAutocomplete/SearchAutocomplete.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useRef, Children } from 'react';
import { useInteractOutside } from '@react-aria/interactions';
import { OverlayTriggerStateContext, Autocomplete, SearchField, Input, Button, Popover, ListBox, ListBoxItem } from 'react-aria-components';
import { useOverlayTriggerState } from '@react-stately/overlays';
import { RiSearch2Line, RiCloseCircleLine } from '@remixicon/react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { SearchAutocompleteDefinition, SearchAutocompleteItemDefinition } from './definition.esm.js';
import { Box } from '../Box/Box.esm.js';
import '../Box/Box.module.css.esm.js';
import { BgReset } from '../../hooks/useBg.esm.js';
import { VisuallyHidden } from '../VisuallyHidden/VisuallyHidden.esm.js';
import '../VisuallyHidden/VisuallyHidden.module.css.esm.js';

const SearchAutocompleteEmptyState = () => {
  const { ownProps } = useDefinition(SearchAutocompleteDefinition, {});
  return /* @__PURE__ */ jsx("div", { className: ownProps.classes.emptyState, children: "No results found." });
};
function SearchAutocomplete(props) {
  const { ownProps, dataAttributes } = useDefinition(
    SearchAutocompleteDefinition,
    props
  );
  const {
    classes,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    inputValue,
    onInputChange,
    placeholder,
    popoverWidth,
    popoverPlacement = "bottom start",
    children,
    isLoading,
    defaultOpen,
    style
  } = ownProps;
  const triggerRef = useRef(null);
  const popoverRef = useRef(null);
  const hasValue = !!inputValue;
  const hasChildren = Children.count(children) > 0;
  let liveMessage = "";
  if (isLoading) {
    liveMessage = "Searching";
  } else if (hasValue && !hasChildren) {
    liveMessage = "No results found";
  }
  const overlayState = useOverlayTriggerState({ defaultOpen });
  useInteractOutside({
    ref: popoverRef,
    onInteractOutside: (e) => {
      const target = e.target;
      if (triggerRef.current?.contains(target)) {
        return;
      }
      overlayState.close();
    },
    isDisabled: !overlayState.isOpen
  });
  return /* @__PURE__ */ jsx(OverlayTriggerStateContext.Provider, { value: overlayState, children: /* @__PURE__ */ jsxs(
    Autocomplete,
    {
      inputValue,
      onInputChange: (value) => {
        onInputChange?.(value);
        if (value) {
          overlayState.open();
        } else {
          overlayState.close();
        }
      },
      children: [
        /* @__PURE__ */ jsx(
          SearchField,
          {
            className: classes.searchField,
            "aria-label": ariaLabel ?? (ariaLabelledBy ? void 0 : placeholder),
            "aria-labelledby": ariaLabelledBy,
            "data-size": dataAttributes["data-size"],
            onKeyDown: (e) => {
              if (e.key === "Enter" && !overlayState.isOpen && hasValue) {
                e.preventDefault();
                overlayState.open();
              }
            },
            children: /* @__PURE__ */ jsxs(
              "div",
              {
                ref: triggerRef,
                className: classes.root,
                ...dataAttributes,
                style,
                children: [
                  /* @__PURE__ */ jsx("div", { "aria-hidden": "true", children: /* @__PURE__ */ jsx(RiSearch2Line, {}) }),
                  /* @__PURE__ */ jsx(
                    Input,
                    {
                      className: classes.searchFieldInput,
                      placeholder
                    }
                  ),
                  /* @__PURE__ */ jsx(
                    Button,
                    {
                      className: classes.searchFieldClear,
                      style: { visibility: hasValue ? "visible" : "hidden" },
                      children: /* @__PURE__ */ jsx(RiCloseCircleLine, {})
                    }
                  )
                ]
              }
            )
          }
        ),
        /* @__PURE__ */ jsx(
          Popover,
          {
            ref: popoverRef,
            className: classes.popover,
            triggerRef,
            isNonModal: true,
            placement: popoverPlacement,
            ...popoverWidth ? { style: { width: popoverWidth } } : {},
            children: /* @__PURE__ */ jsx(BgReset, { children: /* @__PURE__ */ jsx(Box, { bg: "neutral", className: classes.inner, children: /* @__PURE__ */ jsx(
              ListBox,
              {
                className: classes.listBox,
                autoFocus: "first",
                shouldFocusOnHover: true,
                "aria-busy": isLoading || void 0,
                "data-stale": isLoading || void 0,
                renderEmptyState: () => isLoading ? /* @__PURE__ */ jsx("div", { className: classes.loadingState, children: "Searching..." }) : /* @__PURE__ */ jsx(SearchAutocompleteEmptyState, {}),
                onAction: () => {
                  overlayState.close();
                },
                children
              }
            ) }) })
          }
        ),
        /* @__PURE__ */ jsx(VisuallyHidden, { "aria-live": "polite", "aria-atomic": "true", children: liveMessage })
      ]
    }
  ) });
}
function SearchAutocompleteItem(props) {
  const { ownProps, restProps } = useDefinition(
    SearchAutocompleteItemDefinition,
    props
  );
  const { classes, children } = ownProps;
  return /* @__PURE__ */ jsx(
    ListBoxItem,
    {
      textValue: typeof children === "string" ? children : void 0,
      className: classes.root,
      ...restProps,
      children: /* @__PURE__ */ jsx("div", { className: classes.itemContent, children })
    }
  );
}

export { SearchAutocomplete, SearchAutocompleteItem };
