// @backstage/plugin-scaffolder/dist/alpha/components/EditorSubPage.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useCallback } from 'react';
import { Routes, Route, useNavigate } from 'react-router-dom';
import { Content } from '@backstage/core-components';
import { makeStyles } from '@material-ui/core/styles';
import { RequirePermission } from '@backstage/plugin-permission-react';
import { templateManagementPermission } from '@backstage/plugin-scaffolder-common/alpha';
import { SecretsContextProvider } from '@backstage/plugin-scaffolder-react';
import { TemplateEditorIntro } from './TemplateEditorPage/TemplateEditorIntro.esm.js';
import { TemplateEditor } from './TemplateEditorPage/TemplateEditor.esm.js';
import { TemplateFormPreviewer } from './TemplateEditorPage/TemplateFormPreviewer.esm.js';
import { CustomFieldExplorer } from './TemplateEditorPage/CustomFieldExplorer.esm.js';
import { useTemplateDirectory } from './TemplateEditorPage/useTemplateDirectory.esm.js';

const useEditorStyles = makeStyles({
  editorContent: {
    padding: 0,
    height: "calc(100dvh - var(--bui-header-height, 0px))"
  },
  formContent: {
    padding: 0,
    height: "calc(100dvh - var(--bui-header-height, 0px))"
  }
});
function EditorIntroContent() {
  const navigate = useNavigate();
  const { openDirectory, createDirectory } = useTemplateDirectory();
  const handleSelect = useCallback(
    (option) => {
      if (option === "local") {
        openDirectory().then(() => navigate("template")).catch(() => {
        });
      } else if (option === "create-template") {
        createDirectory().then(() => navigate("template")).catch(() => {
        });
      } else if (option === "form") {
        navigate("template-form");
      } else if (option === "field-explorer") {
        navigate("custom-fields");
      }
    },
    [openDirectory, createDirectory, navigate]
  );
  return /* @__PURE__ */ jsx(Content, { children: /* @__PURE__ */ jsx(TemplateEditorIntro, { onSelect: handleSelect }) });
}
function EditorContent() {
  const classes = useEditorStyles();
  return /* @__PURE__ */ jsx(Content, { className: classes.editorContent, children: /* @__PURE__ */ jsx(TemplateEditor, {}) });
}
function FormPreviewContent() {
  const classes = useEditorStyles();
  const navigate = useNavigate();
  const handleClose = useCallback(() => {
    navigate("..");
  }, [navigate]);
  return /* @__PURE__ */ jsx(Content, { className: classes.formContent, children: /* @__PURE__ */ jsx(TemplateFormPreviewer, { onClose: handleClose }) });
}
function CustomFieldsContent() {
  return /* @__PURE__ */ jsx(Content, { children: /* @__PURE__ */ jsx(CustomFieldExplorer, {}) });
}
function EditorSubPage() {
  return /* @__PURE__ */ jsx(RequirePermission, { permission: templateManagementPermission, children: /* @__PURE__ */ jsx(SecretsContextProvider, { children: /* @__PURE__ */ jsxs(Routes, { children: [
    /* @__PURE__ */ jsx(Route, { index: true, element: /* @__PURE__ */ jsx(EditorIntroContent, {}) }),
    /* @__PURE__ */ jsx(Route, { path: "template", element: /* @__PURE__ */ jsx(EditorContent, {}) }),
    /* @__PURE__ */ jsx(Route, { path: "template-form", element: /* @__PURE__ */ jsx(FormPreviewContent, {}) }),
    /* @__PURE__ */ jsx(Route, { path: "custom-fields", element: /* @__PURE__ */ jsx(CustomFieldsContent, {}) })
  ] }) }) });
}

export { EditorSubPage };
