// @backstage/plugin-scaffolder/dist/alpha/components/TemplatesSubPage.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useCallback } from 'react';
import { Routes, Route, useNavigate } from 'react-router-dom';
import { DocsIcon, Content, ContentHeader, SupportButton } from '@backstage/core-components';
import { useRouteRef, useApp } from '@backstage/core-plugin-api';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { EntityListProvider, CatalogFilterLayout, EntitySearchBar, EntityKindPicker, UserListPicker, EntityTagPicker, EntityOwnerPicker } from '@backstage/plugin-catalog-react';
import { TemplateCategoryPicker, TemplateGroups } from '@backstage/plugin-scaffolder-react/alpha';
import { useCustomFieldExtensions, useCustomLayouts, SecretsContextProvider } from '@backstage/plugin-scaffolder-react';
import { parseEntityRef, stringifyEntityRef } from '@backstage/catalog-model';
import { RegisterExistingButton } from './TemplateListPage/RegisterExistingButton.esm.js';
import { TemplateWizardPageContent } from './TemplateWizardPage/TemplateWizardPage.esm.js';
import { registerComponentRouteRef, viewTechDocRouteRef, selectedTemplateRouteRef } from '../../routes.esm.js';
import { scaffolderTranslationRef } from '../../translation.esm.js';
import { DEFAULT_SCAFFOLDER_FIELD_EXTENSIONS } from '../../extensions/default.esm.js';
import { buildTechDocsURL } from '@backstage/plugin-techdocs-react';
import { TECHDOCS_ANNOTATION, TECHDOCS_EXTERNAL_ANNOTATION } from '@backstage/plugin-techdocs-common';
import { OpaqueFormField } from '../../packages/scaffolder-internal/src/wiring/InternalFormField.esm.js';

function TemplateListContent() {
  const registerComponentLink = useRouteRef(registerComponentRouteRef);
  const viewTechDocsLink = useRouteRef(viewTechDocRouteRef);
  const templateRoute = useRouteRef(selectedTemplateRouteRef);
  const navigate = useNavigate();
  const app = useApp();
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const groups = [
    {
      title: t("templateListPage.templateGroups.defaultTitle"),
      filter: () => true
    }
  ];
  const additionalLinksForEntity = useCallback(
    (template) => {
      if (!(template.metadata.annotations?.[TECHDOCS_ANNOTATION] || template.metadata.annotations?.[TECHDOCS_EXTERNAL_ANNOTATION]) || !viewTechDocsLink) {
        return [];
      }
      const url = buildTechDocsURL(template, viewTechDocsLink);
      return url ? [
        {
          icon: app.getSystemIcon("docs") ?? DocsIcon,
          text: t(
            "templateListPage.additionalLinksForEntity.viewTechDocsTitle"
          ),
          url
        }
      ] : [];
    },
    [app, viewTechDocsLink, t]
  );
  const onTemplateSelected = useCallback(
    (template) => {
      const { namespace, name } = parseEntityRef(stringifyEntityRef(template));
      navigate(templateRoute({ namespace, templateName: name }));
    },
    [navigate, templateRoute]
  );
  return /* @__PURE__ */ jsx(EntityListProvider, { children: /* @__PURE__ */ jsxs(Content, { children: [
    /* @__PURE__ */ jsxs(ContentHeader, { children: [
      /* @__PURE__ */ jsx(
        RegisterExistingButton,
        {
          title: t(
            "templateListPage.contentHeader.registerExistingButtonTitle"
          ),
          to: registerComponentLink && registerComponentLink()
        }
      ),
      /* @__PURE__ */ jsx(SupportButton, { children: t("templateListPage.contentHeader.supportButtonTitle") })
    ] }),
    /* @__PURE__ */ jsxs(CatalogFilterLayout, { children: [
      /* @__PURE__ */ jsxs(CatalogFilterLayout.Filters, { children: [
        /* @__PURE__ */ jsx(EntitySearchBar, {}),
        /* @__PURE__ */ jsx(EntityKindPicker, { initialFilter: "template", hidden: true }),
        /* @__PURE__ */ jsx(
          UserListPicker,
          {
            initialFilter: "all",
            availableFilters: ["all", "starred"]
          }
        ),
        /* @__PURE__ */ jsx(TemplateCategoryPicker, {}),
        /* @__PURE__ */ jsx(EntityTagPicker, {}),
        /* @__PURE__ */ jsx(EntityOwnerPicker, {})
      ] }),
      /* @__PURE__ */ jsx(CatalogFilterLayout.Content, { children: /* @__PURE__ */ jsx(
        TemplateGroups,
        {
          groups,
          onTemplateSelected,
          additionalLinksForEntity
        }
      ) })
    ] })
  ] }) });
}
function TemplatesSubPage(props) {
  const customFieldExtensions = useCustomFieldExtensions(void 0);
  const customLayouts = useCustomLayouts(void 0);
  const fieldExtensions = [
    ...customFieldExtensions,
    ...props.formFields?.map(OpaqueFormField.toInternal) ?? [],
    ...DEFAULT_SCAFFOLDER_FIELD_EXTENSIONS.filter(
      ({ name }) => !customFieldExtensions.some(
        (customFieldExtension) => customFieldExtension.name === name
      )
    )
  ];
  return /* @__PURE__ */ jsxs(Routes, { children: [
    /* @__PURE__ */ jsx(Route, { index: true, element: /* @__PURE__ */ jsx(TemplateListContent, {}) }),
    /* @__PURE__ */ jsx(
      Route,
      {
        path: ":namespace/:templateName",
        element: /* @__PURE__ */ jsx(SecretsContextProvider, { children: /* @__PURE__ */ jsx(
          TemplateWizardPageContent,
          {
            customFieldExtensions: fieldExtensions,
            layouts: customLayouts
          }
        ) })
      }
    )
  ] });
}

export { TemplatesSubPage };
