// @backstage/ui/dist/components/Accordion/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Accordion.module.css.esm.js';

const AccordionDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Accordion"
  },
  bg: "provider",
  propDefs: {
    bg: { dataAttribute: true, default: "neutral" },
    children: {},
    className: {}
  }
});
const AccordionTriggerDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-AccordionTrigger",
    button: "bui-AccordionTriggerButton",
    title: "bui-AccordionTriggerTitle",
    subtitle: "bui-AccordionTriggerSubtitle",
    icon: "bui-AccordionTriggerIcon"
  },
  propDefs: {
    className: {},
    title: {},
    subtitle: {},
    children: {}
  }
});
const AccordionPanelDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-AccordionPanel"
  },
  propDefs: {
    className: {}
  }
});
const AccordionGroupDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-AccordionGroup"
  },
  propDefs: {
    className: {},
    allowsMultiple: { default: false }
  }
});

export { AccordionDefinition, AccordionGroupDefinition, AccordionPanelDefinition, AccordionTriggerDefinition };
