// @backstage/plugin-catalog/dist/components/EntityOrphanWarning/EntityOrphanWarning.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { useEntity } from '@backstage/plugin-catalog-react';
import { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { DeleteEntityDialog } from './DeleteEntityDialog.esm.js';
import { useRouteRef } from '@backstage/core-plugin-api';
import { rootRouteRef } from '../../routes.esm.js';
import { catalogTranslationRef } from '../../alpha/translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { Alert, Button } from '@backstage/ui';

function isOrphan(entity) {
  return entity?.metadata?.annotations?.["backstage.io/orphan"] === "true";
}
function EntityOrphanWarning() {
  const navigate = useNavigate();
  const catalogLink = useRouteRef(rootRouteRef);
  const [confirmationDialogOpen, setConfirmationDialogOpen] = useState(false);
  const { entity } = useEntity();
  const { t } = useTranslationRef(catalogTranslationRef);
  const cleanUpAfterRemoval = async () => {
    setConfirmationDialogOpen(false);
    navigate(catalogLink());
  };
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      Alert,
      {
        status: "warning",
        icon: true,
        title: t("deleteEntity.description"),
        customActions: /* @__PURE__ */ jsx(
          Button,
          {
            size: "small",
            variant: "tertiary",
            destructive: true,
            onPress: () => setConfirmationDialogOpen(true),
            children: t("deleteEntity.actionButtonTitle")
          }
        )
      }
    ),
    /* @__PURE__ */ jsx(
      DeleteEntityDialog,
      {
        open: confirmationDialogOpen,
        entity,
        onConfirm: cleanUpAfterRemoval,
        onClose: () => setConfirmationDialogOpen(false)
      }
    )
  ] });
}

export { EntityOrphanWarning, isOrphan };
