// @backstage/frontend-plugin-api/dist/apis/definitions/auth.esm.js
import 'react/jsx-runtime';
import '@backstage/version-bridge';
import '@backstage/errors';
import { createApiRef } from '../system/ApiRef.esm.js';

const SessionState = {
  /**
   * User signed in.
   */
  SignedIn: "SignedIn",
  /**
   * User not signed in.
   */
  SignedOut: "SignedOut"
};
const googleAuthApiRef = createApiRef().with({
  id: "core.auth.google",
  pluginId: "app"
});
const githubAuthApiRef = createApiRef().with({
  id: "core.auth.github",
  pluginId: "app"
});
const oktaAuthApiRef = createApiRef().with({
  id: "core.auth.okta",
  pluginId: "app"
});
const gitlabAuthApiRef = createApiRef().with({
  id: "core.auth.gitlab",
  pluginId: "app"
});
const microsoftAuthApiRef = createApiRef().with({
  id: "core.auth.microsoft",
  pluginId: "app"
});
const oneloginAuthApiRef = createApiRef().with({
  id: "core.auth.onelogin",
  pluginId: "app"
});
const bitbucketAuthApiRef = createApiRef().with({
  id: "core.auth.bitbucket",
  pluginId: "app"
});
const bitbucketServerAuthApiRef = createApiRef().with({
  id: "core.auth.bitbucket-server",
  pluginId: "app"
});
const atlassianAuthApiRef = createApiRef().with({
  id: "core.auth.atlassian",
  pluginId: "app"
});
const vmwareCloudAuthApiRef = createApiRef().with({
  id: "core.auth.vmware-cloud",
  pluginId: "app"
});
const openshiftAuthApiRef = createApiRef().with({
  id: "core.auth.openshift",
  pluginId: "app"
});

export { SessionState, atlassianAuthApiRef, bitbucketAuthApiRef, bitbucketServerAuthApiRef, githubAuthApiRef, gitlabAuthApiRef, googleAuthApiRef, microsoftAuthApiRef, oktaAuthApiRef, oneloginAuthApiRef, openshiftAuthApiRef, vmwareCloudAuthApiRef };
