// @backstage/ui/dist/components/Tabs/Tabs.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { createContext, useRef, useState, useMemo, useCallback, Children, isValidElement, cloneElement, useContext, useEffect } from 'react';
import { useLocation } from 'react-router-dom';
import { TabsIndicators } from './TabsIndicators.esm.js';
import { Tabs as Tabs$1, TabList as TabList$1, Tab as Tab$1, TabPanel as TabPanel$1 } from 'react-aria-components';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { TabsDefinition } from './definition.esm.js';
import { createRoutingRegistration, isInternalLink } from '../InternalLinkProvider/InternalLinkProvider.esm.js';
import styles from './Tabs.module.css.esm.js';
import clsx from 'clsx';

const { RoutingProvider, useRoutingRegistrationEffect } = createRoutingRegistration();
const TabsContext = createContext(void 0);
const useTabsContext = () => {
  const context = useContext(TabsContext);
  if (!context) {
    throw new Error("Tab components must be used within a Tabs component");
  }
  return context;
};
const TabSelectionContext = createContext(
  null
);
const isTabActive = (tabHref, currentPathname, matchStrategy) => {
  if (matchStrategy === "exact") {
    return tabHref === currentPathname;
  }
  if (tabHref === currentPathname) {
    return true;
  }
  return currentPathname.startsWith(`${tabHref}/`);
};
const Tabs = (props) => {
  const { classNames, cleanedProps } = useStyles(TabsDefinition, props);
  const { className, children, ...rest } = cleanedProps;
  const tabsRef = useRef(null);
  const tabRefs = useRef(/* @__PURE__ */ new Map());
  const [hoveredKey, setHoveredKey] = useState(null);
  const prevHoveredKey = useRef(null);
  const [routedTabs, setRoutedTabs] = useState(() => /* @__PURE__ */ new Set());
  const [activeTabs, setActiveTabs] = useState(
    () => /* @__PURE__ */ new Map()
  );
  const setTabRef = (key, element) => {
    if (element) {
      tabRefs.current.set(key, element);
    } else {
      tabRefs.current.delete(key);
    }
  };
  const selectedTabId = useMemo(() => {
    if (routedTabs.size === 0) {
      return void 0;
    }
    if (activeTabs.size === 0) {
      return "";
    }
    let selectedId = null;
    let maxSegments = -1;
    activeTabs.forEach((segmentCount, id) => {
      if (segmentCount > maxSegments) {
        maxSegments = segmentCount;
        selectedId = id;
      }
    });
    return selectedId;
  }, [routedTabs, activeTabs]);
  const registerRoutedTab = useCallback((id) => {
    setRoutedTabs((prev) => new Set(prev).add(id));
  }, []);
  const unregisterRoutedTab = useCallback((id) => {
    setRoutedTabs((prev) => {
      const next = new Set(prev);
      next.delete(id);
      return next;
    });
  }, []);
  const registerActiveTab = useCallback((id, segmentCount) => {
    setActiveTabs((prev) => new Map(prev).set(id, segmentCount));
  }, []);
  const unregisterActiveTab = useCallback((id) => {
    setActiveTabs((prev) => {
      const next = new Map(prev);
      next.delete(id);
      return next;
    });
  }, []);
  if (!children) return null;
  const tabsContextValue = {
    tabsRef,
    tabRefs,
    hoveredKey,
    prevHoveredKey,
    setHoveredKey,
    setTabRef
  };
  const selectionContextValue = useMemo(
    () => ({
      registerRoutedTab,
      unregisterRoutedTab,
      registerActiveTab,
      unregisterActiveTab
    }),
    [
      registerRoutedTab,
      unregisterRoutedTab,
      registerActiveTab,
      unregisterActiveTab
    ]
  );
  return /* @__PURE__ */ jsx(RoutingProvider, { children: /* @__PURE__ */ jsx(TabsContext.Provider, { value: tabsContextValue, children: /* @__PURE__ */ jsx(TabSelectionContext.Provider, { value: selectionContextValue, children: /* @__PURE__ */ jsx(
    Tabs$1,
    {
      className: clsx(
        classNames.tabs,
        styles[classNames.tabs],
        className
      ),
      keyboardActivation: "manual",
      selectedKey: selectedTabId,
      ref: tabsRef,
      ...rest,
      children
    }
  ) }) }) });
};
const TabList = (props) => {
  const { classNames, cleanedProps } = useStyles(TabsDefinition, props);
  const { className, children, ...rest } = cleanedProps;
  const { setHoveredKey, tabRefs, tabsRef, hoveredKey, prevHoveredKey } = useTabsContext();
  const handleHover = (key) => {
    setHoveredKey(key);
  };
  const enhancedChildren = Children.map(children, (child) => {
    if (isValidElement(child)) {
      return cloneElement(child, {
        onHoverStart: () => handleHover(child.props.id),
        onHoverEnd: () => handleHover(null)
      });
    }
    return child;
  });
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: clsx(
        classNames.tabListWrapper,
        styles[classNames.tabListWrapper],
        className
      ),
      children: [
        /* @__PURE__ */ jsx(
          TabList$1,
          {
            className: clsx(classNames.tabList, styles[classNames.tabList]),
            "aria-label": "Toolbar tabs",
            ...rest,
            children: enhancedChildren
          }
        ),
        /* @__PURE__ */ jsx(
          TabsIndicators,
          {
            tabRefs,
            tabsRef,
            hoveredKey,
            prevHoveredKey
          }
        )
      ]
    }
  );
};
function RoutedTabEffects({
  id,
  href,
  matchStrategy = "exact"
}) {
  const selectionCtx = useContext(TabSelectionContext);
  const location = useLocation();
  useRoutingRegistrationEffect(href);
  useEffect(() => {
    if (selectionCtx) {
      selectionCtx.registerRoutedTab(id);
      return () => selectionCtx.unregisterRoutedTab(id);
    }
    return void 0;
  }, [id, selectionCtx]);
  const isActive = isTabActive(href, location.pathname, matchStrategy);
  const segmentCount = href.split("/").filter(Boolean).length;
  useEffect(() => {
    if (isActive && selectionCtx) {
      selectionCtx.registerActiveTab(id, segmentCount);
      return () => selectionCtx.unregisterActiveTab(id);
    }
    return void 0;
  }, [isActive, id, segmentCount, selectionCtx]);
  return null;
}
const Tab = (props) => {
  const { classNames, cleanedProps } = useStyles(TabsDefinition, props);
  const { className, href, children, id, matchStrategy, ...rest } = cleanedProps;
  const { setTabRef } = useTabsContext();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    isInternalLink(href) && /* @__PURE__ */ jsx(
      RoutedTabEffects,
      {
        id,
        href,
        matchStrategy
      }
    ),
    /* @__PURE__ */ jsx(
      Tab$1,
      {
        id,
        className: clsx(classNames.tab, styles[classNames.tab], className),
        ref: (el) => setTabRef(id, el),
        href,
        ...rest,
        children
      }
    )
  ] });
};
const TabPanel = (props) => {
  const { classNames, cleanedProps } = useStyles(TabsDefinition, props);
  const { className, children, ...rest } = cleanedProps;
  return /* @__PURE__ */ jsx(
    TabPanel$1,
    {
      className: clsx(classNames.panel, styles[classNames.panel], className),
      ...rest,
      children
    }
  );
};

export { Tab, TabList, TabPanel, Tabs };
