// @backstage/ui/dist/components/Dialog/Dialog.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { Modal, Dialog as Dialog$1, Heading, DialogTrigger as DialogTrigger$1 } from 'react-aria-components';
import { RiCloseLine } from '@remixicon/react';
import { Button } from '../Button/Button.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../Button/Button.module.css.esm.js';
import { DialogDefinition, DialogHeaderDefinition, DialogBodyDefinition, DialogFooterDefinition } from './definition.esm.js';
import { Box } from '../Box/Box.esm.js';
import '../Box/Box.module.css.esm.js';
import { BgReset } from '../../hooks/useBg.esm.js';
import { Flex } from '../Flex/Flex.esm.js';
import '../Flex/Flex.module.css.esm.js';

const DialogTrigger = (props) => {
  return /* @__PURE__ */ jsx(DialogTrigger$1, { ...props });
};
const Dialog = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(DialogDefinition, props, {
      classNameTarget: "dialog"
    });
    const { classes, children, width, height, style } = ownProps;
    return /* @__PURE__ */ jsx(
      Modal,
      {
        ref,
        className: classes.root,
        isDismissable: true,
        isKeyboardDismissDisabled: false,
        ...restProps,
        children: /* @__PURE__ */ jsx(
          Dialog$1,
          {
            className: classes.dialog,
            style: {
              ["--bui-dialog-min-width"]: typeof width === "number" ? `${width}px` : width || "400px",
              ...height ? {
                ["--bui-dialog-height"]: typeof height === "number" ? `${height}px` : height
              } : {},
              ...style
            },
            children: /* @__PURE__ */ jsx(BgReset, { children: /* @__PURE__ */ jsx(Box, { bg: "neutral", className: classes.content, children }) })
          }
        )
      }
    );
  }
);
Dialog.displayName = "Dialog";
const DialogHeader = forwardRef((props, ref) => {
  const { ownProps, restProps } = useDefinition(DialogHeaderDefinition, props);
  const { classes, children } = ownProps;
  return /* @__PURE__ */ jsxs(Flex, { ref, className: classes.root, ...restProps, children: [
    /* @__PURE__ */ jsx(Heading, { slot: "title", className: classes.title, children }),
    /* @__PURE__ */ jsx(Button, { name: "close", "aria-label": "Close", variant: "tertiary", slot: "close", children: /* @__PURE__ */ jsx(RiCloseLine, {}) })
  ] });
});
DialogHeader.displayName = "DialogHeader";
const DialogBody = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(DialogBodyDefinition, props);
    const { classes, children } = ownProps;
    return /* @__PURE__ */ jsx("div", { className: classes.root, ref, ...restProps, children });
  }
);
DialogBody.displayName = "DialogBody";
const DialogFooter = forwardRef((props, ref) => {
  const { ownProps, restProps } = useDefinition(DialogFooterDefinition, props);
  const { classes, children } = ownProps;
  return /* @__PURE__ */ jsx("div", { ref, className: classes.root, ...restProps, children });
});
DialogFooter.displayName = "DialogFooter";

export { Dialog, DialogBody, DialogFooter, DialogHeader, DialogTrigger };
