// @backstage/plugin-scaffolder-backend/dist/scaffolder/actions/builtin/fetch/templateFile.cjs.js
'use strict';

var pluginScaffolderNode = require('@backstage/plugin-scaffolder-node');
var path = require('path');
var templateFile_examples = require('./templateFile.examples.cjs.js');
var templateFileActionHandler = require('./templateFileActionHandler.cjs.js');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var path__default = /*#__PURE__*/_interopDefaultCompat(path);

function createFetchTemplateFileAction(options) {
  return pluginScaffolderNode.createTemplateAction({
    id: "fetch:template:file",
    description: "Downloads single file and places it in the workspace.",
    examples: templateFile_examples.examples,
    schema: {
      input: {
        type: "object",
        required: ["url", "targetPath"],
        properties: {
          url: {
            title: "Fetch URL",
            description: "Relative path or absolute URL pointing to the single file to fetch.",
            type: "string"
          },
          targetPath: {
            title: "Target Path",
            description: "Target path within the working directory to download the file as.",
            type: "string"
          },
          values: {
            title: "Template Values",
            description: "Values to pass on to the templating engine",
            type: "object"
          },
          cookiecutterCompat: {
            title: "Cookiecutter compatibility mode",
            description: "Enable features to maximise compatibility with templates built for fetch:cookiecutter",
            type: "boolean"
          },
          replace: {
            title: "Replace file",
            description: "If set, replace file in targetPath instead of overwriting existing one.",
            type: "boolean"
          },
          token: {
            title: "Token",
            description: "An optional token to use for authentication when reading the resources.",
            type: "string"
          }
        }
      }
    },
    supportsDryRun: true,
    handler: (ctx) => templateFileActionHandler.createTemplateFileActionHandler({
      ctx,
      resolveTemplateFile: async () => {
        ctx.logger.info("Fetching template file content from remote URL");
        const workDir = await ctx.createTemporaryDirectory();
        const tmpFilePath = path__default.default.join(workDir, "tmp");
        await pluginScaffolderNode.fetchFile({
          baseUrl: ctx.templateInfo?.baseUrl,
          fetchUrl: ctx.input.url,
          outputPath: tmpFilePath,
          token: ctx.input.token,
          ...options
        });
        return tmpFilePath;
      },
      ...options
    })
  });
}

exports.createFetchTemplateFileAction = createFetchTemplateFileAction;
