// @backstage/ui/dist/components/Link/Link.esm.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef, useRef } from 'react';
import { useLink } from 'react-aria';
import clsx from 'clsx';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { LinkDefinition } from './definition.esm.js';
import { InternalLinkProvider } from '../InternalLinkProvider/InternalLinkProvider.esm.js';
import styles from './Link.module.css.esm.js';

const LinkInternal = forwardRef((props, ref) => {
  const { classNames, dataAttributes, cleanedProps } = useStyles(
    LinkDefinition,
    {
      variant: "body",
      weight: "regular",
      color: "primary",
      ...props
    }
  );
  const {
    className,
    href,
    title,
    children,
    onPress,
    variant,
    weight,
    color,
    truncate,
    standalone,
    slot,
    ...restProps
  } = cleanedProps;
  const internalRef = useRef(null);
  const linkRef = ref || internalRef;
  const { linkProps } = useLink(
    {
      href,
      onPress,
      ...restProps
    },
    linkRef
  );
  return /* @__PURE__ */ jsx(
    "a",
    {
      ...linkProps,
      ...dataAttributes,
      ...restProps,
      ref: linkRef,
      href,
      title,
      className: clsx(classNames.root, styles[classNames.root], className),
      children
    }
  );
});
LinkInternal.displayName = "LinkInternal";
const Link = forwardRef((props, ref) => {
  return /* @__PURE__ */ jsx(InternalLinkProvider, { href: props.href, children: /* @__PURE__ */ jsx(LinkInternal, { ...props, ref }) });
});
Link.displayName = "Link";

export { Link };
