// @backstage/plugin-scaffolder/dist/components/fields/RepoOwnerPicker/DefaultRepoOwnerPicker.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import TextField from '@material-ui/core/TextField';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';

const DefaultRepoOwnerPicker = ({
  onChange,
  state,
  rawErrors,
  isDisabled,
  required,
  schema
}) => {
  const { owner } = state;
  const { t } = useTranslationRef(scaffolderTranslationRef);
  return /* @__PURE__ */ jsxs(
    FormControl,
    {
      margin: "normal",
      required,
      error: rawErrors?.length > 0 && !owner,
      children: [
        /* @__PURE__ */ jsx(
          TextField,
          {
            id: "ownerInput",
            label: schema?.title ?? t("fields.repoOwnerPicker.title"),
            disabled: isDisabled,
            onChange: (e) => onChange({ owner: e.target.value }),
            value: owner
          }
        ),
        /* @__PURE__ */ jsx(FormHelperText, { children: schema?.description ?? t("fields.repoOwnerPicker.description") })
      ]
    }
  );
};

export { DefaultRepoOwnerPicker };
