// @backstage/plugin-scaffolder/dist/components/fields/RepoOwnerPicker/GitHubRepoOwnerPicker.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useApi } from '@backstage/core-plugin-api';
import { scaffolderApiRef } from '@backstage/plugin-scaffolder-react';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import { useState, useCallback } from 'react';
import useDebounce from 'react-use/esm/useDebounce';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../../translation.esm.js';

const GitHubRepoOwnerPicker = ({
  onChange,
  state,
  rawErrors,
  accessToken,
  isDisabled,
  required,
  schema,
  excludedOwners = []
}) => {
  const { host, owner } = state;
  const [availableOwners, setAvailableOwners] = useState([]);
  const scaffolderApi = useApi(scaffolderApiRef);
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const updateAvailableOwners = useCallback(() => {
    if (!scaffolderApi.autocomplete || !accessToken || !host) {
      setAvailableOwners([]);
      return;
    }
    scaffolderApi.autocomplete({
      token: accessToken,
      resource: "owners",
      context: { host },
      provider: "github"
    }).then(({ results }) => {
      const owners = results.map((r) => r.id).filter((id) => !excludedOwners.includes(id));
      setAvailableOwners(owners);
    }).catch(() => {
      setAvailableOwners([]);
    });
  }, [host, accessToken, scaffolderApi, excludedOwners]);
  useDebounce(updateAvailableOwners, 500, [updateAvailableOwners]);
  return /* @__PURE__ */ jsxs(
    FormControl,
    {
      margin: "normal",
      required,
      error: rawErrors?.length > 0 && !owner,
      children: [
        /* @__PURE__ */ jsx(
          Autocomplete,
          {
            value: owner,
            onChange: (_, newValue) => {
              onChange({ owner: newValue || "" });
            },
            disabled: isDisabled,
            options: availableOwners,
            renderInput: (params) => /* @__PURE__ */ jsx(
              TextField,
              {
                ...params,
                label: schema?.title ?? t("fields.repoOwnerPicker.title"),
                disabled: isDisabled,
                required
              }
            ),
            freeSolo: true,
            autoSelect: true
          }
        ),
        /* @__PURE__ */ jsx(FormHelperText, { children: schema?.description ?? t("fields.repoOwnerPicker.description") })
      ]
    }
  );
};

export { GitHubRepoOwnerPicker };
