// @backstage/plugin-scaffolder/dist/components/fields/RepoOwnerPicker/RepoOwnerPicker.esm.js
import { jsx } from 'react/jsx-runtime';
import { useApi } from '@backstage/core-plugin-api';
import { scmIntegrationsApiRef, scmAuthApiRef } from '@backstage/integration-react';
import { useState, useMemo, useEffect, useCallback } from 'react';
import useDebounce from 'react-use/esm/useDebounce';
import { useTemplateSecrets } from '@backstage/plugin-scaffolder-react';
import { DefaultRepoOwnerPicker } from './DefaultRepoOwnerPicker.esm.js';
import { GitHubRepoOwnerPicker } from './GitHubRepoOwnerPicker.esm.js';

const RepoOwnerPicker = (props) => {
  const { uiSchema, onChange, rawErrors, formData, schema, required } = props;
  const [state, setState] = useState({
    owner: formData || ""
  });
  const excludedOwners = useMemo(
    () => uiSchema?.["ui:options"]?.excludedOwners ?? [],
    [uiSchema]
  );
  const { host, owner } = state;
  const integrationApi = useApi(scmIntegrationsApiRef);
  const scmAuthApi = useApi(scmAuthApiRef);
  const { secrets, setSecrets } = useTemplateSecrets();
  useDebounce(
    async () => {
      const { requestUserCredentials } = uiSchema?.["ui:options"] ?? {};
      if (!requestUserCredentials || !host) {
        return;
      }
      if (secrets[requestUserCredentials.secretsKey]) {
        return;
      }
      const { token } = await scmAuthApi.getCredentials({
        url: `https://${host}`,
        additionalScope: {
          repoWrite: true,
          customScopes: requestUserCredentials.additionalScopes
        }
      });
      setSecrets({ [requestUserCredentials.secretsKey]: token });
    },
    500,
    [host, uiSchema]
  );
  useEffect(() => {
    if (uiSchema?.["ui:options"]?.host) {
      const hostUiOption = uiSchema["ui:options"].host;
      setState((prevState) => ({ ...prevState, host: hostUiOption }));
    }
  }, [uiSchema]);
  useEffect(() => {
    onChange(owner);
  }, [owner, onChange]);
  const updateLocalState = useCallback(
    (newState) => {
      setState((prevState) => ({ ...prevState, ...newState }));
    },
    [setState]
  );
  const hostType = (host && integrationApi.byHost(host)?.type) ?? null;
  const renderRepoOwnerPicker = () => {
    switch (hostType) {
      case "github":
        return /* @__PURE__ */ jsx(
          GitHubRepoOwnerPicker,
          {
            onChange: updateLocalState,
            state,
            rawErrors,
            accessToken: uiSchema?.["ui:options"]?.requestUserCredentials?.secretsKey && secrets[uiSchema["ui:options"].requestUserCredentials.secretsKey],
            isDisabled: uiSchema?.["ui:disabled"] ?? false,
            required,
            schema,
            excludedOwners
          }
        );
      default:
        return /* @__PURE__ */ jsx(
          DefaultRepoOwnerPicker,
          {
            onChange: updateLocalState,
            state,
            rawErrors,
            isDisabled: uiSchema?.["ui:disabled"] ?? false,
            required,
            schema
          }
        );
    }
  };
  return renderRepoOwnerPicker();
};

export { RepoOwnerPicker };
