// @backstage/ui/dist/components/PluginHeader/PluginHeader.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { PluginHeaderToolbar } from './PluginHeaderToolbar.esm.js';
import { Tabs, TabList, Tab } from '../Tabs/Tabs.esm.js';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { PluginHeaderDefinition } from './definition.esm.js';
import { useRef } from 'react';
import { useIsomorphicLayoutEffect } from '../../hooks/useIsomorphicLayoutEffect.esm.js';
import styles from './PluginHeader.module.css.esm.js';
import clsx from 'clsx';

const PluginHeader = (props) => {
  const { classNames, cleanedProps } = useStyles(PluginHeaderDefinition, props);
  const {
    className,
    tabs,
    icon,
    title,
    titleLink,
    customActions,
    onTabSelectionChange
  } = cleanedProps;
  const hasTabs = tabs && tabs.length > 0;
  const headerRef = useRef(null);
  useIsomorphicLayoutEffect(() => {
    const el = headerRef.current;
    if (!el) return void 0;
    const updateHeight = () => {
      const height = el.offsetHeight;
      document.documentElement.style.setProperty(
        "--bui-header-height",
        `${height}px`
      );
    };
    updateHeight();
    if (typeof ResizeObserver === "undefined") {
      return () => {
        document.documentElement.style.removeProperty("--bui-header-height");
      };
    }
    const observer = new ResizeObserver(updateHeight);
    observer.observe(el);
    return () => {
      observer.disconnect();
      document.documentElement.style.removeProperty("--bui-header-height");
    };
  }, []);
  return /* @__PURE__ */ jsxs(
    "header",
    {
      ref: headerRef,
      className: clsx(classNames.root, styles[classNames.root], className),
      children: [
        /* @__PURE__ */ jsx(
          PluginHeaderToolbar,
          {
            icon,
            title,
            titleLink,
            customActions,
            hasTabs
          }
        ),
        tabs && /* @__PURE__ */ jsx(
          "div",
          {
            className: clsx(
              classNames.tabsWrapper,
              styles[classNames.tabsWrapper]
            ),
            children: /* @__PURE__ */ jsx(Tabs, { onSelectionChange: onTabSelectionChange, children: /* @__PURE__ */ jsx(TabList, { children: tabs?.map((tab) => /* @__PURE__ */ jsx(
              Tab,
              {
                id: tab.id,
                href: tab.href,
                matchStrategy: tab.matchStrategy,
                children: tab.label
              },
              tab.id
            )) }) })
          }
        )
      ]
    }
  );
};

export { PluginHeader };
