// @backstage/plugin-catalog-react/dist/components/EntityInfoCard/EntityInfoCard.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { Card, CardHeader, Flex, Text, CardBody, CardFooter } from '@backstage/ui';
import { makeStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

const useStyles = makeStyles({
  root: {
    height: "100%"
  },
  footer: {
    display: "flex",
    justifyContent: "flex-end"
  }
});
function EntityInfoCard(props) {
  const { title, headerActions, footerActions, children, className } = props;
  const classes = useStyles();
  return /* @__PURE__ */ jsxs(Card, { className: classNames(classes.root, className), children: [
    title && /* @__PURE__ */ jsx(CardHeader, { children: /* @__PURE__ */ jsxs(Flex, { justify: "between", align: "center", children: [
      /* @__PURE__ */ jsx(Text, { as: "h3", variant: "title-x-small", weight: "bold", children: title }),
      headerActions && /* @__PURE__ */ jsx(Flex, { align: "center", gap: "1", children: headerActions })
    ] }) }),
    /* @__PURE__ */ jsx(CardBody, { children }),
    footerActions && /* @__PURE__ */ jsx(CardFooter, { className: classes.footer, children: footerActions })
  ] });
}

export { EntityInfoCard };
