// @backstage/plugin-scaffolder-backend-module-bitbucket-cloud/dist/actions/helpers.cjs.js
'use strict';

var bitbucket = require('bitbucket');
var integration = require('@backstage/integration');

const getBitbucketClient = async (config) => {
  if (config.clientId && config.clientSecret) {
    const token = await integration.getBitbucketCloudOAuthToken(
      config.clientId,
      config.clientSecret
    );
    return new bitbucket.Bitbucket({
      auth: {
        token
      }
    });
  }
  if (config.token && !config.username) {
    return new bitbucket.Bitbucket({
      auth: {
        token: config.token
      }
    });
  }
  if (config.username && config.token) {
    return new bitbucket.Bitbucket({
      auth: {
        username: config.username,
        password: config.token
      }
    });
  }
  if (config.username && config.appPassword) {
    return new bitbucket.Bitbucket({
      auth: {
        username: config.username,
        password: config.appPassword
      }
    });
  }
  throw new Error(
    `Authorization has not been provided for Bitbucket Cloud. Please provide either OAuth credentials (clientId/clientSecret), username and token, or username and appPassword in the Integrations config`
  );
};
const getAuthorizationHeader = async (config) => {
  if (config.clientId && config.clientSecret) {
    const token = await integration.getBitbucketCloudOAuthToken(
      config.clientId,
      config.clientSecret
    );
    return `Bearer ${token}`;
  }
  if (config.username && (config.token ?? config.appPassword)) {
    const buffer = Buffer.from(
      `${config.username}:${config.token ?? config.appPassword}`,
      "utf8"
    );
    return `Basic ${buffer.toString("base64")}`;
  }
  if (config.token) {
    return `Bearer ${config.token}`;
  }
  throw new Error(
    `Authorization has not been provided for Bitbucket Cloud. Please provide either OAuth credentials (clientId/clientSecret), username and token, or username and appPassword in the Integrations config`
  );
};
const getGitAuth = async (config) => {
  if (config.clientId && config.clientSecret) {
    const token = await integration.getBitbucketCloudOAuthToken(
      config.clientId,
      config.clientSecret
    );
    return {
      username: "x-token-auth",
      password: token
    };
  }
  if (config.token && !config.username) {
    return {
      username: "x-token-auth",
      password: config.token
    };
  }
  if (config.username && config.token) {
    return {
      username: "x-bitbucket-api-token-auth",
      password: config.token
    };
  }
  if (config.username && config.appPassword) {
    return {
      username: config.username,
      password: config.appPassword
    };
  }
  throw new Error(
    `Authorization has not been provided for Bitbucket Cloud. Please provide either OAuth credentials (clientId/clientSecret), username and token, or username and appPassword in the Integrations config`
  );
};

exports.getAuthorizationHeader = getAuthorizationHeader;
exports.getBitbucketClient = getBitbucketClient;
exports.getGitAuth = getGitAuth;
