// @backstage/ui/dist/components/Table/hooks/useOffsetPagination.esm.js
import { useState, useCallback } from 'react';
import { usePageCache } from './usePageCache.esm.js';
import { useStableCallback } from './useStableCallback.esm.js';
import { useDebouncedReload } from './useDebouncedReload.esm.js';
import { getEffectivePageSize } from './getEffectivePageSize.esm.js';

function useOffsetPagination(options, query) {
  const { getData: getDataProp, paginationOptions = {} } = options;
  const { initialOffset = 0 } = paginationOptions;
  const defaultPageSize = getEffectivePageSize(paginationOptions);
  const getData = useStableCallback(getDataProp);
  const { sort, filter, search } = query;
  const [pageSize, setPageSize] = useState(defaultPageSize);
  const wrappedGetData = useCallback(
    async ({
      cursor,
      signal
    }) => {
      const currentOffset = cursor ?? 0;
      const params = {
        offset: currentOffset,
        pageSize,
        sort,
        filter,
        search,
        signal
      };
      const response = await getData(params);
      const prevCursor = currentOffset > 0 ? Math.max(0, currentOffset - pageSize) : void 0;
      const nextCursor = currentOffset + pageSize < response.totalCount ? currentOffset + pageSize : void 0;
      return {
        data: response.data,
        prevCursor,
        nextCursor,
        totalCount: response.totalCount
      };
    },
    [getData, pageSize, sort, filter, search]
  );
  const cache = usePageCache({
    getData: wrappedGetData,
    initialCurrentCursor: initialOffset > 0 ? initialOffset : void 0
  });
  useDebouncedReload(query, pageSize, cache.reload);
  const onPageSizeChange = useCallback(
    (newSize) => setPageSize(newSize),
    []
  );
  return {
    data: cache.data,
    loading: cache.loading,
    error: cache.error,
    totalCount: cache.totalCount,
    offset: cache.currentCursor ?? 0,
    pageSize,
    hasNextPage: cache.hasNextPage,
    hasPreviousPage: cache.hasPreviousPage,
    onNextPage: cache.onNextPage,
    onPreviousPage: cache.onPreviousPage,
    onPageSizeChange,
    reload: cache.reload
  };
}

export { useOffsetPagination };
