// @backstage/plugin-catalog-react/dist/alpha/blueprints/EntityContentBlueprint.esm.js
import { createExtensionBlueprint, coreExtensionData, ExtensionBoundary } from '@backstage/frontend-plugin-api';
import { entityContentIconDataRef, entityContentGroupDataRef, entityFilterExpressionDataRef, entityFilterFunctionDataRef, entityContentTitleDataRef } from './extensionData.esm.js';
import { createZodV3FilterPredicateSchema } from '@backstage/filter-predicates';
import { resolveEntityFilterData } from './resolveEntityFilterData.esm.js';

const EntityContentBlueprint = createExtensionBlueprint({
  kind: "entity-content",
  attachTo: { id: "page:catalog/entity", input: "contents" },
  output: [
    coreExtensionData.reactElement,
    coreExtensionData.routePath,
    entityContentTitleDataRef,
    coreExtensionData.routeRef.optional(),
    entityFilterFunctionDataRef.optional(),
    entityFilterExpressionDataRef.optional(),
    entityContentGroupDataRef.optional(),
    entityContentIconDataRef.optional()
  ],
  dataRefs: {
    title: entityContentTitleDataRef,
    filterFunction: entityFilterFunctionDataRef,
    filterExpression: entityFilterExpressionDataRef,
    group: entityContentGroupDataRef,
    icon: entityContentIconDataRef
  },
  config: {
    schema: {
      path: (z) => z.string().optional(),
      title: (z) => z.string().optional(),
      filter: (z) => z.union([z.string(), createZodV3FilterPredicateSchema(z)]).optional(),
      group: (z) => z.literal(false).or(z.string()).optional(),
      icon: (z) => z.string().optional()
    }
  },
  *factory(params, { node, config }) {
    const path = config.path ?? params.path ?? params.defaultPath;
    const title = config.title ?? params.title ?? params.defaultTitle;
    const icon = config.icon ?? params.icon;
    const group = config.group ?? params.group ?? params.defaultGroup;
    yield coreExtensionData.reactElement(
      ExtensionBoundary.lazy(node, params.loader)
    );
    yield coreExtensionData.routePath(path);
    yield entityContentTitleDataRef(title);
    if (params.routeRef) {
      yield coreExtensionData.routeRef(params.routeRef);
    }
    yield* resolveEntityFilterData(params.filter, config, node);
    if (group && typeof group === "string") {
      yield entityContentGroupDataRef(group);
    }
    if (icon) {
      yield entityContentIconDataRef(icon);
    }
  }
});

export { EntityContentBlueprint };
