// @backstage/ui/dist/components/Table/components/Table.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { useId } from 'react-aria';
import { ResizableTableContainer } from 'react-aria-components';
import { TableRoot } from './TableRoot.esm.js';
import { TableHeader } from './TableHeader.esm.js';
import { TableBody } from './TableBody.esm.js';
import { Row } from './Row.esm.js';
import { Column } from './Column.esm.js';
import { TablePagination } from '../../TablePagination/TablePagination.esm.js';
import { useMemo } from 'react';
import { VisuallyHidden } from '../../VisuallyHidden/VisuallyHidden.esm.js';
import { Flex } from '../../Flex/Flex.esm.js';

function isRowRenderFn(rowConfig) {
  return typeof rowConfig === "function";
}
function useDisabledRows({
  data,
  rowConfig
}) {
  return useMemo(() => {
    if (!data || typeof rowConfig === "function" || !rowConfig?.getIsDisabled) {
      return;
    }
    return data.reduce((set, item) => {
      const isDisabled = rowConfig.getIsDisabled?.(item);
      if (isDisabled) {
        set.add(String(item.id));
      }
      return set;
    }, /* @__PURE__ */ new Set());
  }, [data, rowConfig]);
}
function useLiveRegionLabel(pagination, isStale, hasData) {
  if (!hasData || pagination.type === "none") {
    return "";
  }
  const { pageSize, offset, totalCount, getLabel } = pagination;
  if (isStale) {
    return "Loading table data.";
  }
  let liveRegionLabel = "Table page loaded. ";
  if (getLabel) {
    liveRegionLabel += getLabel({ pageSize, offset, totalCount });
  } else if (offset !== void 0) {
    const fromCount = offset + 1;
    const toCount = Math.min(offset + pageSize, totalCount ?? 0);
    liveRegionLabel += `Showing ${fromCount} to ${toCount} of ${totalCount}`;
  }
  return liveRegionLabel;
}
function Table({
  columnConfig,
  data,
  loading = false,
  isStale = false,
  error,
  pagination,
  sort,
  rowConfig,
  selection,
  emptyState,
  className,
  style
}) {
  const liveRegionId = useId();
  const visibleColumns = useMemo(
    () => columnConfig.filter((col) => !col.isHidden),
    [columnConfig]
  );
  const disabledRows = useDisabledRows({ data, rowConfig });
  const {
    mode: selectionMode,
    selected: selectedKeys,
    behavior: selectionBehavior,
    onSelectionChange
  } = selection || {};
  if (loading && !data) {
    return /* @__PURE__ */ jsx("div", { className, style, children: "Loading..." });
  }
  if (error) {
    return /* @__PURE__ */ jsxs("div", { className, style, children: [
      "Error: ",
      error.message
    ] });
  }
  const liveRegionLabel = useLiveRegionLabel(
    pagination,
    isStale,
    data !== void 0
  );
  const manualColumnSizing = columnConfig.some(
    (col) => col.width != null || col.minWidth != null || col.maxWidth != null || col.defaultWidth != null
  );
  const wrapResizable = manualColumnSizing ? (elem) => /* @__PURE__ */ jsx(ResizableTableContainer, { children: elem }) : (elem) => /* @__PURE__ */ jsx(Fragment, { children: elem });
  return /* @__PURE__ */ jsxs("div", { className, style, children: [
    /* @__PURE__ */ jsx(VisuallyHidden, { "aria-live": "polite", id: liveRegionId, children: liveRegionLabel }),
    wrapResizable(
      /* @__PURE__ */ jsxs(
        TableRoot,
        {
          selectionMode,
          selectionBehavior,
          selectedKeys,
          onSelectionChange,
          sortDescriptor: sort?.descriptor ?? void 0,
          onSortChange: sort?.onSortChange,
          disabledKeys: disabledRows,
          stale: isStale,
          "aria-describedby": liveRegionId,
          children: [
            /* @__PURE__ */ jsx(TableHeader, { columns: visibleColumns, children: (column) => column.header ? column.header() : /* @__PURE__ */ jsx(
              Column,
              {
                id: column.id,
                isRowHeader: column.isRowHeader,
                allowsSorting: column.isSortable,
                width: column.width,
                defaultWidth: column.defaultWidth,
                minWidth: column.minWidth,
                maxWidth: column.maxWidth,
                children: column.label
              }
            ) }),
            /* @__PURE__ */ jsx(
              TableBody,
              {
                items: data,
                dependencies: [visibleColumns],
                renderEmptyState: emptyState ? () => /* @__PURE__ */ jsx(Flex, { p: "3", children: emptyState }) : void 0,
                children: (item) => {
                  const itemIndex = data?.indexOf(item) ?? -1;
                  if (isRowRenderFn(rowConfig)) {
                    return rowConfig({
                      item,
                      index: itemIndex
                    });
                  }
                  return /* @__PURE__ */ jsx(
                    Row,
                    {
                      id: String(item.id),
                      columns: visibleColumns,
                      href: rowConfig?.getHref?.(item),
                      onAction: rowConfig?.onClick ? () => rowConfig?.onClick?.(item) : void 0,
                      children: (column) => column.cell(item)
                    }
                  );
                }
              }
            )
          ]
        }
      )
    ),
    pagination.type === "page" && /* @__PURE__ */ jsx(
      TablePagination,
      {
        pageSize: pagination.pageSize,
        pageSizeOptions: pagination.pageSizeOptions,
        offset: pagination.offset,
        totalCount: pagination.totalCount,
        hasNextPage: pagination.hasNextPage,
        hasPreviousPage: pagination.hasPreviousPage,
        onNextPage: pagination.onNextPage,
        onPreviousPage: pagination.onPreviousPage,
        onPageSizeChange: pagination.onPageSizeChange,
        showPageSizeOptions: pagination.showPageSizeOptions,
        getLabel: pagination.getLabel
      }
    )
  ] });
}

export { Table };
