// @backstage/ui/dist/components/Table/components/TableHeader.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { useTableOptions, TableHeader as TableHeader$1, Collection } from 'react-aria-components';
import { Checkbox } from '../../Checkbox/Checkbox.esm.js';
import { Column } from './Column.esm.js';
import { useStyles } from '../../../hooks/useStyles.esm.js';
import { TableDefinition } from '../definition.esm.js';
import styles from '../Table.module.css.esm.js';
import clsx from 'clsx';
import { Flex } from '../../Flex/Flex.esm.js';

const TableHeader = (props) => {
  let { selectionBehavior, selectionMode } = useTableOptions();
  const { classNames, cleanedProps } = useStyles(TableDefinition, props);
  const { columns, children, ...rest } = cleanedProps;
  return /* @__PURE__ */ jsxs(
    TableHeader$1,
    {
      className: clsx(classNames.header, styles[classNames.header]),
      ...rest,
      children: [
        selectionBehavior === "toggle" && selectionMode === "multiple" && /* @__PURE__ */ jsx(
          Column,
          {
            width: 40,
            minWidth: 40,
            maxWidth: 40,
            className: clsx(
              classNames.headSelection,
              styles[classNames.headSelection]
            ),
            children: /* @__PURE__ */ jsx(Flex, { justify: "center", align: "center", children: /* @__PURE__ */ jsx(Checkbox, { slot: "selection", children: /* @__PURE__ */ jsx(Fragment, {}) }) })
          }
        ),
        /* @__PURE__ */ jsx(Collection, { items: columns, children })
      ]
    }
  );
};

export { TableHeader };
