// @backstage/plugin-catalog/dist/alpha/components/EntityTabs/EntityTabsGroup.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { forwardRef, useState } from 'react';
import { Link } from 'react-router-dom';
import classnames from 'classnames';
import Typography from '@material-ui/core/Typography';
import Popover from '@material-ui/core/Popover';
import { capitalize } from '@material-ui/core/utils';
import { withStyles, createStyles } from '@material-ui/core/styles';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import Button from '@material-ui/core/Button';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import List from '@material-ui/core/List';
import { useApi } from '@backstage/core-plugin-api';
import { iconsApiRef } from '@backstage/frontend-plugin-api';

const styles = (theme) => createStyles({
  /* Styles applied to the root element. */
  root: {
    ...theme.typography.button,
    maxWidth: 264,
    minWidth: 72,
    position: "relative",
    boxSizing: "border-box",
    minHeight: 48,
    flexShrink: 0,
    padding: "6px 12px",
    [theme.breakpoints.up("sm")]: {
      padding: "6px 24px"
    },
    overflow: "hidden",
    whiteSpace: "normal",
    textAlign: "center",
    [theme.breakpoints.up("sm")]: {
      minWidth: 160
    }
  },
  defaultTab: {
    ...theme.typography.caption,
    padding: theme.spacing(3, 3),
    textTransform: "uppercase",
    fontWeight: theme.typography.fontWeightBold,
    color: theme.palette.text.secondary
  },
  /* Styles applied to the root element if both `icon` and `label` are provided. */
  labelIcon: {
    minHeight: 72,
    paddingTop: 9,
    "& $wrapper > *:first-child": {
      marginBottom: 6
    }
  },
  /* Styles applied to the root element if the parent [`Tabs`](/api/tabs/) has `textColor="inherit"`. */
  textColorInherit: {
    color: "inherit",
    opacity: 0.7,
    "&$selected": {
      opacity: 1
    },
    "&$disabled": {
      opacity: 0.5
    }
  },
  selectedButton: {
    color: `${theme.palette.text.primary}`,
    opacity: `${1}`
  },
  unselectedButton: {
    color: `${theme.palette.text.secondary}`,
    opacity: `${0.7}`
  },
  /* Styles applied to the root element if the parent [`Tabs`](/api/tabs/) has `textColor="primary"`. */
  textColorPrimary: {
    color: theme.palette.text.secondary,
    "&$selected": {
      color: theme.palette.primary.main
    },
    "&$disabled": {
      color: theme.palette.text.disabled
    }
  },
  /* Styles applied to the root element if the parent [`Tabs`](/api/tabs/) has `textColor="secondary"`. */
  textColorSecondary: {
    color: theme.palette.text.secondary,
    "&$selected": {
      color: theme.palette.secondary.main
    },
    "&$disabled": {
      color: theme.palette.text.disabled
    }
  },
  /* Pseudo-class applied to the root element if `selected={true}` (controlled by the Tabs component). */
  selected: {},
  /* Pseudo-class applied to the root element if `disabled={true}` (controlled by the Tabs component). */
  disabled: {},
  /* Styles applied to the root element if `fullWidth={true}` (controlled by the Tabs component). */
  fullWidth: {
    flexShrink: 1,
    flexGrow: 1,
    flexBasis: 0,
    maxWidth: "none"
  },
  /* Styles applied to the root element if `wrapped={true}`. */
  wrapped: {
    fontSize: theme.typography.pxToRem(12),
    lineHeight: 1.5
  },
  /* Styles applied to the `icon` and `label`'s wrapper element. */
  wrapper: {
    display: "inline-flex",
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
    flexDirection: "row"
  }
});
function resolveIcon(icon, iconsApi, showIcons) {
  if (!showIcons) {
    return void 0;
  }
  if (typeof icon === "string") {
    const Icon = iconsApi.getIcon(icon);
    if (Icon) {
      return /* @__PURE__ */ jsx(Icon, {});
    }
    return void 0;
  }
  return icon;
}
const Tab = forwardRef(function Tab2(props, ref) {
  const [anchorEl, setAnchorEl] = useState(null);
  const iconsApi = useApi(iconsApiRef);
  const open = Boolean(anchorEl);
  const submenuId = open ? "tabbed-submenu" : void 0;
  const {
    classes,
    className,
    disabled = false,
    disableFocusRipple = false,
    items,
    fullWidth,
    indicator,
    label,
    onSelectTab,
    selected,
    textColor = "inherit",
    wrapped = false,
    highlightedButton,
    showIcons = false
  } = props;
  const groupIcon = resolveIcon(props.icon, iconsApi, showIcons);
  const testId = "data-testid" in props && props["data-testid"];
  const handleMenuClose = () => {
    setAnchorEl(null);
  };
  const handleMenuClick = (event) => {
    setAnchorEl(event.currentTarget);
  };
  const classArray = [
    classes?.root,
    classes?.[`textColor${capitalize(textColor)}`],
    classes && {
      [classes.disabled]: disabled,
      [classes.selected]: selected,
      [classes.labelIcon]: label && groupIcon,
      [classes.fullWidth]: fullWidth,
      [classes.wrapped]: wrapped
    },
    className
  ];
  if (items.length === 1) {
    return /* @__PURE__ */ jsxs(
      Button,
      {
        focusRipple: !disableFocusRipple,
        "data-testid": testId,
        className: classnames(
          classArray,
          classes && {
            [classes.labelIcon]: label && (items[0].icon ?? groupIcon)
          }
        ),
        ref,
        role: "tab",
        "aria-selected": selected,
        disabled,
        component: Link,
        onClick: onSelectTab,
        to: items[0]?.path,
        startIcon: resolveIcon(items[0].icon, iconsApi, showIcons),
        children: [
          /* @__PURE__ */ jsx(Typography, { className: classes?.wrapper, variant: "button", children: items[0].label }),
          indicator
        ]
      }
    );
  }
  const hasIcons = showIcons && items.some((i) => i.icon);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(
      Button,
      {
        "data-testid": testId,
        focusRipple: !disableFocusRipple,
        className: classnames(classArray),
        ref,
        role: "tab",
        "aria-selected": selected,
        disabled,
        onClick: handleMenuClick,
        startIcon: groupIcon,
        children: [
          /* @__PURE__ */ jsx(Typography, { className: classes?.wrapper, variant: "button", children: label }),
          /* @__PURE__ */ jsx(ExpandMoreIcon, {})
        ]
      }
    ),
    /* @__PURE__ */ jsx(
      Popover,
      {
        id: submenuId,
        open,
        anchorEl,
        onClose: handleMenuClose,
        anchorOrigin: {
          vertical: "bottom",
          horizontal: "center"
        },
        transformOrigin: {
          vertical: "top",
          horizontal: "center"
        },
        children: /* @__PURE__ */ jsx(List, { component: "nav", children: items.map((i) => {
          const itemIcon = resolveIcon(i.icon, iconsApi, showIcons);
          return /* @__PURE__ */ jsxs(
            ListItem,
            {
              button: true,
              focusRipple: !disableFocusRipple,
              classes: {
                selected: classnames(classes?.selectedButton),
                default: classnames(classes?.unselectedButton),
                disabled: classnames(classes?.disabled)
              },
              ref,
              "aria-selected": selected,
              disabled,
              selected: highlightedButton === i.id,
              component: Link,
              onClick: (e) => {
                handleMenuClose();
                onSelectTab?.(e);
              },
              to: i.path,
              children: [
                itemIcon && /* @__PURE__ */ jsx(ListItemIcon, { children: itemIcon }),
                /* @__PURE__ */ jsx(
                  ListItemText,
                  {
                    inset: !itemIcon && hasIcons,
                    primary: /* @__PURE__ */ jsxs(Fragment, { children: [
                      /* @__PURE__ */ jsx(Typography, { variant: "button", children: i.label }),
                      indicator
                    ] })
                  }
                )
              ]
            },
            `popover_item_${i.id}`
          );
        }) })
      }
    )
  ] });
});
const EntityTabsGroup = withStyles(styles, { name: "MuiTab" })(Tab);

export { EntityTabsGroup };
