// @backstage/plugin-catalog/dist/alpha/components/EntityTabs/EntityTabsList.esm.js
import { jsx } from 'react/jsx-runtime';
import { useMemo } from 'react';
import Box from '@material-ui/core/Box';
import Tabs from '@material-ui/core/Tabs';
import { makeStyles } from '@material-ui/core/styles';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { EntityTabsGroup } from './EntityTabsGroup.esm.js';
import { catalogTranslationRef } from '../../translation.esm.js';

const useStyles = makeStyles(
  (theme) => ({
    tabsWrapper: {
      gridArea: "pageSubheader",
      backgroundColor: theme.palette.background.paper,
      paddingLeft: theme.spacing(3),
      minWidth: 0
    },
    defaultTab: {
      ...theme.typography.caption,
      padding: theme.spacing(3, 3),
      textTransform: "uppercase",
      fontWeight: theme.typography.fontWeightBold,
      color: theme.palette.text.secondary
    },
    selected: {
      color: theme.palette.text.primary
    },
    tabRoot: {
      "&:hover": {
        backgroundColor: theme.palette.background.default,
        color: theme.palette.text.primary
      }
    }
  }),
  { name: "BackstageHeaderTabs" }
);
function EntityTabsList(props) {
  const styles = useStyles();
  const { t } = useTranslationRef(catalogTranslationRef);
  const { tabs: items, selectedIndex = 0, showIcons, groupDefinitions } = props;
  const groups = useMemo(() => {
    const byKey = items.reduce((result, tab) => {
      const group = tab.group ? groupDefinitions[tab.group] : void 0;
      const groupOrId = group && tab.group ? tab.group : tab.id;
      result[groupOrId] = result[groupOrId] ?? {
        group,
        items: []
      };
      result[groupOrId].items.push(tab);
      return result;
    }, {});
    const groupOrder = Object.keys(groupDefinitions);
    return Object.entries(byKey).sort(([a], [b]) => {
      const ai = groupOrder.indexOf(a);
      const bi = groupOrder.indexOf(b);
      if (ai !== -1 && bi !== -1) {
        return ai - bi;
      }
      if (ai !== -1) {
        return -1;
      }
      if (bi !== -1) {
        return 1;
      }
      return 0;
    });
  }, [items, groupDefinitions]);
  const selectedItem = items[selectedIndex];
  return /* @__PURE__ */ jsx(Box, { className: styles.tabsWrapper, children: /* @__PURE__ */ jsx(
    Tabs,
    {
      selectionFollowsFocus: true,
      indicatorColor: "primary",
      textColor: "inherit",
      variant: "scrollable",
      scrollButtons: "auto",
      "aria-label": t("entityTabs.tabsAriaLabel"),
      value: selectedItem?.group ?? selectedItem?.id,
      children: groups.map(([id, tabGroup]) => /* @__PURE__ */ jsx(
        EntityTabsGroup,
        {
          "data-testid": `header-tab-${id}`,
          className: styles.defaultTab,
          classes: { selected: styles.selected, root: styles.tabRoot },
          label: tabGroup.group?.title,
          icon: tabGroup.group?.icon,
          value: id,
          items: tabGroup.items,
          highlightedButton: selectedItem?.id,
          showIcons
        },
        id
      ))
    }
  ) });
}

export { EntityTabsList };
