// @backstage/frontend-plugin-api/dist/blueprints/PageBlueprint.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { Routes, Route, Navigate } from 'react-router-dom';
import { coreExtensionData } from '../wiring/coreExtensionData.esm.js';
import 'zod';
import 'zod-to-json-schema';
import { createExtensionInput } from '../wiring/createExtensionInput.esm.js';
import { createExtensionBlueprint } from '../wiring/createExtensionBlueprint.esm.js';
import { ExtensionBoundary } from '../components/ExtensionBoundary.esm.js';
import '../apis/definitions/AppTreeApi.esm.js';
import '../apis/definitions/auth.esm.js';
import '../apis/definitions/AlertApi.esm.js';
import '../apis/definitions/AppLanguageApi.esm.js';
import '../apis/definitions/AppThemeApi.esm.js';
import '../apis/definitions/SwappableComponentsApi.esm.js';
import '../apis/definitions/ConfigApi.esm.js';
import '../apis/definitions/DiscoveryApi.esm.js';
import '../apis/definitions/ErrorApi.esm.js';
import '../apis/definitions/FeatureFlagsApi.esm.js';
import '../apis/definitions/FetchApi.esm.js';
import '../apis/definitions/IconsApi.esm.js';
import '../apis/definitions/IdentityApi.esm.js';
import '../apis/definitions/DialogApi.esm.js';
import '../apis/definitions/OAuthRequestApi.esm.js';
import '../apis/definitions/RouteResolutionApi.esm.js';
import '../apis/definitions/StorageApi.esm.js';
import '../apis/definitions/AnalyticsApi.esm.js';
import '../apis/definitions/TranslationApi.esm.js';
import { pluginHeaderActionsApiRef } from '../apis/definitions/PluginHeaderActionsApi.esm.js';
import { useApi } from '../apis/system/useApi.esm.js';
import 'react';
import '../components/AppNodeProvider.esm.js';
import '../components/DefaultSwappableComponents.esm.js';
import { PageLayout } from '../components/PageLayout.esm.js';

const PageBlueprint = createExtensionBlueprint({
  kind: "page",
  attachTo: { id: "app/routes", input: "routes" },
  inputs: {
    pages: createExtensionInput([
      coreExtensionData.routePath,
      coreExtensionData.routeRef.optional(),
      coreExtensionData.reactElement,
      coreExtensionData.title.optional(),
      coreExtensionData.icon.optional()
    ])
  },
  output: [
    coreExtensionData.routePath,
    coreExtensionData.reactElement,
    coreExtensionData.routeRef.optional(),
    coreExtensionData.title.optional(),
    coreExtensionData.icon.optional()
  ],
  config: {
    schema: {
      path: (z) => z.string().optional(),
      title: (z) => z.string().optional()
    }
  },
  *factory(params, { config, node, inputs }) {
    const title = config.title ?? params.title;
    const icon = params.icon;
    const pluginId = node.spec.plugin.pluginId;
    const noHeader = params.noHeader ?? false;
    yield coreExtensionData.routePath(config.path ?? params.path);
    if (params.loader) {
      const loader = params.loader;
      const PageContent = () => {
        const headerActionsApi = useApi(pluginHeaderActionsApiRef);
        const headerActions = headerActionsApi.getPluginHeaderActions(pluginId);
        return /* @__PURE__ */ jsx(
          PageLayout,
          {
            title: title ?? node.spec.plugin.title ?? node.spec.plugin.pluginId,
            icon: icon ?? node.spec.plugin.icon,
            noHeader,
            headerActions,
            children: ExtensionBoundary.lazy(node, loader)
          }
        );
      };
      yield coreExtensionData.reactElement(/* @__PURE__ */ jsx(PageContent, {}));
    } else if (inputs.pages.length > 0) {
      const tabs = inputs.pages.map((page) => {
        const path = page.get(coreExtensionData.routePath);
        const tabTitle = page.get(coreExtensionData.title);
        const tabIcon = page.get(coreExtensionData.icon);
        return {
          id: path,
          label: tabTitle || path,
          icon: tabIcon,
          href: path
        };
      });
      const PageContent = () => {
        const firstPagePath = inputs.pages[0]?.get(coreExtensionData.routePath);
        const headerActionsApi = useApi(pluginHeaderActionsApiRef);
        const headerActions = headerActionsApi.getPluginHeaderActions(pluginId);
        return /* @__PURE__ */ jsx(
          PageLayout,
          {
            title,
            icon,
            tabs,
            headerActions,
            children: /* @__PURE__ */ jsxs(Routes, { children: [
              firstPagePath && /* @__PURE__ */ jsx(
                Route,
                {
                  index: true,
                  element: /* @__PURE__ */ jsx(Navigate, { to: firstPagePath, replace: true })
                }
              ),
              inputs.pages.map((page, index) => {
                const path = page.get(coreExtensionData.routePath);
                const element = page.get(coreExtensionData.reactElement);
                return /* @__PURE__ */ jsx(Route, { path: `${path}/*`, element }, index);
              })
            ] })
          }
        );
      };
      yield coreExtensionData.reactElement(/* @__PURE__ */ jsx(PageContent, {}));
    } else {
      const PageContent = () => {
        const headerActionsApi = useApi(pluginHeaderActionsApiRef);
        const headerActions = headerActionsApi.getPluginHeaderActions(pluginId);
        return /* @__PURE__ */ jsx(PageLayout, { title, icon, headerActions });
      };
      yield coreExtensionData.reactElement(/* @__PURE__ */ jsx(PageContent, {}));
    }
    if (params.routeRef) {
      yield coreExtensionData.routeRef(params.routeRef);
    }
    if (title) {
      yield coreExtensionData.title(title);
    }
    if (icon) {
      yield coreExtensionData.icon(icon);
    }
  }
});

export { PageBlueprint };
