// @backstage/frontend-plugin-api/dist/blueprints/SubPageBlueprint.esm.js
import { coreExtensionData } from '../wiring/coreExtensionData.esm.js';
import 'zod';
import 'zod-to-json-schema';
import { createExtensionBlueprint } from '../wiring/createExtensionBlueprint.esm.js';
import { ExtensionBoundary } from '../components/ExtensionBoundary.esm.js';
import 'react/jsx-runtime';
import '../apis/definitions/AppTreeApi.esm.js';
import '../apis/definitions/auth.esm.js';
import '../apis/definitions/AlertApi.esm.js';
import '../apis/definitions/AppLanguageApi.esm.js';
import '../apis/definitions/AppThemeApi.esm.js';
import '../apis/definitions/SwappableComponentsApi.esm.js';
import '../apis/definitions/ConfigApi.esm.js';
import '../apis/definitions/DiscoveryApi.esm.js';
import '../apis/definitions/ErrorApi.esm.js';
import '../apis/definitions/FeatureFlagsApi.esm.js';
import '../apis/definitions/FetchApi.esm.js';
import '../apis/definitions/IconsApi.esm.js';
import '../apis/definitions/IdentityApi.esm.js';
import '../apis/definitions/DialogApi.esm.js';
import '../apis/definitions/OAuthRequestApi.esm.js';
import '../apis/definitions/RouteResolutionApi.esm.js';
import '../apis/definitions/StorageApi.esm.js';
import '../apis/definitions/AnalyticsApi.esm.js';
import '../apis/definitions/TranslationApi.esm.js';
import '../apis/definitions/PluginHeaderActionsApi.esm.js';
import '@backstage/version-bridge';
import '@backstage/errors';
import 'react';
import '../components/AppNodeProvider.esm.js';
import '../components/DefaultSwappableComponents.esm.js';
import '../components/PageLayout.esm.js';

const SubPageBlueprint = createExtensionBlueprint({
  kind: "sub-page",
  attachTo: { relative: { kind: "page" }, input: "pages" },
  output: [
    coreExtensionData.routePath,
    coreExtensionData.reactElement,
    coreExtensionData.title,
    coreExtensionData.routeRef.optional(),
    coreExtensionData.icon.optional()
  ],
  config: {
    schema: {
      path: (z) => z.string().optional(),
      title: (z) => z.string().optional()
    }
  },
  *factory(params, { config, node }) {
    yield coreExtensionData.routePath(config.path ?? params.path);
    yield coreExtensionData.title(config.title ?? params.title);
    yield coreExtensionData.reactElement(
      ExtensionBoundary.lazy(node, params.loader)
    );
    if (params.routeRef) {
      yield coreExtensionData.routeRef(params.routeRef);
    }
    if (params.icon) {
      yield coreExtensionData.icon(params.icon);
    }
  }
});

export { SubPageBlueprint };
