// @backstage/plugin-catalog-react/dist/testUtils/createTestEntityPage.esm.js
import { jsx, jsxs, Fragment as Fragment$1 } from 'react/jsx-runtime';
import { Fragment } from 'react';
import { PageBlueprint, createExtensionInput, coreExtensionData, useApiHolder } from '@backstage/frontend-plugin-api';
import { Routes, Route } from 'react-router-dom';
import { EntityProvider } from '../hooks/useEntity.esm.js';
import '../alpha/blueprints/CatalogFilterBlueprint.esm.js';
import { EntityCardBlueprint } from '../alpha/blueprints/EntityCardBlueprint.esm.js';
import { EntityContentBlueprint } from '../alpha/blueprints/EntityContentBlueprint.esm.js';
import '../alpha/blueprints/EntityContentLayoutBlueprint.esm.js';
import '../alpha/blueprints/EntityHeaderBlueprint.esm.js';
import '../alpha/blueprints/extensionData.esm.js';
import '../alpha/blueprints/EntityContextMenuItemBlueprint.esm.js';
import '../alpha/blueprints/EntityIconLinkBlueprint.esm.js';
import { entityRouteRef } from '../routes.esm.js';
import { catalogApiRef } from '../api.esm.js';
import { catalogApiMock } from './catalogApiMock.esm.js';
import { TestApiProvider } from '@backstage/frontend-test-utils';

function createTestEntityPage(options) {
  const { entity } = options;
  return PageBlueprint.makeWithOverrides({
    name: "test-entity",
    inputs: {
      // Redirect cards from entity-content:catalog/overview
      cards: createExtensionInput(
        [
          coreExtensionData.reactElement,
          EntityContentBlueprint.dataRefs.filterFunction.optional(),
          EntityContentBlueprint.dataRefs.filterExpression.optional(),
          EntityCardBlueprint.dataRefs.type.optional()
        ],
        {
          replaces: [{ id: "entity-content:catalog/overview", input: "cards" }]
        }
      ),
      // Redirect contents from page:catalog/entity
      contents: createExtensionInput(
        [
          coreExtensionData.reactElement,
          coreExtensionData.routePath,
          coreExtensionData.routeRef.optional(),
          EntityContentBlueprint.dataRefs.title,
          EntityContentBlueprint.dataRefs.filterFunction.optional(),
          EntityContentBlueprint.dataRefs.filterExpression.optional(),
          EntityContentBlueprint.dataRefs.group.optional()
        ],
        {
          replaces: [{ id: "page:catalog/entity", input: "contents" }]
        }
      )
    },
    factory: (originalFactory, { inputs }) => {
      return originalFactory({
        path: "/",
        routeRef: entityRouteRef,
        loader: async () => {
          const cards = inputs.cards.filter(
            (card) => (card.get(EntityContentBlueprint.dataRefs.filterFunction) ?? (() => true))(entity)
          ).map((card) => ({
            element: card.get(coreExtensionData.reactElement),
            type: card.get(EntityCardBlueprint.dataRefs.type)
          }));
          const contents = inputs.contents.filter(
            (content) => (content.get(EntityContentBlueprint.dataRefs.filterFunction) ?? (() => true))(entity)
          ).map((content) => ({
            element: content.get(coreExtensionData.reactElement),
            path: content.get(coreExtensionData.routePath),
            title: content.get(EntityContentBlueprint.dataRefs.title)
          }));
          if (contents.length === 0 && cards.length === 0) {
            return /* @__PURE__ */ jsx("div", { "data-testid": "empty-entity-page" });
          }
          return /* @__PURE__ */ jsx(MockEntityApiProvider, { entity, children: /* @__PURE__ */ jsxs(EntityProvider, { entity, children: [
            cards.map((card, index) => /* @__PURE__ */ jsx(Fragment, { children: card.element }, index)),
            contents.length === 1 && contents[0].element,
            contents.length > 1 && /* @__PURE__ */ jsxs(Routes, { children: [
              contents.map((content) => /* @__PURE__ */ jsx(
                Route,
                {
                  path: content.path === "/" ? "/" : `${content.path.replace(/^\//, "")}/*`,
                  element: content.element
                },
                content.path
              )),
              /* @__PURE__ */ jsx(Route, { path: "*", element: contents[0].element })
            ] })
          ] }) });
        }
      });
    }
  });
}
function MockEntityApiProvider({
  entity,
  children
}) {
  const parentHolder = useApiHolder();
  if (parentHolder.get(catalogApiRef)) {
    return /* @__PURE__ */ jsx(Fragment$1, { children });
  }
  const mockApi = catalogApiMock({ entities: [entity] });
  return /* @__PURE__ */ jsx(TestApiProvider, { apis: [[catalogApiRef, mockApi]], children });
}

export { createTestEntityPage };
