// @backstage/ui/dist/components/SearchField/SearchField.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useEffect, useState, useRef } from 'react';
import { SearchField as SearchField$1, Input, Button } from 'react-aria-components';
import { FieldLabel } from '../FieldLabel/FieldLabel.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../FieldLabel/FieldLabel.module.css.esm.js';
import { FieldError } from '../FieldError/FieldError.esm.js';
import '../FieldError/FieldError.module.css.esm.js';
import { RiSearch2Line, RiCloseCircleLine } from '@remixicon/react';
import { SearchFieldDefinition } from './definition.esm.js';

const SearchField = forwardRef(
  (props, ref) => {
    const { ownProps, restProps, dataAttributes } = useDefinition(
      SearchFieldDefinition,
      props
    );
    const {
      classes,
      label,
      icon,
      secondaryLabel,
      placeholder,
      startCollapsed,
      description
    } = ownProps;
    useEffect(() => {
      if (!label && !restProps["aria-label"] && !restProps["aria-labelledby"]) {
        console.warn(
          "SearchField requires either a visible label, aria-label, or aria-labelledby for accessibility"
        );
      }
    }, [label, restProps["aria-label"], restProps["aria-labelledby"]]);
    const [isFocused, setIsFocused] = useState(false);
    const inputRef = useRef(null);
    const secondaryLabelText = secondaryLabel || (restProps.isRequired ? "Required" : null);
    const handleFocusChange = (isFocused2) => {
      restProps.onFocusChange?.(isFocused2);
      setIsFocused(isFocused2);
    };
    const handleContainerClick = () => {
      inputRef.current?.focus();
    };
    const hasInputRef = !!inputRef.current;
    const hasValue = !!inputRef.current?.value;
    const isCollapsed = hasInputRef ? startCollapsed && !hasValue && !isFocused : startCollapsed && !restProps.value && !restProps.defaultValue && !isFocused;
    return /* @__PURE__ */ jsxs(
      SearchField$1,
      {
        className: classes.root,
        ...dataAttributes,
        "data-collapsed": isCollapsed,
        ...restProps,
        onFocusChange: handleFocusChange,
        ref,
        children: [
          /* @__PURE__ */ jsx(
            FieldLabel,
            {
              label,
              secondaryLabel: secondaryLabelText,
              description
            }
          ),
          /* @__PURE__ */ jsxs(
            "div",
            {
              className: classes.inputWrapper,
              "data-size": dataAttributes["data-size"],
              onClick: handleContainerClick,
              children: [
                icon !== false && /* @__PURE__ */ jsx(
                  "div",
                  {
                    className: classes.inputIcon,
                    "data-size": dataAttributes["data-size"],
                    "aria-hidden": "true",
                    children: icon || /* @__PURE__ */ jsx(RiSearch2Line, {})
                  }
                ),
                /* @__PURE__ */ jsx(
                  Input,
                  {
                    ref: inputRef,
                    className: classes.input,
                    ...icon !== false && { "data-icon": true },
                    placeholder
                  }
                ),
                /* @__PURE__ */ jsx(
                  Button,
                  {
                    className: classes.clear,
                    "data-size": dataAttributes["data-size"],
                    children: /* @__PURE__ */ jsx(RiCloseCircleLine, {})
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ jsx(FieldError, {})
        ]
      }
    );
  }
);
SearchField.displayName = "searchField";

export { SearchField };
