// @backstage/ui/dist/components/Table/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Table.module.css.esm.js';

const TableWrapperDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TableWrapper",
    resizableContainer: "bui-TableResizableContainer"
  },
  propDefs: {
    className: {}
  }
});
const TableDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Table"
  },
  propDefs: {
    stale: { dataAttribute: true },
    loading: { dataAttribute: true }
  }
});
const TableHeaderDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TableHeader",
    headSelection: "bui-TableHeadSelection"
  },
  propDefs: {
    columns: {},
    children: {}
  }
});
const TableBodyDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TableBody"
  },
  propDefs: {}
});
const RowDefinition = defineComponent()({
  styles,
  analytics: true,
  bg: "consumer",
  classNames: {
    root: "bui-TableRow",
    cell: "bui-TableCell",
    cellSelection: "bui-TableCellSelection"
  },
  propDefs: {
    columns: {},
    children: {},
    href: {},
    noTrack: {}
  }
});
const ColumnDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TableHead",
    headContent: "bui-TableHeadContent",
    headLabel: "bui-TableHeadLabel",
    headSortButton: "bui-TableHeadSortButton"
  },
  propDefs: {
    children: {},
    className: {}
  }
});
const CellDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TableCell"
  },
  propDefs: {
    className: {}
  }
});
const CellTextDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TableCell",
    cellContentWrapper: "bui-TableCellContentWrapper",
    cellContent: "bui-TableCellContent",
    cellIcon: "bui-TableCellIcon"
  },
  propDefs: {
    title: {},
    description: {},
    color: { default: "primary" },
    leadingIcon: {},
    href: {},
    className: {}
  }
});
const CellProfileDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TableCell",
    cellContentWrapper: "bui-TableCellContentWrapper",
    cellContent: "bui-TableCellContent"
  },
  propDefs: {
    src: {},
    name: {},
    href: {},
    description: {},
    color: { default: "primary" },
    className: {}
  }
});

export { CellDefinition, CellProfileDefinition, CellTextDefinition, ColumnDefinition, RowDefinition, TableBodyDefinition, TableDefinition, TableHeaderDefinition, TableWrapperDefinition };
