// @backstage/integration/dist/bitbucketCloud/core.cjs.js
'use strict';

var fetch = require('cross-fetch');
var parseGitUrl = require('git-url-parse');
var helpers = require('../helpers.cjs.js');
var luxon = require('luxon');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fetch__default = /*#__PURE__*/_interopDefaultCompat(fetch);
var parseGitUrl__default = /*#__PURE__*/_interopDefaultCompat(parseGitUrl);

let cachedToken;
let refreshPromise;
async function getBitbucketCloudOAuthToken(clientId, clientSecret) {
  if (cachedToken && luxon.DateTime.now() < cachedToken.expiresAt) {
    return cachedToken.token;
  }
  if (refreshPromise) {
    return refreshPromise;
  }
  refreshPromise = (async () => {
    try {
      const credentials = Buffer.from(
        `${clientId}:${clientSecret}`,
        "utf8"
      ).toString("base64");
      const response = await fetch__default.default(
        "https://bitbucket.org/site/oauth2/access_token",
        {
          method: "POST",
          headers: {
            "Content-Type": "application/x-www-form-urlencoded",
            Authorization: `Basic ${credentials}`
          },
          body: "grant_type=client_credentials"
        }
      );
      if (!response.ok) {
        throw new Error(
          `Failed to fetch OAuth token from Bitbucket Cloud: ${response.status} ${response.statusText}`
        );
      }
      const data = await response.json();
      if (!data.access_token) {
        throw new Error("OAuth token response missing access_token field");
      }
      const expiresIn = data.expires_in || 3600;
      const expiresAt = luxon.DateTime.now().plus({ seconds: expiresIn }).minus({ minutes: 10 });
      cachedToken = {
        token: data.access_token,
        expiresAt
      };
      return data.access_token;
    } catch (error) {
      throw new Error(
        `Failed to fetch OAuth token for Bitbucket Cloud: ${error}`
      );
    } finally {
      refreshPromise = void 0;
    }
  })();
  return refreshPromise;
}
async function getBitbucketCloudDefaultBranch(url, config) {
  const { name: repoName, owner: project } = parseGitUrl__default.default(url);
  const branchUrl = `${config.apiBaseUrl}/repositories/${project}/${repoName}`;
  const response = await fetch__default.default(
    branchUrl,
    await getBitbucketCloudRequestOptions(config)
  );
  if (!response.ok) {
    const message = `Failed to retrieve default branch from ${branchUrl}, ${response.status} ${response.statusText}`;
    throw new Error(message);
  }
  const repoInfo = await response.json();
  const defaultBranch = repoInfo.mainbranch.name;
  if (!defaultBranch) {
    throw new Error(
      `Failed to read default branch from ${branchUrl}. Response ${response.status} ${response.json()}`
    );
  }
  return defaultBranch;
}
async function getBitbucketCloudDownloadUrl(url, config) {
  const {
    name: repoName,
    owner: project,
    ref,
    protocol,
    resource
  } = helpers.parseGitUrlSafe(url);
  let branch = ref;
  if (!branch) {
    branch = await getBitbucketCloudDefaultBranch(url, config);
  }
  return `${protocol}://${resource}/${project}/${repoName}/get/${branch}.tar.gz`;
}
function getBitbucketCloudFileFetchUrl(url, config) {
  try {
    const { owner, name, ref, filepathtype, filepath } = helpers.parseGitUrlSafe(url);
    if (!owner || !name || filepathtype !== "src" && filepathtype !== "raw") {
      throw new Error("Invalid Bitbucket Cloud URL or file path");
    }
    const pathWithoutSlash = filepath.replace(/^\//, "");
    if (!ref) {
      throw new Error("Invalid Bitbucket Cloud URL or file path");
    }
    return `${config.apiBaseUrl}/repositories/${owner}/${name}/src/${ref}/${pathWithoutSlash}`;
  } catch (e) {
    throw new Error(`Incorrect URL: ${url}, ${e}`);
  }
}
async function getBitbucketCloudRequestOptions(config) {
  const headers = {};
  if (config.clientId && config.clientSecret) {
    const token = await getBitbucketCloudOAuthToken(
      config.clientId,
      config.clientSecret
    );
    headers.Authorization = `Bearer ${token}`;
    return { headers };
  }
  if (config.username && (config.token ?? config.appPassword)) {
    const buffer = Buffer.from(
      `${config.username}:${config.token ?? config.appPassword}`,
      "utf8"
    );
    headers.Authorization = `Basic ${buffer.toString("base64")}`;
  }
  return { headers };
}

exports.getBitbucketCloudDefaultBranch = getBitbucketCloudDefaultBranch;
exports.getBitbucketCloudDownloadUrl = getBitbucketCloudDownloadUrl;
exports.getBitbucketCloudFileFetchUrl = getBitbucketCloudFileFetchUrl;
exports.getBitbucketCloudOAuthToken = getBitbucketCloudOAuthToken;
exports.getBitbucketCloudRequestOptions = getBitbucketCloudRequestOptions;
