// @backstage/plugin-scaffolder/dist/components/OngoingTask/OngoingTaskContextMenu.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { Header, MenuTrigger, ButtonIcon, Menu, MenuItem } from '@backstage/ui';
import { RiMore2Line, RiFileListLine, RiAddCircleLine, RiRepeatLine, RiReplay10Line, RiCloseCircleLine } from '@remixicon/react';
import { usePermission } from '@backstage/plugin-permission-react';
import { taskReadPermission, taskCreatePermission } from '@backstage/plugin-scaffolder-common/alpha';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { scaffolderTranslationRef } from '../../translation.esm.js';

function OngoingTaskContextMenu(props) {
  const {
    title,
    cancelEnabled,
    canRetry,
    isRetryableTask,
    logsVisible,
    buttonBarVisible,
    onRetry,
    onStartOver,
    onToggleLogs,
    onToggleButtonBar,
    taskId
  } = props;
  const { t } = useTranslationRef(scaffolderTranslationRef);
  const { allowed: canReadTask } = usePermission({
    permission: taskReadPermission,
    resourceRef: taskId
  });
  const { allowed: canCreateTask } = usePermission({
    permission: taskCreatePermission
  });
  const canStartOver = canReadTask && canCreateTask;
  return /* @__PURE__ */ jsx(
    Header,
    {
      title,
      customActions: /* @__PURE__ */ jsxs(MenuTrigger, { children: [
        /* @__PURE__ */ jsx(
          ButtonIcon,
          {
            variant: "tertiary",
            icon: /* @__PURE__ */ jsx(RiMore2Line, {}),
            "aria-label": t("ongoingTask.contextMenu.moreOptions")
          }
        ),
        /* @__PURE__ */ jsxs(Menu, { placement: "bottom end", children: [
          /* @__PURE__ */ jsx(
            MenuItem,
            {
              onAction: () => onToggleLogs?.(!logsVisible),
              iconStart: /* @__PURE__ */ jsx(RiFileListLine, { size: 16 }),
              children: logsVisible ? t("ongoingTask.contextMenu.hideLogs") : t("ongoingTask.contextMenu.showLogs")
            }
          ),
          /* @__PURE__ */ jsx(
            MenuItem,
            {
              onAction: () => onToggleButtonBar?.(!buttonBarVisible),
              iconStart: /* @__PURE__ */ jsx(RiAddCircleLine, { size: 16 }),
              children: buttonBarVisible ? t("ongoingTask.contextMenu.hideButtonBar") : t("ongoingTask.contextMenu.showButtonBar")
            }
          ),
          /* @__PURE__ */ jsx(
            MenuItem,
            {
              onAction: onStartOver,
              isDisabled: cancelEnabled || !canStartOver,
              iconStart: /* @__PURE__ */ jsx(RiRepeatLine, { size: 16 }),
              children: t("ongoingTask.contextMenu.startOver")
            }
          ),
          isRetryableTask && /* @__PURE__ */ jsx(
            MenuItem,
            {
              onAction: onRetry,
              isDisabled: cancelEnabled || !canRetry,
              iconStart: /* @__PURE__ */ jsx(RiReplay10Line, { size: 16 }),
              children: t("ongoingTask.contextMenu.retry")
            }
          ),
          /* @__PURE__ */ jsx(
            MenuItem,
            {
              onAction: props.onCancel,
              isDisabled: props.isCancelButtonDisabled,
              iconStart: /* @__PURE__ */ jsx(RiCloseCircleLine, { size: 16 }),
              color: "danger",
              children: t("ongoingTask.contextMenu.cancel")
            }
          )
        ] })
      ] })
    }
  );
}

export { OngoingTaskContextMenu };
