// @backstage/plugin-scaffolder/dist/alpha/extensions.esm.js
import { jsx } from 'react/jsx-runtime';
import { PageBlueprint, createExtensionInput, SubPageBlueprint, NavItemBlueprint, ApiBlueprint, identityApiRef, fetchApiRef, discoveryApiRef } from '@backstage/frontend-plugin-api';
import { rootRouteRef } from '../routes.esm.js';
import CreateComponentIcon from '@material-ui/icons/AddCircleOutline';
import { FormFieldBlueprint, formFieldsApiRef } from '@backstage/plugin-scaffolder-react/alpha';
import { scmIntegrationsApiRef } from '@backstage/integration-react';
import { scaffolderApiRef } from '@backstage/plugin-scaffolder-react';
import { ScaffolderClient } from '../api.esm.js';

const scaffolderPage = PageBlueprint.makeWithOverrides({
  inputs: {
    formFields: createExtensionInput([
      FormFieldBlueprint.dataRefs.formFieldLoader
    ])
  },
  factory(originalFactory) {
    return originalFactory({
      routeRef: rootRouteRef,
      path: "/create",
      title: "Create"
    });
  }
});
const scaffolderTemplatesSubPage = SubPageBlueprint.makeWithOverrides({
  name: "templates",
  factory(originalFactory, { apis }) {
    const formFieldsApi = apis.get(formFieldsApiRef);
    return originalFactory({
      path: "templates",
      title: "Templates",
      loader: async () => {
        const formFields = await formFieldsApi?.loadFormFields() ?? [];
        return import('./components/TemplatesSubPage.esm.js').then((m) => /* @__PURE__ */ jsx(m.TemplatesSubPage, { formFields }));
      }
    });
  }
});
const scaffolderTasksSubPage = SubPageBlueprint.make({
  name: "tasks",
  params: {
    path: "tasks",
    title: "Tasks",
    loader: () => import('./components/TasksSubPage.esm.js').then((m) => /* @__PURE__ */ jsx(m.TasksSubPage, {}))
  }
});
const scaffolderActionsSubPage = SubPageBlueprint.make({
  name: "actions",
  params: {
    path: "actions",
    title: "Actions",
    loader: () => Promise.all([
      import('../components/ActionsPage/index.esm.js'),
      import('@backstage/core-components')
    ]).then(([m, { Content }]) => /* @__PURE__ */ jsx(Content, { children: /* @__PURE__ */ jsx(m.ActionPageContent, {}) }))
  }
});
const scaffolderEditorSubPage = SubPageBlueprint.make({
  name: "editor",
  params: {
    path: "edit",
    title: "Template Editor",
    loader: () => import('./components/EditorSubPage.esm.js').then((m) => /* @__PURE__ */ jsx(m.EditorSubPage, {}))
  }
});
const scaffolderTemplatingExtensionsSubPage = SubPageBlueprint.make({
  name: "templating-extensions",
  params: {
    path: "templating-extensions",
    title: "Templating Extensions",
    loader: () => Promise.all([
      import('../components/TemplatingExtensionsPage/index.esm.js'),
      import('@backstage/core-components')
    ]).then(([m, { Content }]) => /* @__PURE__ */ jsx(Content, { children: /* @__PURE__ */ jsx(m.TemplatingExtensionsPageContent, { linkLocal: true }) }))
  }
});
const scaffolderNavItem = NavItemBlueprint.make({
  params: {
    routeRef: rootRouteRef,
    title: "Create...",
    icon: CreateComponentIcon
  }
});
const repoUrlPickerFormField = FormFieldBlueprint.make({
  name: "repo-url-picker",
  params: {
    field: () => import('./fields/RepoUrlPicker.esm.js').then((m) => m.RepoUrlPicker)
  }
});
const entityNamePickerFormField = FormFieldBlueprint.make({
  name: "entity-name-picker",
  params: {
    field: () => import('./fields/EntityNamePicker.esm.js').then((m) => m.EntityNamePicker)
  }
});
const entityPickerFormField = FormFieldBlueprint.make({
  name: "entity-picker",
  params: {
    field: () => import('./fields/EntityPicker.esm.js').then((m) => m.EntityPicker)
  }
});
const ownerPickerFormField = FormFieldBlueprint.make({
  name: "owner-picker",
  params: {
    field: () => import('./fields/OwnerPicker.esm.js').then((m) => m.OwnerPicker)
  }
});
const entityTagsPickerFormField = FormFieldBlueprint.make({
  name: "entity-tags-picker",
  params: {
    field: () => import('./fields/EntityTagsPicker.esm.js').then((m) => m.EntityTagsPicker)
  }
});
const multiEntityPickerFormField = FormFieldBlueprint.make({
  name: "multi-entity-picker",
  params: {
    field: () => import('./fields/MultiEntityPicker.esm.js').then((m) => m.MultiEntityPicker)
  }
});
const myGroupsPickerFormField = FormFieldBlueprint.make({
  name: "my-groups-picker",
  params: {
    field: () => import('./fields/MyGroupsPicker.esm.js').then((m) => m.MyGroupsPicker)
  }
});
const ownedEntityPickerFormField = FormFieldBlueprint.make({
  name: "owned-entity-picker",
  params: {
    field: () => import('./fields/OwnedEntityPicker.esm.js').then((m) => m.OwnedEntityPicker)
  }
});
const repoBranchPickerFormField = FormFieldBlueprint.make({
  name: "repo-branch-picker",
  params: {
    field: () => import('./fields/RepoBranchPicker.esm.js').then((m) => m.RepoBranchPicker)
  }
});
const repoOwnerPickerFormField = FormFieldBlueprint.make({
  name: "repo-owner-picker",
  params: {
    field: () => import('./fields/RepoOwnerPicker.esm.js').then((m) => m.RepoOwnerPicker)
  }
});
const scaffolderApi = ApiBlueprint.make({
  params: (defineParams) => defineParams({
    api: scaffolderApiRef,
    deps: {
      discoveryApi: discoveryApiRef,
      scmIntegrationsApi: scmIntegrationsApiRef,
      fetchApi: fetchApiRef,
      identityApi: identityApiRef
    },
    factory: ({ discoveryApi, scmIntegrationsApi, fetchApi, identityApi }) => new ScaffolderClient({
      discoveryApi,
      scmIntegrationsApi,
      fetchApi,
      identityApi
    })
  })
});

export { entityNamePickerFormField, entityPickerFormField, entityTagsPickerFormField, multiEntityPickerFormField, myGroupsPickerFormField, ownedEntityPickerFormField, ownerPickerFormField, repoBranchPickerFormField, repoOwnerPickerFormField, repoUrlPickerFormField, scaffolderActionsSubPage, scaffolderApi, scaffolderEditorSubPage, scaffolderNavItem, scaffolderPage, scaffolderTasksSubPage, scaffolderTemplatesSubPage, scaffolderTemplatingExtensionsSubPage };
