// @backstage/plugin-scaffolder/dist/alpha/formFieldsApi.esm.js
import { ApiBlueprint, createExtensionInput, appTreeApiRef } from '@backstage/frontend-plugin-api';
import { FormFieldBlueprint, formFieldsApiRef } from '@backstage/plugin-scaffolder-react/alpha';
export { formFieldsApiRef } from '@backstage/plugin-scaffolder-react/alpha';
import { OpaqueFormField } from '../packages/scaffolder-internal/src/wiring/InternalFormField.esm.js';
import '../packages/scaffolder-internal/src/wiring/InternalFormDecorator.esm.js';

const formFieldsApi = ApiBlueprint.makeWithOverrides({
  name: "form-fields",
  inputs: {
    formFields: createExtensionInput([
      FormFieldBlueprint.dataRefs.formFieldLoader
    ])
  },
  factory(originalFactory, { inputs }) {
    const formFieldLoaders = inputs.formFields.map(
      (e) => e.get(FormFieldBlueprint.dataRefs.formFieldLoader)
    );
    return originalFactory(
      (defineParams) => defineParams({
        api: formFieldsApiRef,
        deps: { appTreeApi: appTreeApiRef },
        factory: ({ appTreeApi }) => ({
          async loadFormFields() {
            const pageFormFieldLoaders = getPageFormFieldLoaders(appTreeApi);
            const formFields = await Promise.all(
              [...formFieldLoaders, ...pageFormFieldLoaders].map(
                (loader) => loader()
              )
            );
            return formFields.map(OpaqueFormField.toInternal);
          }
        })
      })
    );
  }
});
function getPageFormFieldLoaders(appTreeApi) {
  const { tree } = appTreeApi.getTree();
  const pageNode = tree.nodes.get("page:scaffolder");
  if (!pageNode?.instance) {
    return [];
  }
  const formFieldNodes = pageNode.edges.attachments.get("formFields") ?? [];
  return formFieldNodes.flatMap((node) => {
    if (!node.instance) {
      return [];
    }
    const loader = node.instance.getData(
      FormFieldBlueprint.dataRefs.formFieldLoader
    );
    return loader ? [loader] : [];
  });
}

export { formFieldsApi };
