// @backstage/plugin-scaffolder/dist/alpha/plugin.esm.js
import { jsx } from 'react/jsx-runtime';
import { createFrontendPlugin } from '@backstage/frontend-plugin-api';
import CreateComponentIcon from '@material-ui/icons/AddCircleOutline';
import { viewTechDocRouteRef, registerComponentRouteRef, templatingExtensionsRouteRef, editRouteRef, scaffolderListTaskRouteRef, actionsRouteRef, scaffolderTaskRouteRef, selectedTemplateRouteRef, rootRouteRef } from '../routes.esm.js';
import { scaffolderApi, scaffolderPage, scaffolderTemplatesSubPage, scaffolderTasksSubPage, scaffolderActionsSubPage, scaffolderEditorSubPage, scaffolderTemplatingExtensionsSubPage, scaffolderNavItem, repoUrlPickerFormField, entityNamePickerFormField, entityPickerFormField, ownerPickerFormField, entityTagsPickerFormField, multiEntityPickerFormField, myGroupsPickerFormField, ownedEntityPickerFormField, repoBranchPickerFormField, repoOwnerPickerFormField } from './extensions.esm.js';
import { isTemplateEntityV1beta3 } from '@backstage/plugin-scaffolder-common';
import { formFieldsApi } from './formFieldsApi.esm.js';
import './api/ref.esm.js';
import { formDecoratorsApi } from './api/FormDecoratorsApi.esm.js';
import { EntityIconLinkBlueprint } from '@backstage/plugin-catalog-react/alpha';
import { useScaffolderTemplateIconLinkProps } from './hooks/useScaffolderTemplateIconLinkProps.esm.js';

const scaffolderEntityIconLink = EntityIconLinkBlueprint.make({
  name: "launch-template",
  params: {
    filter: isTemplateEntityV1beta3,
    useProps: useScaffolderTemplateIconLinkProps
  }
});
var plugin = createFrontendPlugin({
  pluginId: "scaffolder",
  title: "Create",
  icon: /* @__PURE__ */ jsx(CreateComponentIcon, { fontSize: "inherit" }),
  info: { packageJson: () => import('../plugins/scaffolder/package.json.esm.js') },
  routes: {
    root: rootRouteRef,
    selectedTemplate: selectedTemplateRouteRef,
    ongoingTask: scaffolderTaskRouteRef,
    actions: actionsRouteRef,
    listTasks: scaffolderListTaskRouteRef,
    edit: editRouteRef,
    templatingExtensions: templatingExtensionsRouteRef
  },
  externalRoutes: {
    registerComponent: registerComponentRouteRef,
    viewTechDoc: viewTechDocRouteRef
  },
  extensions: [
    scaffolderApi,
    scaffolderPage,
    scaffolderTemplatesSubPage,
    scaffolderTasksSubPage,
    scaffolderActionsSubPage,
    scaffolderEditorSubPage,
    scaffolderTemplatingExtensionsSubPage,
    scaffolderNavItem,
    scaffolderEntityIconLink,
    formDecoratorsApi,
    formFieldsApi,
    repoUrlPickerFormField,
    entityNamePickerFormField,
    entityPickerFormField,
    ownerPickerFormField,
    entityTagsPickerFormField,
    multiEntityPickerFormField,
    myGroupsPickerFormField,
    ownedEntityPickerFormField,
    repoBranchPickerFormField,
    repoOwnerPickerFormField
  ]
});

export { plugin as default };
