// @backstage/plugin-catalog-react/dist/components/EntityDataTable/EntityDataTable.esm.js
import { jsx } from 'react/jsx-runtime';
import { stringifyEntityRef } from '@backstage/catalog-model';
import { useTable, Table } from '@backstage/ui';
import { useMemo, useCallback } from 'react';

function EntityDataTable(props) {
  const { columnConfig, data, loading, error, emptyState } = props;
  const tableData = useMemo(
    () => data.map((entity) => ({
      ...entity,
      id: stringifyEntityRef(entity)
    })),
    [data]
  );
  const sortFn = useCallback(
    (items, sort) => {
      const column = columnConfig.find((c) => c.id === sort.column);
      if (!column?.sortValue) {
        return items;
      }
      const getValue = column.sortValue;
      const direction = sort.direction === "descending" ? -1 : 1;
      return [...items].sort(
        (a, b) => getValue(a).localeCompare(getValue(b)) * direction
      );
    },
    [columnConfig]
  );
  const { tableProps } = useTable({
    mode: "complete",
    data: tableData,
    sortFn,
    paginationOptions: { pageSize: tableData.length || 1 }
  });
  return /* @__PURE__ */ jsx(
    Table,
    {
      columnConfig,
      ...tableProps,
      loading,
      error,
      emptyState,
      pagination: { type: "none" }
    }
  );
}

export { EntityDataTable };
