// @backstage/ui/dist/components/Tabs/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Tabs.module.css.esm.js';

const TabsDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Tabs"
  },
  propDefs: {
    className: {},
    children: {}
  }
});
const TabListDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TabListWrapper",
    tabList: "bui-TabList"
  },
  propDefs: {
    className: {},
    children: {}
  }
});
const TabDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Tab"
  },
  analytics: true,
  propDefs: {
    className: {},
    matchStrategy: {},
    href: {},
    id: {},
    noTrack: {}
  }
});
const TabPanelDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TabPanel"
  },
  propDefs: {
    className: {}
  }
});
const TabsIndicatorsDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TabActive",
    hovered: "bui-TabHovered"
  },
  propDefs: {
    tabRefs: {},
    tabsRef: {},
    hoveredKey: {},
    prevHoveredKey: {}
  }
});

export { TabDefinition, TabListDefinition, TabPanelDefinition, TabsDefinition, TabsIndicatorsDefinition };
