// @backstage/ui/dist/components/List/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './List.module.css.esm.js';

const ListDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-List"
  },
  propDefs: {
    items: {},
    children: {},
    renderEmptyState: {},
    className: {}
  }
});
const ListRowDefinition = defineComponent()({
  styles,
  bg: "consumer",
  classNames: {
    root: "bui-ListRow",
    check: "bui-ListRowCheck",
    icon: "bui-ListRowIcon",
    label: "bui-ListRowLabel",
    description: "bui-ListRowDescription",
    actions: "bui-ListRowActions"
  },
  propDefs: {
    children: {},
    description: {},
    icon: {},
    menuItems: {},
    customActions: {},
    className: {}
  }
});

export { ListDefinition, ListRowDefinition };
