// @backstage/ui/dist/components/List/List.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { GridList, GridListItem, Text } from 'react-aria-components';
import { RiCheckLine, RiMoreLine } from '@remixicon/react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { ListDefinition, ListRowDefinition } from './definition.esm.js';
import { Box } from '../Box/Box.esm.js';
import { ButtonIcon } from '../ButtonIcon/ButtonIcon.esm.js';
import '../ButtonIcon/ButtonIcon.module.css.esm.js';
import { MenuTrigger, Menu } from '../Menu/Menu.esm.js';
import '../Menu/definition.esm.js';

const List = (props) => {
  const { ownProps, restProps, dataAttributes } = useDefinition(
    ListDefinition,
    props
  );
  const { classes, items, children, renderEmptyState } = ownProps;
  return /* @__PURE__ */ jsx(
    GridList,
    {
      className: classes.root,
      items,
      renderEmptyState,
      ...dataAttributes,
      ...restProps,
      children
    }
  );
};
const ListRow = (props) => {
  const { ownProps, restProps, dataAttributes } = useDefinition(
    ListRowDefinition,
    props
  );
  const { classes, children, description, icon, menuItems, customActions } = ownProps;
  const textValue = typeof children === "string" ? children : void 0;
  return /* @__PURE__ */ jsx(
    GridListItem,
    {
      textValue,
      className: classes.root,
      ...dataAttributes,
      ...restProps,
      children: ({ isSelected }) => /* @__PURE__ */ jsxs(Fragment, { children: [
        isSelected && /* @__PURE__ */ jsx("div", { className: classes.check, children: /* @__PURE__ */ jsx(RiCheckLine, {}) }),
        icon && /* @__PURE__ */ jsx(Box, { bg: "neutral", className: classes.icon, children: icon }),
        /* @__PURE__ */ jsxs("div", { className: classes.label, children: [
          /* @__PURE__ */ jsx("span", { children }),
          description && /* @__PURE__ */ jsx(Text, { slot: "description", className: classes.description, children: description })
        ] }),
        customActions && /* @__PURE__ */ jsx("div", { className: classes.actions, children: customActions }),
        menuItems && /* @__PURE__ */ jsx("div", { className: classes.actions, children: /* @__PURE__ */ jsxs(MenuTrigger, { children: [
          /* @__PURE__ */ jsx(
            ButtonIcon,
            {
              icon: /* @__PURE__ */ jsx(RiMoreLine, {}),
              size: "small",
              "aria-label": "More actions",
              variant: "tertiary"
            }
          ),
          /* @__PURE__ */ jsx(Menu, { placement: "bottom end", children: menuItems })
        ] }) })
      ] })
    }
  );
};

export { List, ListRow };
