// @backstage/frontend-plugin-api/dist/components/PageLayout.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { createSwappableComponent } from './createSwappableComponent.esm.js';

function DefaultPageLayout(props) {
  const { title, icon, headerActions, tabs, children } = props;
  return /* @__PURE__ */ jsxs(
    "div",
    {
      "data-component": "page-layout",
      style: {
        display: "flex",
        flexDirection: "column",
        flexGrow: 1,
        minHeight: 0
      },
      children: [
        (title || tabs) && /* @__PURE__ */ jsxs(
          "header",
          {
            style: {
              borderBottom: "1px solid #ddd",
              backgroundColor: "#fff",
              flexShrink: 0
            },
            children: [
              title && /* @__PURE__ */ jsxs(
                "div",
                {
                  style: {
                    display: "flex",
                    alignItems: "center",
                    gap: "8px",
                    padding: "12px 24px 8px",
                    fontSize: "18px",
                    fontWeight: 500
                  },
                  children: [
                    icon,
                    title,
                    headerActions && /* @__PURE__ */ jsx("div", { style: { marginLeft: "auto" }, children: headerActions })
                  ]
                }
              ),
              tabs && tabs.length > 0 && /* @__PURE__ */ jsx(
                "nav",
                {
                  style: {
                    display: "flex",
                    gap: "4px",
                    padding: "0 24px"
                  },
                  children: tabs.map((tab) => /* @__PURE__ */ jsxs(
                    "a",
                    {
                      href: tab.href,
                      style: {
                        display: "flex",
                        alignItems: "center",
                        gap: "4px",
                        padding: "8px 12px",
                        textDecoration: "none",
                        color: "#333",
                        borderBottom: "2px solid transparent"
                      },
                      children: [
                        tab.icon,
                        tab.label
                      ]
                    },
                    tab.id
                  ))
                }
              )
            ]
          }
        ),
        /* @__PURE__ */ jsx(
          "div",
          {
            style: {
              display: "flex",
              flexDirection: "column",
              flexGrow: 1,
              minHeight: 0
            },
            children
          }
        )
      ]
    }
  );
}
const PageLayout = createSwappableComponent({
  id: "core.page-layout",
  loader: () => DefaultPageLayout
});

export { PageLayout };
