// @backstage/ui/dist/components/Card/Card.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useRef, useCallback } from 'react';
import { Button } from 'react-aria-components';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { CardDefinition, CardHeaderDefinition, CardBodyDefinition, CardFooterDefinition } from './definition.esm.js';
import { Box } from '../Box/Box.esm.js';
import { Link } from '../Link/Link.esm.js';
import '../Link/Link.module.css.esm.js';

const INTERACTIVE_ELEMENT_SELECTOR = 'a[href],button,input,select,textarea,[role="button"],[role="link"],[tabindex]:not([tabindex="-1"])';
const Card = forwardRef((props, ref) => {
  const { ownProps, restProps, dataAttributes } = useDefinition(
    CardDefinition,
    props
  );
  const {
    classes,
    children,
    onPress,
    href,
    label,
    target: linkTarget,
    rel,
    download
  } = ownProps;
  const isInteractive = !!(onPress || href);
  const triggerRef = useRef(null);
  const handleClick = useCallback(
    (e) => {
      if (!isInteractive || !triggerRef.current) return;
      if (triggerRef.current.contains(e.target)) return;
      const targetNode = e.target;
      const targetElement = targetNode instanceof Element ? targetNode : targetNode?.parentElement;
      if (targetElement?.closest(INTERACTIVE_ELEMENT_SELECTOR)) return;
      if (window.getSelection()?.toString()) return;
      triggerRef.current.dispatchEvent(
        new MouseEvent("click", {
          bubbles: true,
          cancelable: true,
          ctrlKey: e.ctrlKey,
          metaKey: e.metaKey,
          shiftKey: e.shiftKey
        })
      );
    },
    [isInteractive]
  );
  return /* @__PURE__ */ jsxs(
    Box,
    {
      as: "article",
      bg: "neutral",
      ref,
      className: classes.root,
      "data-interactive": isInteractive || void 0,
      ...dataAttributes,
      ...restProps,
      onClick: isInteractive ? handleClick : void 0,
      children: [
        href && /* @__PURE__ */ jsx(
          Link,
          {
            ref: triggerRef,
            className: classes.trigger,
            href,
            target: linkTarget,
            rel,
            download,
            "aria-label": label
          }
        ),
        onPress && !href && /* @__PURE__ */ jsx(
          Button,
          {
            ref: triggerRef,
            className: classes.trigger,
            onPress,
            "aria-label": label
          }
        ),
        children
      ]
    }
  );
});
Card.displayName = "Card";
const CardHeader = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(CardHeaderDefinition, props);
    const { classes, children } = ownProps;
    return /* @__PURE__ */ jsx("div", { ref, className: classes.root, ...restProps, children });
  }
);
CardHeader.displayName = "CardHeader";
const CardBody = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(CardBodyDefinition, props);
    const { classes, children } = ownProps;
    return /* @__PURE__ */ jsx("div", { ref, className: classes.root, ...restProps, children });
  }
);
CardBody.displayName = "CardBody";
const CardFooter = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(CardFooterDefinition, props);
    const { classes, children } = ownProps;
    return /* @__PURE__ */ jsx("div", { ref, className: classes.root, ...restProps, children });
  }
);
CardFooter.displayName = "CardFooter";

export { Card, CardBody, CardFooter, CardHeader };
