// @backstage/ui/dist/components/Accordion/Accordion.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { Disclosure, Heading, Button, DisclosurePanel, DisclosureGroup } from 'react-aria-components';
import { RiArrowDownSLine } from '@remixicon/react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { AccordionDefinition, AccordionTriggerDefinition, AccordionPanelDefinition, AccordionGroupDefinition } from './definition.esm.js';
import { Flex } from '../Flex/Flex.esm.js';

const Accordion = forwardRef(
  (props, ref) => {
    const { ownProps, restProps, dataAttributes } = useDefinition(
      AccordionDefinition,
      props
    );
    const { classes } = ownProps;
    return /* @__PURE__ */ jsx(
      Disclosure,
      {
        ref,
        className: classes.root,
        ...dataAttributes,
        ...restProps
      }
    );
  }
);
Accordion.displayName = "Accordion";
const AccordionTrigger = forwardRef(
  (props, ref) => {
    const { ownProps, restProps, dataAttributes } = useDefinition(
      AccordionTriggerDefinition,
      props
    );
    const { classes, title, subtitle, children } = ownProps;
    return /* @__PURE__ */ jsx(
      Heading,
      {
        ref,
        className: classes.root,
        ...dataAttributes,
        ...restProps,
        children: /* @__PURE__ */ jsxs(Button, { slot: "trigger", className: classes.button, children: [
          children ? children : /* @__PURE__ */ jsxs(Flex, { gap: "2", align: "center", children: [
            /* @__PURE__ */ jsx("span", { className: classes.title, children: title }),
            subtitle && /* @__PURE__ */ jsx("span", { className: classes.subtitle, children: subtitle })
          ] }),
          /* @__PURE__ */ jsx(RiArrowDownSLine, { className: classes.icon, size: 16 })
        ] })
      }
    );
  }
);
AccordionTrigger.displayName = "AccordionTrigger";
const AccordionPanel = forwardRef(
  (props, ref) => {
    const { ownProps, restProps, dataAttributes } = useDefinition(
      AccordionPanelDefinition,
      props
    );
    const { classes } = ownProps;
    return /* @__PURE__ */ jsx(
      DisclosurePanel,
      {
        ref,
        className: classes.root,
        ...dataAttributes,
        ...restProps
      }
    );
  }
);
AccordionPanel.displayName = "AccordionPanel";
const AccordionGroup = forwardRef(
  (props, ref) => {
    const { ownProps, restProps, dataAttributes } = useDefinition(
      AccordionGroupDefinition,
      props
    );
    const { classes, allowsMultiple } = ownProps;
    return /* @__PURE__ */ jsx(
      DisclosureGroup,
      {
        ref,
        allowsMultipleExpanded: allowsMultiple,
        className: classes.root,
        ...dataAttributes,
        ...restProps
      }
    );
  }
);
AccordionGroup.displayName = "AccordionGroup";

export { Accordion, AccordionGroup, AccordionPanel, AccordionTrigger };
