// @backstage/ui/dist/components/Menu/Menu.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { MenuTrigger as MenuTrigger$1, SubmenuTrigger as SubmenuTrigger$1, Menu as Menu$1, Popover, Virtualizer, ListLayout, ListBox, useFilter, Autocomplete, SearchField, Input, Button, MenuItem as MenuItem$1, ListBoxItem, MenuSection as MenuSection$1, Header, Separator } from 'react-aria-components';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { MenuDefinition, MenuListBoxDefinition, MenuAutocompleteDefinition, MenuAutocompleteListboxDefinition, MenuItemDefinition, MenuListBoxItemDefinition, MenuSectionDefinition, MenuSeparatorDefinition, MenuEmptyStateDefinition } from './definition.esm.js';
import { RiCloseCircleLine, RiArrowRightSLine, RiCheckLine } from '@remixicon/react';
import { isInternalLink } from '../../utils/linkUtils.esm.js';
import { getNodeText } from '../../analytics/getNodeText.esm.js';
import { Box } from '../Box/Box.esm.js';
import '../Box/Box.module.css.esm.js';
import { BgReset } from '../../hooks/useBg.esm.js';

const rowHeight = 32;
const MenuEmptyState = () => {
  const { ownProps } = useDefinition(MenuEmptyStateDefinition, {});
  return /* @__PURE__ */ jsx("div", { className: ownProps.classes.root, children: "No results found." });
};
const MenuTrigger = (props) => {
  return /* @__PURE__ */ jsx(MenuTrigger$1, { ...props });
};
const SubmenuTrigger = (props) => {
  return /* @__PURE__ */ jsx(SubmenuTrigger$1, { ...props });
};
const Menu = (props) => {
  const { ownProps, restProps } = useDefinition(MenuDefinition, props);
  const { classes, placement, virtualized, maxWidth, maxHeight, style } = ownProps;
  let newMaxWidth = maxWidth || (virtualized ? "260px" : "undefined");
  const menuContent = /* @__PURE__ */ jsx(
    Menu$1,
    {
      className: classes.content,
      renderEmptyState: () => /* @__PURE__ */ jsx(MenuEmptyState, {}),
      style: { width: newMaxWidth, maxHeight, ...style },
      ...restProps
    }
  );
  return /* @__PURE__ */ jsx(Popover, { className: classes.root, placement, children: /* @__PURE__ */ jsx(BgReset, { children: /* @__PURE__ */ jsx(Box, { bg: "neutral", className: classes.inner, children: virtualized ? /* @__PURE__ */ jsx(
    Virtualizer,
    {
      layout: ListLayout,
      layoutOptions: {
        rowHeight
      },
      children: menuContent
    }
  ) : menuContent }) }) });
};
const MenuListBox = (props) => {
  const { ownProps, restProps } = useDefinition(MenuListBoxDefinition, props);
  const {
    classes,
    selectionMode,
    placement,
    virtualized,
    maxWidth,
    maxHeight,
    style
  } = ownProps;
  let newMaxWidth = maxWidth || (virtualized ? "260px" : "undefined");
  const listBoxContent = /* @__PURE__ */ jsx(
    ListBox,
    {
      className: classes.content,
      selectionMode,
      style: { width: newMaxWidth, maxHeight, ...style },
      ...restProps
    }
  );
  return /* @__PURE__ */ jsx(Popover, { className: classes.root, placement, children: /* @__PURE__ */ jsx(BgReset, { children: /* @__PURE__ */ jsx(Box, { bg: "neutral", className: classes.inner, children: virtualized ? /* @__PURE__ */ jsx(
    Virtualizer,
    {
      layout: ListLayout,
      layoutOptions: {
        rowHeight
      },
      children: listBoxContent
    }
  ) : listBoxContent }) }) });
};
const MenuAutocomplete = (props) => {
  const { ownProps, restProps } = useDefinition(
    MenuAutocompleteDefinition,
    props
  );
  const {
    classes,
    placement,
    virtualized,
    maxWidth,
    maxHeight,
    style,
    placeholder
  } = ownProps;
  const { contains } = useFilter({ sensitivity: "base" });
  let newMaxWidth = maxWidth || (virtualized ? "260px" : "undefined");
  const menuContent = /* @__PURE__ */ jsx(
    Menu$1,
    {
      className: classes.content,
      renderEmptyState: () => /* @__PURE__ */ jsx(MenuEmptyState, {}),
      style: { width: newMaxWidth, maxHeight, ...style },
      ...restProps
    }
  );
  return /* @__PURE__ */ jsx(Popover, { className: classes.root, placement, children: /* @__PURE__ */ jsx(BgReset, { children: /* @__PURE__ */ jsx(Box, { bg: "neutral", className: classes.inner, children: /* @__PURE__ */ jsxs(Autocomplete, { filter: contains, children: [
    /* @__PURE__ */ jsxs(
      SearchField,
      {
        className: classes.searchField,
        "aria-label": placeholder || "Search",
        children: [
          /* @__PURE__ */ jsx(
            Input,
            {
              className: classes.searchFieldInput,
              placeholder: placeholder || "Search..."
            }
          ),
          /* @__PURE__ */ jsx(Button, { className: classes.searchFieldClear, children: /* @__PURE__ */ jsx(RiCloseCircleLine, {}) })
        ]
      }
    ),
    virtualized ? /* @__PURE__ */ jsx(
      Virtualizer,
      {
        layout: ListLayout,
        layoutOptions: {
          rowHeight
        },
        children: menuContent
      }
    ) : menuContent
  ] }) }) }) });
};
const MenuAutocompleteListbox = (props) => {
  const { ownProps, restProps } = useDefinition(
    MenuAutocompleteListboxDefinition,
    props
  );
  const {
    classes,
    selectionMode,
    placement,
    virtualized,
    maxWidth,
    maxHeight,
    style,
    placeholder
  } = ownProps;
  const { contains } = useFilter({ sensitivity: "base" });
  let newMaxWidth = maxWidth || (virtualized ? "260px" : "undefined");
  const listBoxContent = /* @__PURE__ */ jsx(
    ListBox,
    {
      className: classes.content,
      renderEmptyState: () => /* @__PURE__ */ jsx(MenuEmptyState, {}),
      selectionMode,
      style: { width: newMaxWidth, maxHeight, ...style },
      ...restProps
    }
  );
  return /* @__PURE__ */ jsx(Popover, { className: classes.root, placement, children: /* @__PURE__ */ jsx(BgReset, { children: /* @__PURE__ */ jsx(Box, { bg: "neutral", className: classes.inner, children: /* @__PURE__ */ jsxs(Autocomplete, { filter: contains, children: [
    /* @__PURE__ */ jsxs(
      SearchField,
      {
        className: classes.searchField,
        "aria-label": placeholder || "Search",
        children: [
          /* @__PURE__ */ jsx(
            Input,
            {
              className: classes.searchFieldInput,
              placeholder: placeholder || "Search..."
            }
          ),
          /* @__PURE__ */ jsx(Button, { className: classes.searchFieldClear, children: /* @__PURE__ */ jsx(RiCloseCircleLine, {}) })
        ]
      }
    ),
    virtualized ? /* @__PURE__ */ jsx(
      Virtualizer,
      {
        layout: ListLayout,
        layoutOptions: {
          rowHeight
        },
        children: listBoxContent
      }
    ) : listBoxContent
  ] }) }) }) });
};
const MenuItem = (props) => {
  const { ownProps, restProps, dataAttributes, analytics } = useDefinition(
    MenuItemDefinition,
    props
  );
  const { classes, iconStart, children, href } = ownProps;
  const handleAction = () => {
    if (href) {
      const text = restProps["aria-label"] ?? getNodeText(children) ?? String(href);
      analytics.captureEvent("click", text, {
        attributes: { to: String(href) }
      });
    }
  };
  if (href && !isInternalLink(href)) {
    return /* @__PURE__ */ jsx(
      MenuItem$1,
      {
        className: classes.root,
        ...dataAttributes,
        textValue: typeof children === "string" ? children : void 0,
        ...restProps,
        onAction: () => {
          restProps.onAction?.();
          handleAction();
          window.open(href, "_blank", "noopener,noreferrer");
        },
        children: /* @__PURE__ */ jsxs("div", { className: classes.itemWrapper, children: [
          /* @__PURE__ */ jsxs("div", { className: classes.itemContent, children: [
            iconStart,
            children
          ] }),
          /* @__PURE__ */ jsx("div", { className: classes.itemArrow, children: /* @__PURE__ */ jsx(RiArrowRightSLine, {}) })
        ] })
      }
    );
  }
  return /* @__PURE__ */ jsx(
    MenuItem$1,
    {
      className: classes.root,
      ...dataAttributes,
      href,
      textValue: typeof children === "string" ? children : void 0,
      ...restProps,
      onAction: () => {
        restProps.onAction?.();
        handleAction();
      },
      children: /* @__PURE__ */ jsxs("div", { className: classes.itemWrapper, children: [
        /* @__PURE__ */ jsxs("div", { className: classes.itemContent, children: [
          iconStart,
          children
        ] }),
        /* @__PURE__ */ jsx("div", { className: classes.itemArrow, children: /* @__PURE__ */ jsx(RiArrowRightSLine, {}) })
      ] })
    }
  );
};
const MenuListBoxItem = (props) => {
  const { ownProps, restProps } = useDefinition(
    MenuListBoxItemDefinition,
    props
  );
  const { classes, children } = ownProps;
  return /* @__PURE__ */ jsx(
    ListBoxItem,
    {
      textValue: typeof children === "string" ? children : void 0,
      className: classes.root,
      ...restProps,
      children: /* @__PURE__ */ jsx("div", { className: classes.itemWrapper, children: /* @__PURE__ */ jsxs("div", { className: classes.itemContent, children: [
        /* @__PURE__ */ jsx("div", { className: classes.check, children: /* @__PURE__ */ jsx(RiCheckLine, {}) }),
        children
      ] }) })
    }
  );
};
const MenuSection = (props) => {
  const { ownProps, restProps } = useDefinition(MenuSectionDefinition, props);
  const { classes, children, title } = ownProps;
  return /* @__PURE__ */ jsxs(MenuSection$1, { className: classes.root, ...restProps, children: [
    /* @__PURE__ */ jsx(Header, { className: classes.header, children: title }),
    children
  ] });
};
const MenuSeparator = (props) => {
  const { ownProps, restProps } = useDefinition(MenuSeparatorDefinition, props);
  return /* @__PURE__ */ jsx(Separator, { className: ownProps.classes.root, ...restProps });
};

export { Menu, MenuAutocomplete, MenuAutocompleteListbox, MenuItem, MenuListBox, MenuListBoxItem, MenuSection, MenuSeparator, MenuTrigger, SubmenuTrigger };
