// @backstage/ui/dist/components/Table/components/Row.esm.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { useTableOptions, Cell, Collection, Row as Row$1 } from 'react-aria-components';
import { Checkbox } from '../../Checkbox/Checkbox.esm.js';
import { useDefinition } from '../../../hooks/useDefinition/useDefinition.esm.js';
import '../../Checkbox/Checkbox.module.css.esm.js';
import { RowDefinition } from '../definition.esm.js';
import { isExternalLink } from '../../../utils/linkUtils.esm.js';
import clsx from 'clsx';
import { Flex } from '../../Flex/Flex.esm.js';
import '../../Flex/Flex.module.css.esm.js';

function Row(props) {
  const { ownProps, restProps, dataAttributes, analytics } = useDefinition(
    RowDefinition,
    props
  );
  const { classes, columns, children, href } = ownProps;
  const isExternal = isExternalLink(href);
  const hasInternalHref = !!href && !isExternal;
  const hasExternalHref = !!href && isExternal;
  const hasInteraction = !!restProps.onAction || !!href;
  const effectiveTarget = hasExternalHref ? "_blank" : restProps.target;
  const effectiveRel = effectiveTarget === "_blank" ? [
    .../* @__PURE__ */ new Set([
      "noopener",
      "noreferrer",
      ...restProps.rel?.split(/\s+/).filter(Boolean) ?? []
    ])
  ].join(" ") : restProps.rel;
  const handlePress = hasInteraction ? () => {
    restProps.onAction?.();
    if (href) {
      analytics.captureEvent("click", href, {
        attributes: { to: String(href) }
      });
    }
  } : void 0;
  let { selectionBehavior, selectionMode } = useTableOptions();
  const content = /* @__PURE__ */ jsxs(Fragment, { children: [
    selectionBehavior === "toggle" && selectionMode === "multiple" && /* @__PURE__ */ jsx(Cell, { className: clsx(classes.cell, classes.cellSelection), children: /* @__PURE__ */ jsx(Flex, { justify: "center", align: "center", children: /* @__PURE__ */ jsx(Checkbox, { slot: "selection", "aria-label": "Select row" }) }) }),
    /* @__PURE__ */ jsx(Collection, { items: columns, children })
  ] });
  return /* @__PURE__ */ jsx(
    Row$1,
    {
      href,
      ...restProps,
      ...dataAttributes,
      target: effectiveTarget,
      rel: effectiveRel,
      className: clsx(classes.root, restProps.className),
      "data-react-aria-pressable": hasInternalHref ? "true" : void 0,
      onAction: handlePress,
      children: content
    }
  );
}

export { Row };
