// @backstage/ui/dist/components/Table/components/TableHeader.esm.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useTableOptions, TableHeader as TableHeader$1, Collection } from 'react-aria-components';
import { Checkbox } from '../../Checkbox/Checkbox.esm.js';
import { useDefinition } from '../../../hooks/useDefinition/useDefinition.esm.js';
import '../../Checkbox/Checkbox.module.css.esm.js';
import { Column } from './Column.esm.js';
import { TableHeaderDefinition } from '../definition.esm.js';
import { Flex } from '../../Flex/Flex.esm.js';
import '../../Flex/Flex.module.css.esm.js';

const TableHeader = (props) => {
  let { selectionBehavior, selectionMode } = useTableOptions();
  const { ownProps, restProps } = useDefinition(TableHeaderDefinition, props);
  const { classes, columns, children } = ownProps;
  return /* @__PURE__ */ jsxs(TableHeader$1, { className: classes.root, ...restProps, children: [
    selectionBehavior === "toggle" && selectionMode === "multiple" && /* @__PURE__ */ jsx(
      Column,
      {
        width: 40,
        minWidth: 40,
        maxWidth: 40,
        className: classes.headSelection,
        children: /* @__PURE__ */ jsx(Flex, { justify: "center", align: "center", "aria-label": "Row selection", children: /* @__PURE__ */ jsx(Checkbox, { slot: "selection", "aria-label": "Select all" }) })
      }
    ),
    /* @__PURE__ */ jsx(Collection, { items: columns, children })
  ] });
};

export { TableHeader };
