// @backstage/ui/dist/components/Table/components/CellText.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { Text } from '../../Text/Text.esm.js';
import { useDefinition } from '../../../hooks/useDefinition/useDefinition.esm.js';
import '../../Text/Text.module.css.esm.js';
import { Link } from '../../Link/Link.esm.js';
import '../../Link/Link.module.css.esm.js';
import { Cell } from 'react-aria-components';
import { CellTextDefinition } from '../definition.esm.js';

const CellText = (props) => {
  const { ownProps, restProps } = useDefinition(CellTextDefinition, props);
  const { classes, title, description, color, leadingIcon, href } = ownProps;
  return /* @__PURE__ */ jsx(Cell, { className: classes.root, ...restProps, children: /* @__PURE__ */ jsxs("div", { className: classes.cellContentWrapper, children: [
    leadingIcon && /* @__PURE__ */ jsx("div", { className: classes.cellIcon, children: leadingIcon }),
    /* @__PURE__ */ jsxs("div", { className: classes.cellContent, children: [
      href ? /* @__PURE__ */ jsx(
        Link,
        {
          href,
          variant: "body-medium",
          color,
          truncate: true,
          title,
          children: title
        }
      ) : /* @__PURE__ */ jsx(
        Text,
        {
          as: "p",
          variant: "body-medium",
          color,
          truncate: true,
          title,
          children: title
        }
      ),
      description && /* @__PURE__ */ jsx(
        Text,
        {
          variant: "body-medium",
          color: "secondary",
          truncate: true,
          title: description,
          children: description
        }
      )
    ] })
  ] }) });
};
CellText.displayName = "CellText";

export { CellText };
