// @backstage/frontend-plugin-api/dist/blueprints/SignInPageBlueprint.esm.js
import { jsx } from 'react/jsx-runtime';
import { lazy } from 'react';
import '../wiring/coreExtensionData.esm.js';
import 'zod';
import 'zod-to-json-schema';
import { createExtensionDataRef } from '../wiring/createExtensionDataRef.esm.js';
import { createExtensionBlueprint } from '../wiring/createExtensionBlueprint.esm.js';
import { ExtensionBoundary } from '../components/ExtensionBoundary.esm.js';
import '../apis/definitions/AppTreeApi.esm.js';
import '../apis/definitions/auth.esm.js';
import '../apis/definitions/AlertApi.esm.js';
import '../apis/definitions/AppLanguageApi.esm.js';
import '../apis/definitions/AppThemeApi.esm.js';
import '../apis/definitions/SwappableComponentsApi.esm.js';
import '../apis/definitions/ConfigApi.esm.js';
import '../apis/definitions/DiscoveryApi.esm.js';
import '../apis/definitions/ErrorApi.esm.js';
import '../apis/definitions/FeatureFlagsApi.esm.js';
import '../apis/definitions/FetchApi.esm.js';
import '../apis/definitions/IconsApi.esm.js';
import '../apis/definitions/IdentityApi.esm.js';
import '../apis/definitions/DialogApi.esm.js';
import '../apis/definitions/OAuthRequestApi.esm.js';
import '../apis/definitions/RouteResolutionApi.esm.js';
import '../apis/definitions/StorageApi.esm.js';
import '../apis/definitions/AnalyticsApi.esm.js';
import '../apis/definitions/TranslationApi.esm.js';
import '@backstage/version-bridge';
import '@backstage/errors';
import '../components/AppNodeProvider.esm.js';
import '../components/DefaultSwappableComponents.esm.js';

const componentDataRef = createExtensionDataRef().with({ id: "core.sign-in-page.component" });
const SignInPageBlueprint = createExtensionBlueprint({
  kind: "sign-in-page",
  attachTo: { id: "app/root", input: "signInPage" },
  output: [componentDataRef],
  dataRefs: {
    component: componentDataRef
  },
  *factory({
    loader
  }, { node }) {
    const ExtensionComponent = lazy(
      () => loader().then((component) => ({ default: component }))
    );
    yield componentDataRef((props) => /* @__PURE__ */ jsx(ExtensionBoundary, { node, children: /* @__PURE__ */ jsx(ExtensionComponent, { ...props }) }));
  }
});

export { SignInPageBlueprint };
