// @backstage/plugin-catalog-react/dist/components/UnregisterEntityDialog/UnregisterEntityDialog.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { EntityRefLink } from '../EntityRefLink/EntityRefLink.esm.js';
import '@backstage/catalog-model';
import { useState, useCallback } from 'react';
import 'lodash/get';
import { makeStyles } from '@material-ui/core/styles';
import { useUnregisterEntityDialogState } from './useUnregisterEntityDialogState.esm.js';
import { useApi, configApiRef, alertApiRef } from '@backstage/core-plugin-api';
import { ResponseErrorPanel, Progress } from '@backstage/core-components';
import { assertError } from '@backstage/errors';
import { catalogReactTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';
import { Dialog, DialogHeader, DialogBody, DialogFooter, Button, Alert, Text, Box, Accordion, AccordionTrigger, AccordionPanel } from '@backstage/ui';

const useStyles = makeStyles({
  bodyContent: {
    overflowWrap: "break-word"
  }
});
function useUnregisterDialogHandlers(entity, onConfirm, onClose) {
  const alertApi = useApi(alertApiRef);
  const state = useUnregisterEntityDialogState(entity);
  const [busyAction, setBusyAction] = useState(
    null
  );
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const onUnregister = useCallback(async () => {
    if ("unregisterLocation" in state) {
      setBusyAction("unregister");
      try {
        await state.unregisterLocation();
        onConfirm();
      } catch (err) {
        assertError(err);
        alertApi.post({ message: err.message });
      } finally {
        setBusyAction(null);
      }
    }
  }, [alertApi, onConfirm, state]);
  const onDelete = useCallback(async () => {
    if ("deleteEntity" in state) {
      setBusyAction("delete");
      try {
        await state.deleteEntity();
        const entityName = entity.metadata.title ?? entity.metadata.name;
        onConfirm();
        alertApi.post({
          message: t("unregisterEntityDialog.deleteEntitySuccessMessage", {
            entityName
          }),
          severity: "success",
          display: "transient"
        });
      } catch (err) {
        assertError(err);
        alertApi.post({ message: err.message });
      } finally {
        setBusyAction(null);
      }
    }
  }, [alertApi, onConfirm, state, entity, t]);
  return { state, busyAction, onUnregister, onDelete, onClose };
}
function AdvancedDeleteAccordion({
  triggerTitle,
  description,
  onDelete,
  busyAction
}) {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ jsx(Box, { mt: "4", children: /* @__PURE__ */ jsxs(Accordion, { children: [
    /* @__PURE__ */ jsx(AccordionTrigger, { title: triggerTitle }),
    /* @__PURE__ */ jsxs(AccordionPanel, { children: [
      /* @__PURE__ */ jsx(Text, { as: "p", children: description }),
      /* @__PURE__ */ jsx(Box, { mt: "4", children: /* @__PURE__ */ jsx(
        Button,
        {
          variant: "primary",
          destructive: true,
          loading: busyAction === "delete",
          isDisabled: busyAction !== null && busyAction !== "delete",
          onPress: onDelete,
          children: t("unregisterEntityDialog.deleteButtonTitle")
        }
      ) })
    ] })
  ] }) });
}
function BootstrapBody({
  location,
  appTitle,
  onDelete,
  busyAction
}) {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(
      Alert,
      {
        status: "info",
        icon: true,
        description: t("unregisterEntityDialog.bootstrapState.title", {
          appTitle,
          location
        })
      }
    ),
    /* @__PURE__ */ jsx(
      AdvancedDeleteAccordion,
      {
        triggerTitle: t(
          "unregisterEntityDialog.bootstrapState.advancedOptions"
        ),
        description: t(
          "unregisterEntityDialog.bootstrapState.advancedDescription"
        ),
        onDelete,
        busyAction
      }
    )
  ] });
}
function OnlyDeleteBody() {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ jsx(Text, { as: "p", children: t("unregisterEntityDialog.onlyDeleteStateTitle") });
}
function UnregisterBody({
  state,
  appTitle,
  onDelete,
  busyAction
}) {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Text, { as: "p", children: t("unregisterEntityDialog.unregisterState.title") }),
    /* @__PURE__ */ jsx("ul", { children: state.colocatedEntities.map((e) => /* @__PURE__ */ jsx("li", { children: /* @__PURE__ */ jsx(EntityRefLink, { entityRef: e }) }, `${e.kind}:${e.namespace}/${e.name}`)) }),
    /* @__PURE__ */ jsx(Text, { as: "p", children: t("unregisterEntityDialog.unregisterState.subTitle") }),
    /* @__PURE__ */ jsx("ul", { children: /* @__PURE__ */ jsx("li", { children: state.location }) }),
    /* @__PURE__ */ jsx(Text, { as: "p", children: t("unregisterEntityDialog.unregisterState.description", {
      appTitle
    }) }),
    /* @__PURE__ */ jsx(
      AdvancedDeleteAccordion,
      {
        triggerTitle: t(
          "unregisterEntityDialog.unregisterState.advancedOptions"
        ),
        description: t(
          "unregisterEntityDialog.unregisterState.advancedDescription"
        ),
        onDelete,
        busyAction
      }
    )
  ] });
}
function useDialogContent(handlers, appTitle) {
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const { state, busyAction, onUnregister, onDelete } = handlers;
  switch (state.type) {
    case "loading":
      return { body: /* @__PURE__ */ jsx(Progress, {}), actionButton: null };
    case "error":
      return {
        body: /* @__PURE__ */ jsx(ResponseErrorPanel, { error: state.error }),
        actionButton: null
      };
    case "bootstrap":
      return {
        body: /* @__PURE__ */ jsx(
          BootstrapBody,
          {
            location: state.location,
            appTitle,
            onDelete,
            busyAction
          }
        ),
        actionButton: null
      };
    case "only-delete":
      return {
        body: /* @__PURE__ */ jsx(OnlyDeleteBody, {}),
        actionButton: /* @__PURE__ */ jsx(
          Button,
          {
            variant: "primary",
            destructive: true,
            loading: busyAction === "delete",
            isDisabled: busyAction !== null && busyAction !== "delete",
            onPress: onDelete,
            children: t("unregisterEntityDialog.deleteButtonTitle")
          }
        )
      };
    case "unregister":
      return {
        body: /* @__PURE__ */ jsx(
          UnregisterBody,
          {
            state,
            appTitle,
            onDelete,
            busyAction
          }
        ),
        actionButton: /* @__PURE__ */ jsx(
          Button,
          {
            variant: "primary",
            destructive: true,
            loading: busyAction === "unregister",
            isDisabled: busyAction !== null && busyAction !== "unregister",
            onPress: onUnregister,
            children: t("unregisterEntityDialog.unregisterState.unregisterButtonTitle")
          }
        )
      };
    default:
      return {
        body: /* @__PURE__ */ jsx(
          Alert,
          {
            status: "danger",
            description: t("unregisterEntityDialog.errorStateTitle")
          }
        ),
        actionButton: null
      };
  }
}
function DialogContents({
  entity,
  onConfirm,
  onClose
}) {
  const classes = useStyles();
  const { t } = useTranslationRef(catalogReactTranslationRef);
  const configApi = useApi(configApiRef);
  const appTitle = configApi.getOptionalString("app.title") ?? "Backstage";
  const handlers = useUnregisterDialogHandlers(entity, onConfirm, onClose);
  const { body, actionButton } = useDialogContent(handlers, appTitle);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(DialogHeader, { children: t("unregisterEntityDialog.title") }),
    /* @__PURE__ */ jsx(DialogBody, { className: classes.bodyContent, children: body }),
    /* @__PURE__ */ jsxs(DialogFooter, { children: [
      /* @__PURE__ */ jsx(Button, { variant: "secondary", onPress: onClose, children: t("unregisterEntityDialog.cancelButtonTitle") }),
      actionButton
    ] })
  ] });
}
const UnregisterEntityDialog = (props) => {
  const { open, onConfirm, onClose, entity } = props;
  return /* @__PURE__ */ jsx(
    Dialog,
    {
      isOpen: open,
      onOpenChange: (isOpen) => !isOpen && onClose(),
      width: 600,
      children: open && /* @__PURE__ */ jsx(
        DialogContents,
        {
          entity,
          onConfirm,
          onClose
        }
      )
    }
  );
};

export { UnregisterEntityDialog };
