// @backstage/ui/dist/components/Grid/Grid.esm.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import clsx from 'clsx';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { GridDefinition, GridItemDefinition } from './definition.esm.js';
import styles from './Grid.module.css.esm.js';
import { useBgProvider, BgProvider } from '../../hooks/useBg.esm.js';

const GridRoot = forwardRef((props, ref) => {
  const { bg: resolvedBg } = useBgProvider(props.bg);
  const { classNames, dataAttributes, utilityClasses, style, cleanedProps } = useStyles(GridDefinition, {
    columns: "auto",
    gap: "4",
    ...props,
    bg: resolvedBg
    // Use resolved bg for data attribute
  });
  const { className, bg, ...rest } = cleanedProps;
  const content = /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: clsx(
        classNames.root,
        utilityClasses,
        styles[classNames.root],
        className
      ),
      style,
      ...dataAttributes,
      ...rest
    }
  );
  return resolvedBg ? /* @__PURE__ */ jsx(BgProvider, { bg: resolvedBg, children: content }) : content;
});
const GridItem = forwardRef((props, ref) => {
  const { bg: resolvedBg } = useBgProvider(props.bg);
  const { classNames, dataAttributes, utilityClasses, style, cleanedProps } = useStyles(GridItemDefinition, {
    ...props,
    bg: resolvedBg
    // Use resolved bg for data attribute
  });
  const { className, bg, ...rest } = cleanedProps;
  const content = /* @__PURE__ */ jsx(
    "div",
    {
      ref,
      className: clsx(
        classNames.root,
        utilityClasses,
        styles[classNames.root],
        className
      ),
      style,
      ...dataAttributes,
      ...rest
    }
  );
  return resolvedBg ? /* @__PURE__ */ jsx(BgProvider, { bg: resolvedBg, children: content }) : content;
});
const Grid = {
  Root: GridRoot,
  Item: GridItem
};

export { Grid };
