// @backstage/ui/dist/components/Menu/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Menu.module.css.esm.js';

const menuPopoverClassNames = {
  root: "bui-MenuPopover",
  inner: "bui-MenuInner",
  content: "bui-MenuContent"
};
const menuAutocompleteClassNames = {
  ...menuPopoverClassNames,
  searchField: "bui-MenuSearchField",
  searchFieldInput: "bui-MenuSearchFieldInput",
  searchFieldClear: "bui-MenuSearchFieldClear"
};
const menuPopoverPropDefs = {
  placement: { default: "bottom start" },
  virtualized: { default: false },
  maxWidth: {},
  maxHeight: {},
  style: {},
  className: {}
};
const MenuDefinition = defineComponent()({
  styles,
  classNames: menuPopoverClassNames,
  propDefs: menuPopoverPropDefs
});
const MenuListBoxDefinition = defineComponent()({
  styles,
  classNames: menuPopoverClassNames,
  propDefs: {
    ...menuPopoverPropDefs,
    selectionMode: { default: "single" }
  }
});
const MenuAutocompleteDefinition = defineComponent()({
  styles,
  classNames: menuAutocompleteClassNames,
  propDefs: {
    ...menuPopoverPropDefs,
    placeholder: {}
  }
});
const MenuAutocompleteListboxDefinition = defineComponent()({
  styles,
  classNames: menuAutocompleteClassNames,
  propDefs: {
    ...menuPopoverPropDefs,
    placeholder: {},
    selectionMode: { default: "single" }
  }
});
const MenuItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-MenuItem",
    itemWrapper: "bui-MenuItemWrapper",
    itemContent: "bui-MenuItemContent",
    itemArrow: "bui-MenuItemArrow"
  },
  analytics: true,
  propDefs: {
    iconStart: {},
    children: {},
    color: { dataAttribute: true, default: "primary" },
    href: {},
    noTrack: {},
    className: {}
  }
});
const MenuListBoxItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-MenuItemListBox",
    itemWrapper: "bui-MenuItemWrapper",
    itemContent: "bui-MenuItemContent",
    check: "bui-MenuItemListBoxCheck"
  },
  propDefs: {
    children: {},
    className: {}
  }
});
const MenuSectionDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-MenuSection",
    header: "bui-MenuSectionHeader"
  },
  propDefs: {
    title: {},
    children: {},
    className: {}
  }
});
const MenuSeparatorDefinition = defineComponent()(
  {
    styles,
    classNames: {
      root: "bui-MenuSeparator"
    },
    propDefs: {
      className: {}
    }
  }
);
const MenuEmptyStateDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-MenuEmptyState"
  },
  propDefs: {}
});

export { MenuAutocompleteDefinition, MenuAutocompleteListboxDefinition, MenuDefinition, MenuEmptyStateDefinition, MenuItemDefinition, MenuListBoxDefinition, MenuListBoxItemDefinition, MenuSectionDefinition, MenuSeparatorDefinition };
