// @backstage/ui/dist/components/Table/hooks/useTable.esm.js
import { useRef, useMemo } from 'react';
import { useQueryState } from './useQueryState.esm.js';
import { useCompletePagination } from './useCompletePagination.esm.js';
import { useCursorPagination } from './useCursorPagination.esm.js';
import { useOffsetPagination } from './useOffsetPagination.esm.js';

function useTableProps(paginationResult, sortState, paginationOptions = {}) {
  const {
    showPageSizeOptions = true,
    pageSizeOptions,
    onPageSizeChange: onPageSizeChangeCallback,
    onNextPage: onNextPageCallback,
    onPreviousPage: onPreviousPageCallback,
    getLabel
  } = paginationOptions;
  const previousDataRef = useRef(paginationResult.data);
  if (paginationResult.data) {
    previousDataRef.current = paginationResult.data;
  }
  const displayData = paginationResult.data ?? previousDataRef.current;
  const isStale = paginationResult.loading && displayData !== void 0;
  const pagination = useMemo(
    () => ({
      type: "page",
      pageSize: paginationResult.pageSize,
      pageSizeOptions,
      offset: paginationResult.offset,
      totalCount: paginationResult.totalCount,
      hasNextPage: paginationResult.hasNextPage,
      hasPreviousPage: paginationResult.hasPreviousPage,
      onNextPage: () => {
        paginationResult.onNextPage();
        onNextPageCallback?.();
      },
      onPreviousPage: () => {
        paginationResult.onPreviousPage();
        onPreviousPageCallback?.();
      },
      onPageSizeChange: (size) => {
        paginationResult.onPageSizeChange(size);
        onPageSizeChangeCallback?.(size);
      },
      showPageSizeOptions,
      getLabel
    }),
    [
      paginationResult.pageSize,
      pageSizeOptions,
      paginationResult.offset,
      paginationResult.totalCount,
      paginationResult.hasNextPage,
      paginationResult.hasPreviousPage,
      paginationResult.onNextPage,
      paginationResult.onPreviousPage,
      paginationResult.onPageSizeChange,
      onNextPageCallback,
      onPreviousPageCallback,
      onPageSizeChangeCallback
    ]
  );
  return useMemo(
    () => ({
      data: displayData,
      loading: paginationResult.loading,
      isStale,
      error: paginationResult.error,
      pagination,
      sort: sortState
    }),
    [
      displayData,
      paginationResult.loading,
      isStale,
      paginationResult.error,
      pagination,
      showPageSizeOptions,
      getLabel,
      sortState
    ]
  );
}
function useTable(options) {
  const query = useQueryState(options);
  const initialModeRef = useRef(options.mode);
  if (initialModeRef.current !== options.mode) {
    throw new Error(
      `useTable mode cannot change from '${initialModeRef.current}' to '${options.mode}'. The mode must remain stable for the lifetime of the component.`
    );
  }
  let pagination;
  if (options.mode === "complete") {
    pagination = useCompletePagination(options, query);
  } else if (options.mode === "offset") {
    pagination = useOffsetPagination(options, query);
  } else if (options.mode === "cursor") {
    pagination = useCursorPagination(options, query);
  } else {
    throw new Error("Invalid mode");
  }
  const sortState = useMemo(
    () => ({ descriptor: query.sort, onSortChange: query.setSort }),
    [query.sort, query.setSort]
  );
  const tableProps = useTableProps(
    pagination,
    sortState,
    options.paginationOptions ?? {}
  );
  return {
    tableProps,
    reload: pagination.reload,
    filter: { value: query.filter, onChange: query.setFilter },
    search: { value: query.search, onChange: query.setSearch }
  };
}

export { useTable };
