// @backstage/ui/dist/components/Card/Card.esm.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { CardDefinition, CardHeaderDefinition, CardBodyDefinition, CardFooterDefinition } from './definition.esm.js';
import { Box } from '../Box/Box.esm.js';

const Card = forwardRef((props, ref) => {
  const { ownProps, restProps, dataAttributes } = useDefinition(
    CardDefinition,
    props
  );
  const { classes, children } = ownProps;
  return /* @__PURE__ */ jsx(
    Box,
    {
      bg: "neutral-auto",
      ref,
      className: classes.root,
      ...dataAttributes,
      ...restProps,
      children
    }
  );
});
Card.displayName = "Card";
const CardHeader = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(CardHeaderDefinition, props);
    const { classes, children } = ownProps;
    return /* @__PURE__ */ jsx("div", { ref, className: classes.root, ...restProps, children });
  }
);
CardHeader.displayName = "CardHeader";
const CardBody = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(CardBodyDefinition, props);
    const { classes, children } = ownProps;
    return /* @__PURE__ */ jsx("div", { ref, className: classes.root, ...restProps, children });
  }
);
CardBody.displayName = "CardBody";
const CardFooter = forwardRef(
  (props, ref) => {
    const { ownProps, restProps } = useDefinition(CardFooterDefinition, props);
    const { classes, children } = ownProps;
    return /* @__PURE__ */ jsx("div", { ref, className: classes.root, ...restProps, children });
  }
);
CardFooter.displayName = "CardFooter";

export { Card, CardBody, CardFooter, CardHeader };
