// @backstage/ui/dist/components/Card/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import 'react';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Card.module.css.esm.js';

const CardDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Card"
  },
  propDefs: {
    children: {},
    className: {}
  }
});
const CardHeaderDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-CardHeader"
  },
  propDefs: {
    children: {},
    className: {}
  }
});
const CardBodyDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-CardBody"
  },
  propDefs: {
    children: {},
    className: {}
  }
});
const CardFooterDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-CardFooter"
  },
  propDefs: {
    children: {},
    className: {}
  }
});

export { CardBodyDefinition, CardDefinition, CardFooterDefinition, CardHeaderDefinition };
