// @backstage/ui/dist/components/Grid/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './Grid.module.css.esm.js';

const GridDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-Grid"
  },
  bg: "provider",
  propDefs: {
    bg: { dataAttribute: true },
    children: {},
    className: {},
    style: {}
  },
  utilityProps: [
    "columns",
    "gap",
    "m",
    "mb",
    "ml",
    "mr",
    "mt",
    "mx",
    "my",
    "p",
    "pb",
    "pl",
    "pr",
    "pt",
    "px",
    "py"
  ]
});
const GridItemDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-GridItem"
  },
  bg: "provider",
  propDefs: {
    bg: { dataAttribute: true },
    children: {},
    className: {},
    style: {}
  },
  utilityProps: ["colSpan", "colEnd", "colStart", "rowSpan"]
});

export { GridDefinition, GridItemDefinition };
