// @backstage/ui/dist/hooks/useDefinition/helpers.esm.js
import { breakpoints } from '../useBreakpoint.esm.js';
import { utilityClassMap } from '../../utils/utilityClassMap.esm.js';

const namedBreakpoints = breakpoints.filter((b) => b.id !== "initial");
function isResponsiveObject(value) {
  return typeof value === "object" && value !== null && namedBreakpoints.some((b) => b.id in value);
}
function resolveResponsiveValue(value, breakpoint) {
  if (!isResponsiveObject(value)) {
    return value;
  }
  const index = breakpoints.findIndex((b) => b.id === breakpoint);
  for (let i = index; i >= 0; i--) {
    const key = breakpoints[i].id;
    if (key in value && value[key] !== void 0) {
      return value[key];
    }
  }
  for (let i = 0; i < breakpoints.length; i++) {
    const key = breakpoints[i].id;
    if (key in value && value[key] !== void 0) {
      return value[key];
    }
  }
  return value;
}
function resolveDefinitionProps(definition, props, breakpoint) {
  const ownPropKeys = new Set(Object.keys(definition.propDefs));
  const utilityPropKeys = new Set(definition.utilityProps ?? []);
  const ownPropsRaw = {};
  const restProps = {};
  for (const [key, value] of Object.entries(props)) {
    if (ownPropKeys.has(key)) {
      ownPropsRaw[key] = value;
    } else if (!utilityPropKeys.has(key)) {
      restProps[key] = value;
    }
  }
  const ownPropsResolved = {};
  for (const [key, config] of Object.entries(definition.propDefs)) {
    const rawValue = ownPropsRaw[key];
    const resolvedValue = resolveResponsiveValue(rawValue, breakpoint);
    const finalValue = resolvedValue ?? config.default;
    if (finalValue !== void 0) {
      ownPropsResolved[key] = finalValue;
    }
  }
  return { ownPropsResolved, restProps };
}
function processUtilityProps(props, utilityPropKeys) {
  const utilityClassList = [];
  const generatedStyle = {};
  const handleUtilityValue = (key, val, prefix = "") => {
    const utilityConfig = utilityClassMap[key];
    if (!utilityConfig) {
      return;
    }
    const values = utilityConfig.values;
    if (values.length > 0 && values.includes(val)) {
      const className = prefix ? `${prefix}${utilityConfig.class}-${val}` : `${utilityConfig.class}-${val}`;
      utilityClassList.push(className);
    } else if ("cssVar" in utilityConfig && utilityConfig.cssVar) {
      const cssVar = utilityConfig.cssVar;
      const cssVarKey = prefix ? `${cssVar}-${prefix.slice(0, -1)}` : cssVar;
      generatedStyle[cssVarKey] = val;
      const className = prefix ? `${prefix}${utilityConfig.class}` : utilityConfig.class;
      utilityClassList.push(className);
    }
  };
  for (const key of utilityPropKeys) {
    const value = props[key];
    if (value === void 0 || value === null) {
      continue;
    }
    if (typeof value === "object" && value !== null) {
      const breakpointValues = value;
      for (const bp in breakpointValues) {
        const prefix = bp === "initial" ? "" : `${bp}:`;
        handleUtilityValue(key, breakpointValues[bp], prefix);
      }
    } else {
      handleUtilityValue(key, value);
    }
  }
  return {
    utilityClasses: utilityClassList.join(" "),
    utilityStyle: generatedStyle
  };
}

export { processUtilityProps, resolveDefinitionProps, resolveResponsiveValue };
