// @backstage/ui/dist/components/Select/SelectListBox.esm.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { ListBox, ListBoxItem, Text } from 'react-aria-components';
import { RiCheckLine } from '@remixicon/react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { SelectListBoxDefinition } from './definition.esm.js';

const NoResults = () => {
  const { ownProps } = useDefinition(SelectListBoxDefinition, {});
  const { classes } = ownProps;
  return /* @__PURE__ */ jsx("div", { className: classes.noResults, children: "No results found." });
};
function SelectListBox(props) {
  const { ownProps } = useDefinition(SelectListBoxDefinition, props);
  const { classes, options } = ownProps;
  return /* @__PURE__ */ jsx(ListBox, { className: classes.root, renderEmptyState: () => /* @__PURE__ */ jsx(NoResults, {}), children: options?.map((option) => /* @__PURE__ */ jsxs(
    ListBoxItem,
    {
      id: option.value,
      textValue: option.label,
      className: classes.item,
      isDisabled: option.disabled,
      children: [
        /* @__PURE__ */ jsx("div", { className: classes.itemIndicator, children: /* @__PURE__ */ jsx(RiCheckLine, {}) }),
        /* @__PURE__ */ jsx(Text, { slot: "label", className: classes.itemLabel, children: option.label })
      ]
    },
    option.value
  )) });
}

export { SelectListBox };
