// @backstage/frontend-plugin-api/dist/index.esm.js
export { AnalyticsContext } from './analytics/AnalyticsContext.esm.js';
export { useAnalytics } from './analytics/useAnalytics.esm.js';
export { appTreeApiRef } from './apis/definitions/AppTreeApi.esm.js';
export { SessionState, atlassianAuthApiRef, bitbucketAuthApiRef, bitbucketServerAuthApiRef, githubAuthApiRef, gitlabAuthApiRef, googleAuthApiRef, microsoftAuthApiRef, oktaAuthApiRef, oneloginAuthApiRef, openshiftAuthApiRef, vmwareCloudAuthApiRef } from './apis/definitions/auth.esm.js';
export { alertApiRef } from './apis/definitions/AlertApi.esm.js';
export { appLanguageApiRef } from './apis/definitions/AppLanguageApi.esm.js';
export { appThemeApiRef } from './apis/definitions/AppThemeApi.esm.js';
export { swappableComponentsApiRef } from './apis/definitions/SwappableComponentsApi.esm.js';
export { configApiRef } from './apis/definitions/ConfigApi.esm.js';
export { discoveryApiRef } from './apis/definitions/DiscoveryApi.esm.js';
export { errorApiRef } from './apis/definitions/ErrorApi.esm.js';
export { FeatureFlagState, featureFlagsApiRef } from './apis/definitions/FeatureFlagsApi.esm.js';
export { fetchApiRef } from './apis/definitions/FetchApi.esm.js';
export { iconsApiRef } from './apis/definitions/IconsApi.esm.js';
export { identityApiRef } from './apis/definitions/IdentityApi.esm.js';
export { dialogApiRef } from './apis/definitions/DialogApi.esm.js';
export { oauthRequestApiRef } from './apis/definitions/OAuthRequestApi.esm.js';
export { routeResolutionApiRef } from './apis/definitions/RouteResolutionApi.esm.js';
export { storageApiRef } from './apis/definitions/StorageApi.esm.js';
export { analyticsApiRef } from './apis/definitions/AnalyticsApi.esm.js';
export { translationApiRef } from './apis/definitions/TranslationApi.esm.js';
export { useApi, useApiHolder, withApis } from './apis/system/useApi.esm.js';
export { createApiRef } from './apis/system/ApiRef.esm.js';
export { createApiFactory } from './apis/system/helpers.esm.js';
export { AnalyticsImplementationBlueprint } from './blueprints/AnalyticsImplementationBlueprint.esm.js';
export { ApiBlueprint } from './blueprints/ApiBlueprint.esm.js';
export { AppRootElementBlueprint } from './blueprints/AppRootElementBlueprint.esm.js';
export { AppRootWrapperBlueprint } from './blueprints/AppRootWrapperBlueprint.esm.js';
export { IconBundleBlueprint } from './blueprints/IconBundleBlueprint.esm.js';
export { NavContentBlueprint } from './blueprints/NavContentBlueprint.esm.js';
export { NavItemBlueprint } from './blueprints/NavItemBlueprint.esm.js';
export { PageBlueprint } from './blueprints/PageBlueprint.esm.js';
export { RouterBlueprint } from './blueprints/RouterBlueprint.esm.js';
export { SignInPageBlueprint } from './blueprints/SignInPageBlueprint.esm.js';
export { ThemeBlueprint } from './blueprints/ThemeBlueprint.esm.js';
export { TranslationBlueprint } from './blueprints/TranslationBlueprint.esm.js';
export { SwappableComponentBlueprint } from './blueprints/SwappableComponentBlueprint.esm.js';
export { ExtensionBoundary } from './components/ExtensionBoundary.esm.js';
export { createSwappableComponent } from './components/createSwappableComponent.esm.js';
export { useAppNode } from './components/AppNodeProvider.esm.js';
export { ErrorDisplay, NotFoundErrorPage, Progress } from './components/DefaultSwappableComponents.esm.js';
export { createRouteRef } from './routing/RouteRef.esm.js';
export { createSubRouteRef } from './routing/SubRouteRef.esm.js';
export { createExternalRouteRef } from './routing/ExternalRouteRef.esm.js';
export { useRouteRef } from './routing/useRouteRef.esm.js';
export { useRouteRefParams } from './routing/useRouteRefParams.esm.js';
export { createTranslationMessages } from './translation/TranslationMessages.esm.js';
export { createTranslationResource } from './translation/TranslationResource.esm.js';
export { createTranslationRef } from './translation/TranslationRef.esm.js';
export { useTranslationRef } from './translation/useTranslationRef.esm.js';
export { coreExtensionData } from './wiring/coreExtensionData.esm.js';
export { createExtension } from './wiring/createExtension.esm.js';
export { createExtensionInput } from './wiring/createExtensionInput.esm.js';
export { createExtensionDataRef } from './wiring/createExtensionDataRef.esm.js';
export { createFrontendPlugin } from './wiring/createFrontendPlugin.esm.js';
export { createFrontendModule } from './wiring/createFrontendModule.esm.js';
export { createFrontendFeatureLoader } from './wiring/createFrontendFeatureLoader.esm.js';
export { createExtensionBlueprint, createExtensionBlueprintParams } from './wiring/createExtensionBlueprint.esm.js';
