// @backstage/plugin-catalog-react/dist/alpha/blueprints/extensionData.esm.js
import { createExtensionDataRef } from '@backstage/frontend-plugin-api';

const entityContentTitleDataRef = createExtensionDataRef().with({
  id: "catalog.entity-content-title"
});
const entityContentIconDataRef = createExtensionDataRef().with({
  id: "catalog.entity-content-icon"
});
const entityFilterFunctionDataRef = createExtensionDataRef().with({ id: "catalog.entity-filter-function" });
const entityFilterExpressionDataRef = createExtensionDataRef().with({
  id: "catalog.entity-filter-expression"
});
const defaultEntityContentGroupDefinitions = {
  overview: {
    title: "Overview"
  },
  documentation: {
    title: "Documentation"
  },
  development: {
    title: "Development"
  },
  deployment: {
    title: "Deployment"
  },
  operation: {
    title: "Operation"
  },
  observability: {
    title: "Observability"
  }
};
const defaultEntityContentGroups = Object.fromEntries(
  Object.entries(defaultEntityContentGroupDefinitions).map(
    ([key, { title }]) => [key, title]
  )
);
const entityContentGroupDataRef = createExtensionDataRef().with({
  id: "catalog.entity-content-group"
});
const entityCardTypes = [
  "info",
  "content"
];
const entityCardTypeDataRef = createExtensionDataRef().with({
  id: "catalog.entity-card-type"
});

export { defaultEntityContentGroupDefinitions, defaultEntityContentGroups, entityCardTypeDataRef, entityCardTypes, entityContentGroupDataRef, entityContentIconDataRef, entityContentTitleDataRef, entityFilterExpressionDataRef, entityFilterFunctionDataRef };
