// @backstage/backend-app-api/dist/wiring/BackendInitializer.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var errors = require('@backstage/errors');
var DependencyGraph = require('../lib/DependencyGraph.cjs.js');
var ServiceRegistry = require('./ServiceRegistry.cjs.js');
var createInitializationResultCollector = require('./createInitializationResultCollector.cjs.js');
var helpers = require('./helpers.cjs.js');
var BackendStartupError = require('./BackendStartupError.cjs.js');
var createAllowBootFailurePredicate = require('./createAllowBootFailurePredicate.cjs.js');

const instanceRegistry = new class InstanceRegistry {
  #registered = false;
  #instances = /* @__PURE__ */ new Set();
  register(instance) {
    if (!this.#registered) {
      this.#registered = true;
      process.addListener("SIGTERM", this.#exitHandler);
      process.addListener("SIGINT", this.#exitHandler);
      process.addListener("beforeExit", this.#exitHandler);
    }
    this.#instances.add(instance);
  }
  unregister(instance) {
    this.#instances.delete(instance);
  }
  #exitHandler = async () => {
    try {
      const results = await Promise.allSettled(
        Array.from(this.#instances).map((b) => b.stop())
      );
      const errors = results.flatMap(
        (r) => r.status === "rejected" ? [r.reason] : []
      );
      if (errors.length > 0) {
        for (const error of errors) {
          console.error(error);
        }
        process.exit(1);
      } else {
        process.exit(0);
      }
    } catch (error) {
      console.error(error);
      process.exit(1);
    }
  };
}();
function createRootInstanceMetadataServiceFactory(rawRegistrations) {
  const installedPlugins = /* @__PURE__ */ new Map();
  const registrations = rawRegistrations.filter((registration) => registration.featureType === "registrations").flatMap((registration) => registration.getRegistrations());
  const plugins = registrations.filter(
    (registration) => registration.type === "plugin" || registration.type === "plugin-v1.1"
  );
  const modules = registrations.filter(
    (registration) => registration.type === "module" || registration.type === "module-v1.1"
  );
  for (const plugin of plugins) {
    const { pluginId } = plugin;
    if (!installedPlugins.get(pluginId)) {
      installedPlugins.set(pluginId, {
        pluginId,
        modules: []
      });
    }
  }
  for (const module of modules) {
    const { pluginId, moduleId } = module;
    const installedPlugin = installedPlugins.get(pluginId);
    if (installedPlugin) {
      installedPlugin.modules.push({
        moduleId
      });
    }
  }
  return backendPluginApi.createServiceFactory({
    service: backendPluginApi.coreServices.rootInstanceMetadata,
    deps: {},
    factory: async () => {
      const readonlyInstalledPlugins = helpers.deepFreeze([
        ...installedPlugins.values()
      ]);
      const instanceMetadata = {
        getInstalledPlugins: () => Promise.resolve(readonlyInstalledPlugins)
      };
      return instanceMetadata;
    }
  });
}
class BackendInitializer {
  #startPromise;
  #stopPromise;
  #registrations = new Array();
  #extensionPoints = /* @__PURE__ */ new Map();
  #serviceRegistry;
  #registeredFeatures = new Array();
  #registeredFeatureLoaders = new Array();
  #unhandledRejectionHandler;
  #uncaughtExceptionHandler;
  constructor(defaultApiFactories) {
    this.#serviceRegistry = ServiceRegistry.ServiceRegistry.create([...defaultApiFactories]);
  }
  async #getInitDeps(deps, resultCollector, pluginId, moduleId) {
    const result = /* @__PURE__ */ new Map();
    const missingRefs = /* @__PURE__ */ new Set();
    for (const [name, ref] of Object.entries(deps)) {
      const ep = this.#extensionPoints.get(ref.id);
      if (ep) {
        if (ep.pluginId !== pluginId) {
          throw new Error(
            `Illegal dependency: Module '${moduleId}' for plugin '${pluginId}' attempted to depend on extension point '${ref.id}' for plugin '${ep.pluginId}'. Extension points can only be used within their plugin's scope.`
          );
        }
        if (!moduleId) {
          throw new Error(
            `Rejected dependency on extension point ${ref.id} from outside of a module`
          );
        }
        result.set(
          name,
          ep.factory({
            reportModuleStartupFailure: ({ error }) => {
              resultCollector.amendPluginModuleResult(
                pluginId,
                moduleId,
                error
              );
            }
          })
        );
      } else {
        const impl = await this.#serviceRegistry.get(
          ref,
          pluginId
        );
        if (impl) {
          result.set(name, impl);
        } else {
          missingRefs.add(ref);
        }
      }
    }
    if (missingRefs.size > 0) {
      const missing = Array.from(missingRefs).join(", ");
      const target = moduleId ? `module '${moduleId}' for plugin '${pluginId}'` : `plugin '${pluginId}'`;
      throw new Error(
        `Service or extension point dependencies of ${target} are missing for the following ref(s): ${missing}`
      );
    }
    return Object.fromEntries(result);
  }
  add(feature) {
    if (this.#startPromise) {
      throw new Error("feature can not be added after the backend has started");
    }
    this.#registeredFeatures.push(Promise.resolve(feature));
  }
  #addFeature(feature) {
    if (isServiceFactory(feature)) {
      this.#serviceRegistry.add(feature);
    } else if (isBackendFeatureLoader(feature)) {
      this.#registeredFeatureLoaders.push(feature);
    } else if (isBackendRegistrations(feature)) {
      this.#registrations.push(feature);
    } else {
      throw new Error(
        `Failed to add feature, invalid feature ${JSON.stringify(feature)}`
      );
    }
  }
  async start() {
    if (this.#startPromise) {
      throw new Error("Backend has already started");
    }
    if (this.#stopPromise) {
      throw new Error("Backend has already stopped");
    }
    instanceRegistry.register(this);
    this.#startPromise = this.#doStart();
    return await this.#startPromise;
  }
  async #doStart() {
    this.#serviceRegistry.checkForCircularDeps();
    for (const feature of this.#registeredFeatures) {
      this.#addFeature(await feature);
    }
    await this.#applyBackendFeatureLoaders(this.#registeredFeatureLoaders);
    this.#serviceRegistry.add(
      createRootInstanceMetadataServiceFactory(this.#registrations)
    );
    if (process.env.NODE_ENV !== "test") {
      const rootLogger2 = await this.#serviceRegistry.get(
        backendPluginApi.coreServices.rootLogger,
        "root"
      );
      this.#unhandledRejectionHandler = (reason) => {
        rootLogger2?.child({ type: "unhandledRejection" })?.error("Unhandled rejection", reason);
      };
      this.#uncaughtExceptionHandler = (error) => {
        rootLogger2?.child({ type: "uncaughtException" })?.error("Uncaught exception", error);
      };
      process.on("unhandledRejection", this.#unhandledRejectionHandler);
      process.on("uncaughtException", this.#uncaughtExceptionHandler);
    }
    await this.#serviceRegistry.initializeEagerServicesWithScope("root");
    const rootConfig = await this.#serviceRegistry.get(
      backendPluginApi.coreServices.rootConfig,
      "root"
    );
    const rootLogger = await this.#serviceRegistry.get(
      backendPluginApi.coreServices.rootLogger,
      "root"
    );
    const allRegistrations = this.#registrations.flatMap(
      (f) => f.getRegistrations()
    );
    const allPluginIds = [
      ...new Set(
        allRegistrations.flatMap(
          (r) => "pluginId" in r && typeof r.pluginId === "string" ? [r.pluginId] : []
        )
      )
    ];
    const resultCollector = createInitializationResultCollector.createInitializationResultCollector({
      pluginIds: allPluginIds,
      logger: rootLogger,
      allowBootFailurePredicate: createAllowBootFailurePredicate.createAllowBootFailurePredicate(rootConfig)
    });
    const { pluginInits, moduleInits } = this.#enumerateRegistrations(
      allRegistrations,
      resultCollector
    );
    await Promise.all(
      [...pluginInits.keys()].map(async (pluginId) => {
        try {
          await this.#serviceRegistry.initializeEagerServicesWithScope(
            "plugin",
            pluginId
          );
          const modules = moduleInits.get(pluginId);
          if (modules) {
            const tree = DependencyGraph.DependencyGraph.fromIterable(
              Array.from(modules).map(([moduleId, moduleInit]) => ({
                value: { moduleId, moduleInit },
                // Relationships are reversed at this point since we're only interested in the extension points.
                // If a modules provides extension point A we want it to be initialized AFTER all modules
                // that depend on extension point A, so that they can provide their extensions.
                consumes: Array.from(moduleInit.provides).map((p) => p.id),
                provides: Array.from(moduleInit.consumes).map((c) => c.id)
              }))
            );
            const circular = tree.detectCircularDependency();
            if (circular) {
              throw new errors.ConflictError(
                `Circular dependency detected for modules of plugin '${pluginId}', ${circular.map(({ moduleId }) => `'${moduleId}'`).join(" -> ")}`
              );
            }
            await tree.parallelTopologicalTraversal(
              async ({ moduleId, moduleInit }) => {
                try {
                  const moduleDeps = await this.#getInitDeps(
                    moduleInit.init.deps,
                    resultCollector,
                    pluginId,
                    moduleId
                  );
                  await moduleInit.init.func(moduleDeps);
                  resultCollector.onPluginModuleResult(pluginId, moduleId);
                } catch (error) {
                  errors.assertError(error);
                  resultCollector.onPluginModuleResult(
                    pluginId,
                    moduleId,
                    error
                  );
                }
              }
            );
          }
          const pluginInit = pluginInits.get(pluginId);
          if (pluginInit) {
            const pluginDeps = await this.#getInitDeps(
              pluginInit.init.deps,
              resultCollector,
              pluginId
            );
            await pluginInit.init.func(pluginDeps);
          }
          resultCollector.onPluginResult(pluginId);
          const lifecycleService2 = await this.#getPluginLifecycleImpl(pluginId);
          await lifecycleService2.startup();
        } catch (error) {
          errors.assertError(error);
          resultCollector.onPluginResult(pluginId, error);
        }
      })
    ).catch((error) => {
      throw new errors.ForwardedError(
        "Unexpected uncaught backend startup error",
        error
      );
    });
    const result = resultCollector.finalize();
    if (result.outcome === "failure") {
      throw new BackendStartupError.BackendStartupError(result);
    }
    const lifecycleService = await this.#getRootLifecycleImpl();
    await lifecycleService.startup();
    return { result };
  }
  #enumerateRegistrations(allRegistrations, resultCollector) {
    const pluginInits = /* @__PURE__ */ new Map();
    const moduleInits = /* @__PURE__ */ new Map();
    for (const r of allRegistrations) {
      const addedExtensionPointIds = [];
      try {
        const provides = /* @__PURE__ */ new Set();
        if (r.type === "plugin" || r.type === "module") {
          for (const [extRef, extImpl] of r.extensionPoints) {
            if (this.#extensionPoints.has(extRef.id)) {
              throw new Error(
                `ExtensionPoint with ID '${extRef.id}' is already registered`
              );
            }
            this.#extensionPoints.set(extRef.id, {
              pluginId: r.pluginId,
              factory: () => extImpl
            });
            addedExtensionPointIds.push(extRef.id);
            provides.add(extRef);
          }
        } else if (r.type === "plugin-v1.1" || r.type === "module-v1.1") {
          for (const extReg of r.extensionPoints) {
            if (this.#extensionPoints.has(extReg.extensionPoint.id)) {
              throw new Error(
                `ExtensionPoint with ID '${extReg.extensionPoint.id}' is already registered`
              );
            }
            this.#extensionPoints.set(extReg.extensionPoint.id, {
              pluginId: r.pluginId,
              factory: extReg.factory
            });
            addedExtensionPointIds.push(extReg.extensionPoint.id);
            provides.add(extReg.extensionPoint);
          }
        }
        if (r.type === "plugin" || r.type === "plugin-v1.1") {
          if (pluginInits.has(r.pluginId)) {
            throw new Error(`Plugin '${r.pluginId}' is already registered`);
          }
          pluginInits.set(r.pluginId, {
            provides,
            consumes: new Set(Object.values(r.init.deps)),
            init: r.init
          });
        } else if (r.type === "module" || r.type === "module-v1.1") {
          let modules = moduleInits.get(r.pluginId);
          if (!modules) {
            modules = /* @__PURE__ */ new Map();
            moduleInits.set(r.pluginId, modules);
          }
          if (modules.has(r.moduleId)) {
            throw new Error(
              `Module '${r.moduleId}' for plugin '${r.pluginId}' is already registered`
            );
          }
          modules.set(r.moduleId, {
            provides,
            consumes: new Set(Object.values(r.init.deps)),
            init: r.init
          });
        } else {
          throw new Error(`Invalid registration type '${r.type}'`);
        }
      } catch (error) {
        errors.assertError(error);
        for (const id of addedExtensionPointIds) {
          this.#extensionPoints.delete(id);
        }
        if ("pluginId" in r && "moduleId" in r) {
          resultCollector.onPluginModuleResult(r.pluginId, r.moduleId, error);
        } else if ("pluginId" in r) {
          pluginInits.delete(r.pluginId);
          moduleInits.delete(r.pluginId);
          resultCollector.onPluginResult(r.pluginId, error);
        } else {
          throw error;
        }
      }
    }
    return { pluginInits, moduleInits };
  }
  // It's fine to call .stop() multiple times, which for example can happen with manual stop + process exit
  async stop() {
    instanceRegistry.unregister(this);
    if (!this.#stopPromise) {
      this.#stopPromise = this.#doStop();
    }
    await this.#stopPromise;
  }
  async #doStop() {
    if (!this.#startPromise) {
      return;
    }
    try {
      await this.#startPromise;
    } catch (error) {
    }
    const rootLifecycleService = await this.#getRootLifecycleImpl();
    await rootLifecycleService.beforeShutdown();
    const allPlugins = /* @__PURE__ */ new Set();
    for (const feature of this.#registrations) {
      for (const r of feature.getRegistrations()) {
        if (r.type === "plugin" || r.type === "plugin-v1.1") {
          allPlugins.add(r.pluginId);
        }
      }
    }
    await Promise.allSettled(
      [...allPlugins].map(async (pluginId) => {
        const lifecycleService = await this.#getPluginLifecycleImpl(pluginId);
        await lifecycleService.shutdown();
      })
    );
    await rootLifecycleService.shutdown();
    if (this.#unhandledRejectionHandler) {
      process.off("unhandledRejection", this.#unhandledRejectionHandler);
      this.#unhandledRejectionHandler = void 0;
    }
    if (this.#uncaughtExceptionHandler) {
      process.off("uncaughtException", this.#uncaughtExceptionHandler);
      this.#uncaughtExceptionHandler = void 0;
    }
  }
  // Bit of a hacky way to grab the lifecycle services, potentially find a nicer way to do this
  async #getRootLifecycleImpl() {
    const lifecycleService = await this.#serviceRegistry.get(
      backendPluginApi.coreServices.rootLifecycle,
      "root"
    );
    const service = lifecycleService;
    if (service && typeof service.startup === "function" && typeof service.shutdown === "function") {
      return service;
    }
    throw new Error("Unexpected root lifecycle service implementation");
  }
  async #getPluginLifecycleImpl(pluginId) {
    const lifecycleService = await this.#serviceRegistry.get(
      backendPluginApi.coreServices.lifecycle,
      pluginId
    );
    const service = lifecycleService;
    if (service && typeof service.startup === "function" && typeof service.shutdown === "function") {
      return service;
    }
    throw new Error("Unexpected plugin lifecycle service implementation");
  }
  async #applyBackendFeatureLoaders(loaders) {
    const servicesAddedByLoaders = /* @__PURE__ */ new Map();
    for (const loader of loaders) {
      const deps = /* @__PURE__ */ new Map();
      const missingRefs = /* @__PURE__ */ new Set();
      for (const [name, ref] of Object.entries(loader.deps ?? {})) {
        if (ref.scope !== "root") {
          throw new Error(
            `Feature loaders can only depend on root scoped services, but '${name}' is scoped to '${ref.scope}'. Offending loader is ${loader.description}`
          );
        }
        const impl = await this.#serviceRegistry.get(
          ref,
          "root"
        );
        if (impl) {
          deps.set(name, impl);
        } else {
          missingRefs.add(ref);
        }
      }
      if (missingRefs.size > 0) {
        const missing = Array.from(missingRefs).join(", ");
        throw new Error(
          `No service available for the following ref(s): ${missing}, depended on by feature loader ${loader.description}`
        );
      }
      const result = await loader.loader(Object.fromEntries(deps)).then((features) => features.map(helpers.unwrapFeature)).catch((error) => {
        throw new errors.ForwardedError(
          `Feature loader ${loader.description} failed`,
          error
        );
      });
      let didAddServiceFactory = false;
      const newLoaders = new Array();
      for await (const feature of result) {
        if (isBackendFeatureLoader(feature)) {
          newLoaders.push(feature);
        } else {
          if (isServiceFactory(feature) && !feature.service.multiton) {
            const conflictingLoader = servicesAddedByLoaders.get(
              feature.service.id
            );
            if (conflictingLoader) {
              throw new Error(
                `Duplicate service implementations provided for ${feature.service.id} by both feature loader ${loader.description} and feature loader ${conflictingLoader.description}`
              );
            }
            if (!this.#serviceRegistry.hasBeenAdded(feature.service)) {
              didAddServiceFactory = true;
              servicesAddedByLoaders.set(feature.service.id, loader);
              this.#addFeature(feature);
            }
          } else {
            this.#addFeature(feature);
          }
        }
      }
      if (didAddServiceFactory) {
        this.#serviceRegistry.checkForCircularDeps();
      }
      if (newLoaders.length > 0) {
        await this.#applyBackendFeatureLoaders(newLoaders);
      }
    }
  }
}
function toInternalBackendFeature(feature) {
  if (feature.$$type !== "@backstage/BackendFeature") {
    throw new Error(`Invalid BackendFeature, bad type '${feature.$$type}'`);
  }
  const internal = feature;
  if (internal.version !== "v1") {
    throw new Error(
      `Invalid BackendFeature, bad version '${internal.version}'`
    );
  }
  return internal;
}
function isServiceFactory(feature) {
  const internal = toInternalBackendFeature(feature);
  if (internal.featureType === "service") {
    return true;
  }
  return "service" in internal;
}
function isBackendRegistrations(feature) {
  const internal = toInternalBackendFeature(feature);
  if (internal.featureType === "registrations") {
    return true;
  }
  return "getRegistrations" in internal;
}
function isBackendFeatureLoader(feature) {
  return toInternalBackendFeature(feature).featureType === "loader";
}

exports.BackendInitializer = BackendInitializer;
