// @backstage/plugin-catalog-react/dist/components/EntityDataTable/columnFactories.esm.js
import { jsx } from 'react/jsx-runtime';
import { RELATION_PART_OF, RELATION_OWNED_BY } from '@backstage/catalog-model';
import { CellText, Column, Cell } from '@backstage/ui';
import { EntityRefLink } from '../EntityRefLink/EntityRefLink.esm.js';
import { EntityRefLinks } from '../EntityRefLink/EntityRefLinks.esm.js';
import { humanizeEntityRef } from '../EntityRefLink/humanize.esm.js';
import { EntityTableColumnTitle } from '../EntityTable/TitleColumn.esm.js';
import { getEntityRelations } from '../../utils/getEntityRelations.esm.js';

const columnFactories = Object.freeze({
  createEntityRefColumn(options) {
    const isRowHeader = options.isRowHeader ?? true;
    return {
      id: "name",
      label: "Name",
      header: () => /* @__PURE__ */ jsx(Column, { id: "name", isRowHeader, allowsSorting: true, children: /* @__PURE__ */ jsx(EntityTableColumnTitle, { translationKey: "name" }) }),
      isRowHeader,
      isSortable: true,
      cell: (entity) => /* @__PURE__ */ jsx(Cell, { children: /* @__PURE__ */ jsx(
        EntityRefLink,
        {
          entityRef: entity,
          defaultKind: options.defaultKind,
          title: entity.metadata?.title
        }
      ) }),
      sortValue: (entity) => entity.metadata?.title || humanizeEntityRef(entity, { defaultKind: options.defaultKind })
    };
  },
  createEntityRelationColumn(options) {
    return {
      id: options.id,
      label: options.id.charAt(0).toUpperCase() + options.id.slice(1),
      header: () => /* @__PURE__ */ jsx(Column, { id: options.id, allowsSorting: true, children: /* @__PURE__ */ jsx(EntityTableColumnTitle, { translationKey: options.translationKey }) }),
      isSortable: true,
      cell: (entity) => /* @__PURE__ */ jsx(Cell, { children: /* @__PURE__ */ jsx(
        EntityRefLinks,
        {
          entityRefs: getEntityRelations(
            entity,
            options.relation,
            options.filter
          ),
          defaultKind: options.defaultKind
        }
      ) }),
      sortValue: (entity) => getEntityRelations(entity, options.relation, options.filter).map((r) => humanizeEntityRef(r, { defaultKind: options.defaultKind })).join(", ")
    };
  },
  createOwnerColumn() {
    return columnFactories.createEntityRelationColumn({
      id: "owner",
      translationKey: "owner",
      relation: RELATION_OWNED_BY,
      defaultKind: "group"
    });
  },
  createSystemColumn() {
    return columnFactories.createEntityRelationColumn({
      id: "system",
      translationKey: "system",
      relation: RELATION_PART_OF,
      defaultKind: "system",
      filter: { kind: "system" }
    });
  },
  createDomainColumn() {
    return columnFactories.createEntityRelationColumn({
      id: "domain",
      translationKey: "domain",
      relation: RELATION_PART_OF,
      defaultKind: "domain",
      filter: { kind: "domain" }
    });
  },
  createMetadataDescriptionColumn() {
    return {
      id: "description",
      label: "Description",
      header: () => /* @__PURE__ */ jsx(Column, { id: "description", allowsSorting: true, children: /* @__PURE__ */ jsx(EntityTableColumnTitle, { translationKey: "description" }) }),
      isSortable: true,
      cell: (entity) => /* @__PURE__ */ jsx(CellText, { title: entity.metadata.description ?? "" }),
      sortValue: (entity) => entity.metadata.description ?? ""
    };
  },
  createSpecTypeColumn() {
    return {
      id: "type",
      label: "Type",
      header: () => /* @__PURE__ */ jsx(Column, { id: "type", allowsSorting: true, children: /* @__PURE__ */ jsx(EntityTableColumnTitle, { translationKey: "type" }) }),
      isSortable: true,
      cell: (entity) => /* @__PURE__ */ jsx(
        CellText,
        {
          title: entity.spec?.type ?? ""
        }
      ),
      sortValue: (entity) => entity.spec?.type ?? ""
    };
  },
  createSpecLifecycleColumn() {
    return {
      id: "lifecycle",
      label: "Lifecycle",
      header: () => /* @__PURE__ */ jsx(Column, { id: "lifecycle", allowsSorting: true, children: /* @__PURE__ */ jsx(EntityTableColumnTitle, { translationKey: "lifecycle" }) }),
      isSortable: true,
      cell: (entity) => /* @__PURE__ */ jsx(
        CellText,
        {
          title: entity.spec?.lifecycle ?? ""
        }
      ),
      sortValue: (entity) => entity.spec?.lifecycle ?? ""
    };
  }
});

export { columnFactories };
