// @backstage/cli-common/dist/paths.cjs.js
'use strict';

var fs = require('node:fs');
var node_path = require('node:path');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

function findRootPath(searchDir, filterFunc) {
  let path = searchDir;
  for (let i = 0; i < 1e3; i++) {
    const packagePath = node_path.resolve(path, "package.json");
    const exists = fs__default.default.existsSync(packagePath);
    if (exists && filterFunc(packagePath)) {
      return path;
    }
    const newPath = node_path.dirname(path);
    if (newPath === path) {
      return void 0;
    }
    path = newPath;
  }
  throw new Error(
    `Iteration limit reached when searching for root package.json at ${searchDir}`
  );
}
function findOwnRootDir(ownDir) {
  const isLocal = fs__default.default.existsSync(node_path.resolve(ownDir, "src"));
  if (!isLocal) {
    throw new Error(
      "Tried to access monorepo package root dir outside of Backstage repository"
    );
  }
  const rootDir = findRootPath(ownDir, (pkgJsonPath) => {
    try {
      const content = fs__default.default.readFileSync(pkgJsonPath, "utf8");
      const data = JSON.parse(content);
      return Boolean(data.workspaces);
    } catch (error) {
      throw new Error(
        `Failed to read package.json at '${pkgJsonPath}', ${error}`
      );
    }
  });
  if (!rootDir) {
    throw new Error(`No monorepo root found when searching from '${ownDir}'`);
  }
  return rootDir;
}
const dirCache = /* @__PURE__ */ new Map();
class OwnPathsImpl {
  static #instanceCache = /* @__PURE__ */ new Map();
  static find(searchDir) {
    const dir = OwnPathsImpl.findDir(searchDir);
    let instance = OwnPathsImpl.#instanceCache.get(dir);
    if (!instance) {
      instance = new OwnPathsImpl(dir);
      OwnPathsImpl.#instanceCache.set(dir, instance);
    }
    return instance;
  }
  static findDir(searchDir) {
    const visited = [];
    let dir = searchDir;
    for (let i = 0; i < 1e3; i++) {
      const cached = dirCache.get(dir);
      if (cached !== void 0) {
        for (const d of visited) {
          dirCache.set(d, cached);
        }
        return cached;
      }
      visited.push(dir);
      if (fs__default.default.existsSync(node_path.resolve(dir, "package.json"))) {
        for (const d of visited) {
          dirCache.set(d, dir);
        }
        return dir;
      }
      const newDir = node_path.dirname(dir);
      if (newDir === dir) {
        break;
      }
      dir = newDir;
    }
    throw new Error(
      `No package.json found while searching for package root of ${searchDir}`
    );
  }
  #dir;
  #rootDir;
  constructor(dir) {
    this.#dir = dir;
  }
  get dir() {
    return this.#dir;
  }
  get rootDir() {
    this.#rootDir ??= findOwnRootDir(this.#dir);
    return this.#rootDir;
  }
  resolve = (...paths) => {
    return node_path.resolve(this.#dir, ...paths);
  };
  resolveRoot = (...paths) => {
    return node_path.resolve(this.rootDir, ...paths);
  };
}
exports.targetPathsOverride = void 0;
function setTargetPathsOverride(override) {
  exports.targetPathsOverride = override;
}
class TargetPathsImpl {
  #cwd;
  #dir;
  #rootDir;
  get dir() {
    if (exports.targetPathsOverride) {
      return exports.targetPathsOverride.dir;
    }
    const cwd = process.cwd();
    if (this.#dir !== void 0 && this.#cwd === cwd) {
      return this.#dir;
    }
    this.#cwd = cwd;
    this.#rootDir = void 0;
    this.#dir = fs__default.default.realpathSync(cwd).replace(/^[a-z]:/, (str) => str.toLocaleUpperCase("en-US"));
    return this.#dir;
  }
  get rootDir() {
    if (exports.targetPathsOverride) {
      return exports.targetPathsOverride.rootDir;
    }
    const dir = this.dir;
    if (this.#rootDir !== void 0) {
      return this.#rootDir;
    }
    this.#rootDir = findRootPath(dir, (path) => {
      try {
        const content = fs__default.default.readFileSync(path, "utf8");
        const data = JSON.parse(content);
        return Boolean(data.workspaces);
      } catch (error) {
        throw new Error(
          `Failed to parse package.json file while searching for root, ${error}`
        );
      }
    }) ?? dir;
    return this.#rootDir;
  }
  resolve = (...paths) => {
    if (exports.targetPathsOverride) {
      return exports.targetPathsOverride.resolve(...paths);
    }
    return node_path.resolve(this.dir, ...paths);
  };
  resolveRoot = (...paths) => {
    if (exports.targetPathsOverride) {
      return exports.targetPathsOverride.resolveRoot(...paths);
    }
    return node_path.resolve(this.rootDir, ...paths);
  };
}
const targetPaths = new TargetPathsImpl();
function findOwnPaths(searchDir) {
  return OwnPathsImpl.find(searchDir);
}
function findPaths(searchDir) {
  const own = findOwnPaths(searchDir);
  return {
    get ownDir() {
      return own.dir;
    },
    get ownRoot() {
      return own.rootDir;
    },
    get targetDir() {
      return targetPaths.dir;
    },
    get targetRoot() {
      return targetPaths.rootDir;
    },
    resolveOwn: own.resolve,
    resolveOwnRoot: own.resolveRoot,
    resolveTarget: targetPaths.resolve,
    resolveTargetRoot: targetPaths.resolveRoot
  };
}
const BACKSTAGE_JSON = "backstage.json";

exports.BACKSTAGE_JSON = BACKSTAGE_JSON;
exports.findOwnPaths = findOwnPaths;
exports.findOwnRootDir = findOwnRootDir;
exports.findPaths = findPaths;
exports.findRootPath = findRootPath;
exports.setTargetPathsOverride = setTargetPathsOverride;
exports.targetPaths = targetPaths;
