// @backstage/ui/dist/components/TagGroup/TagGroup.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { Tag as Tag$1, Button, TagGroup as TagGroup$1, TagList } from 'react-aria-components';
import { forwardRef } from 'react';
import { RiCloseCircleLine } from '@remixicon/react';
import clsx from 'clsx';
import { useStyles } from '../../hooks/useStyles.esm.js';
import { TagGroupDefinition } from './definition.esm.js';
import { createRoutingRegistration } from '../InternalLinkProvider/InternalLinkProvider.esm.js';
import styles from './TagGroup.module.css.esm.js';

const { RoutingProvider, useRoutingRegistrationEffect } = createRoutingRegistration();
const TagGroup = (props) => {
  const { classNames, cleanedProps } = useStyles(TagGroupDefinition, props);
  const { items, children, renderEmptyState, ...rest } = cleanedProps;
  return /* @__PURE__ */ jsx(RoutingProvider, { children: /* @__PURE__ */ jsx(
    TagGroup$1,
    {
      className: clsx(classNames.group, styles[classNames.group]),
      ...rest,
      children: /* @__PURE__ */ jsx(
        TagList,
        {
          className: clsx(classNames.list, styles[classNames.list]),
          items,
          renderEmptyState,
          children
        }
      )
    }
  ) });
};
const Tag = forwardRef(
  (props, ref) => {
    const { classNames, cleanedProps } = useStyles(TagGroupDefinition, {
      size: "small",
      ...props
    });
    const { children, className, icon, size, href, ...rest } = cleanedProps;
    const textValue = typeof children === "string" ? children : void 0;
    useRoutingRegistrationEffect(href);
    return /* @__PURE__ */ jsx(
      Tag$1,
      {
        ref,
        textValue,
        className: clsx(classNames.tag, styles[classNames.tag], className),
        "data-size": size,
        href,
        ...rest,
        children: ({ allowsRemoving }) => /* @__PURE__ */ jsxs(Fragment, { children: [
          icon && /* @__PURE__ */ jsx(
            "span",
            {
              className: clsx(classNames.tagIcon, styles[classNames.tagIcon]),
              children: icon
            }
          ),
          children,
          allowsRemoving && /* @__PURE__ */ jsx(
            Button,
            {
              className: clsx(
                classNames.tagRemoveButton,
                styles[classNames.tagRemoveButton]
              ),
              slot: "remove",
              children: /* @__PURE__ */ jsx(RiCloseCircleLine, { size: 16 })
            }
          )
        ] })
      }
    );
  }
);

export { Tag, TagGroup };
