// @backstage/ui/dist/components/PluginHeader/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './PluginHeader.module.css.esm.js';

const PluginHeaderDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-PluginHeader",
    toolbar: "bui-PluginHeaderToolbar",
    toolbarWrapper: "bui-PluginHeaderToolbarWrapper",
    toolbarContent: "bui-PluginHeaderToolbarContent",
    toolbarControls: "bui-PluginHeaderToolbarControls",
    toolbarIcon: "bui-PluginHeaderToolbarIcon",
    toolbarName: "bui-PluginHeaderToolbarName",
    tabs: "bui-PluginHeaderTabsWrapper"
  },
  propDefs: {
    icon: {},
    title: {},
    titleLink: {},
    customActions: {},
    tabs: {},
    onTabSelectionChange: {},
    className: {}
  }
});

export { PluginHeaderDefinition };
