// @backstage/core-plugin-api/dist/extensions/useElementFilter.esm.js
import { useMemo, Children, isValidElement, Fragment } from 'react';
import { getComponentData } from './componentData.esm.js';
import { useApi, featureFlagsApiRef } from '@backstage/frontend-plugin-api';
import '../apis/definitions/AnalyticsApi.esm.js';

function selectChildren(rootNode, featureFlagsApi, selector, strictError) {
  return Children.toArray(rootNode).flatMap((node) => {
    if (!isValidElement(node)) {
      return [];
    }
    if (node.type === Fragment) {
      return selectChildren(
        node.props.children,
        featureFlagsApi,
        selector,
        strictError
      );
    }
    if (getComponentData(node, "core.featureFlagged")) {
      const props = node.props;
      const isEnabled = "with" in props ? featureFlagsApi.isActive(props.with) : !featureFlagsApi.isActive(props.without);
      if (isEnabled) {
        return selectChildren(
          node.props.children,
          featureFlagsApi,
          selector,
          strictError
        );
      }
      return [];
    }
    if (selector === void 0 || selector(node)) {
      return [node];
    }
    if (strictError) {
      throw new Error(strictError);
    }
    return selectChildren(
      node.props.children,
      featureFlagsApi,
      selector,
      strictError
    );
  });
}
class Collection {
  node;
  featureFlagsApi;
  constructor(node, featureFlagsApi) {
    this.node = node;
    this.featureFlagsApi = featureFlagsApi;
  }
  selectByComponentData(query) {
    const selection = selectChildren(
      this.node,
      this.featureFlagsApi,
      (node) => getComponentData(node, query.key) !== void 0,
      query.withStrictError
    );
    return new Collection(selection, this.featureFlagsApi);
  }
  findComponentData(query) {
    const selection = selectChildren(
      this.node,
      this.featureFlagsApi,
      (node) => getComponentData(node, query.key) !== void 0
    );
    return selection.map((node) => getComponentData(node, query.key)).filter((data) => data !== void 0);
  }
  getElements() {
    return selectChildren(this.node, this.featureFlagsApi);
  }
}
function useElementFilter(node, filterFn, dependencies = []) {
  const featureFlagsApi = useApi(featureFlagsApiRef);
  const elements = new Collection(node, featureFlagsApi);
  return useMemo(() => filterFn(elements), [node, ...dependencies]);
}

export { useElementFilter };
