// @backstage/ui/dist/components/TablePagination/definition.esm.js
import 'react/jsx-runtime';
import 'clsx';
import '../../hooks/useBreakpoint.esm.js';
import '../../hooks/useBg.esm.js';
import '../../hooks/useDefinition/helpers.esm.js';
import '../../analytics/useAnalytics.esm.js';
import { defineComponent } from '../../hooks/useDefinition/defineComponent.esm.js';
import styles from './TablePagination.module.css.esm.js';

const DEFAULT_PAGE_SIZE_OPTIONS = [
  { label: "Show 5 results", value: 5 },
  { label: "Show 10 results", value: 10 },
  { label: "Show 20 results", value: 20 },
  { label: "Show 30 results", value: 30 },
  { label: "Show 40 results", value: 40 },
  { label: "Show 50 results", value: 50 }
];
const TablePaginationDefinition = defineComponent()({
  styles,
  classNames: {
    root: "bui-TablePagination",
    left: "bui-TablePaginationLeft",
    right: "bui-TablePaginationRight",
    select: "bui-TablePaginationSelect"
  },
  propDefs: {
    pageSize: {},
    pageSizeOptions: { default: DEFAULT_PAGE_SIZE_OPTIONS },
    offset: {},
    totalCount: {},
    hasNextPage: {},
    hasPreviousPage: {},
    onNextPage: {},
    onPreviousPage: {},
    onPageSizeChange: {},
    showPageSizeOptions: { default: true },
    getLabel: {}
  }
});

export { TablePaginationDefinition };
