// @backstage/ui/dist/components/TagGroup/TagGroup.esm.js
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { Tag as Tag$1, Button, TagGroup as TagGroup$1, TagList } from 'react-aria-components';
import { forwardRef } from 'react';
import { RiCloseCircleLine } from '@remixicon/react';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import { TagDefinition, TagGroupDefinition } from './definition.esm.js';
import { getNodeText } from '../../analytics/getNodeText.esm.js';

const TagGroup = (props) => {
  const { ownProps, restProps } = useDefinition(TagGroupDefinition, props);
  const { classes, items, children, renderEmptyState } = ownProps;
  return /* @__PURE__ */ jsx(TagGroup$1, { className: classes.root, ...restProps, children: /* @__PURE__ */ jsx(
    TagList,
    {
      className: classes.list,
      items,
      renderEmptyState,
      children
    }
  ) });
};
const Tag = forwardRef((props, ref) => {
  const { ownProps, restProps, dataAttributes, analytics } = useDefinition(
    TagDefinition,
    props
  );
  const { classes, children, icon, href } = ownProps;
  const textValue = typeof children === "string" ? children : void 0;
  const handlePress = () => {
    if (href) {
      const text = props["aria-label"] ?? textValue ?? getNodeText(children) ?? String(href);
      analytics.captureEvent("click", text, {
        attributes: { to: String(href) }
      });
    }
  };
  return /* @__PURE__ */ jsx(
    Tag$1,
    {
      ref,
      textValue,
      className: classes.root,
      href,
      ...dataAttributes,
      ...restProps,
      onPress: (e) => {
        restProps.onPress?.(e);
        handlePress();
      },
      children: ({ allowsRemoving }) => /* @__PURE__ */ jsxs(Fragment, { children: [
        icon && /* @__PURE__ */ jsx("span", { className: classes.icon, children: icon }),
        children,
        allowsRemoving && /* @__PURE__ */ jsx(Button, { className: classes.removeButton, slot: "remove", children: /* @__PURE__ */ jsx(RiCloseCircleLine, { size: 16 }) })
      ] })
    }
  );
});

export { Tag, TagGroup };
