// @backstage/ui/dist/components/Header/Header.esm.js
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { Text } from '../Text/Text.esm.js';
import { useDefinition } from '../../hooks/useDefinition/useDefinition.esm.js';
import '../Text/Text.module.css.esm.js';
import { RiArrowRightSLine } from '@remixicon/react';
import { Tabs, TabList, Tab } from '../Tabs/Tabs.esm.js';
import '../Tabs/Tabs.module.css.esm.js';
import { HeaderDefinition } from './definition.esm.js';
import { Container } from '../Container/Container.esm.js';
import '../Container/Container.module.css.esm.js';
import { Link } from '../Link/Link.esm.js';
import '../Link/Link.module.css.esm.js';

const Header = (props) => {
  const { ownProps } = useDefinition(HeaderDefinition, props);
  const { classes, title, tabs, customActions, breadcrumbs } = ownProps;
  return /* @__PURE__ */ jsxs(Container, { className: classes.root, children: [
    /* @__PURE__ */ jsxs("div", { className: classes.content, children: [
      /* @__PURE__ */ jsxs("div", { className: classes.breadcrumbs, children: [
        breadcrumbs && breadcrumbs.map((breadcrumb) => /* @__PURE__ */ jsxs(Fragment, { children: [
          /* @__PURE__ */ jsx(
            Link,
            {
              href: breadcrumb.href,
              variant: "title-small",
              weight: "bold",
              color: "secondary",
              truncate: true,
              style: { maxWidth: "240px" },
              standalone: true,
              children: breadcrumb.label
            }
          ),
          /* @__PURE__ */ jsx(RiArrowRightSLine, { size: 16, color: "var(--bui-fg-secondary)" })
        ] }, breadcrumb.label)),
        /* @__PURE__ */ jsx(Text, { variant: "title-small", weight: "bold", as: "h2", children: title })
      ] }),
      /* @__PURE__ */ jsx("div", { className: classes.controls, children: customActions })
    ] }),
    tabs && /* @__PURE__ */ jsx("div", { className: classes.tabsWrapper, children: /* @__PURE__ */ jsx(Tabs, { children: /* @__PURE__ */ jsx(TabList, { children: tabs.map((tab) => /* @__PURE__ */ jsx(
      Tab,
      {
        id: tab.id,
        href: tab.href,
        matchStrategy: tab.matchStrategy,
        children: tab.label
      },
      tab.id
    )) }) }) })
  ] });
};
const HeaderPage = Header;

export { Header, HeaderPage };
